### Parameters to Set - this is for the user if they want to use this Markdown file

```{r, echo=FALSE}
run_number <- 1 # different runs will be saved in different paths (same run number will overwrite the original file)

#Initial Number of each feature after feature selection
n_lncs <- 500 
ngenes <- 500
n_mirna <- 500
nprobes <- 500
n_clinical <- 18

#How many features you want the autoencoder to reduce the feature space of each to
meth_features <- 30
gene_features <- 30
mirna_features <- 30
lnc_features <- 30
fullAE_features <- 120


#Do you want to try just linear feature selection (no autoencoder)?
linear_feature_selection <- "no"

#Do you want to compare results of the autoencoder to PCA? (Not included in dissertation)
PCA <- "no"

#Do you want to try another round of feature selection after the autoencoder to select the most important features? (Not included in)
another_round_of_featureselection <- "no"

#Will allow you to save feature space visualizations to the file path
visualizations <- "no"

#Which autoencoder type do you want?
autoencoder_type <- "basic"
# "basic", "denoising_gaussian", "stacked_denoising_zeros" "sparse","denoising_gaussian_stacked"

#Do you want to compare one overall autoencoder with the individual autoencoders?
full_AE <- "no"

#This will be the run info that will be incorporated into the file name later (to make sure you are not overwriting the same file when you change one single piece of information)
runinfo <- paste0("Multi", n_lncs, ngenes, n_mirna, nprobes, meth_features, gene_features, mirna_features, lnc_features, linear_feature_selection,p_value, PCA,autoencoder_type, full_AE, PCA, run_number)
```


## Loading Packages 

```{r, warning=FALSE, echo = FALSE, message=FALSE, results='hide'}
library(survivalmodels)
library(SummarizedExperiment)
library(TCGAbiolinks)
library(AnnotationDbi)
library(GenomicFeatures)
library(ruta)
library(tidyverse)
library(dplyr)
library(tibble)
library(keras)
library(h2o)
library(caret)
library(readr)
library(data.table)
library(impute)
library(e1071)
library(purrr)
library(MethylMix)
library(ELMER)
library(MultiAssayExperiment)
library(GenomicRanges)
library(org.Hs.eg.db)
library(compound.Cox)
library(survminer)
library(survival)
library(ranger)
library(survivalmodels)
```


### Clinical Data Processing

```{r, echo=FALSE}
#Clinical data downloaded using TCGA biolinks - can be found in the download script
load(file="~/desktop/Multimodal/LUSC_Clin_Harm.RData")
load(file="~/desktop/Multimodal/LUAD_Clin_Harm.RData")

LUSCclin <- bind_rows(LUSCclin, LUADclin)

LUSCclin_half <- LUSCclin %>% 
  dplyr::select(c(barcode, patient, shortLetterCode, treatments, days_to_last_follow_up, days_to_death, vital_status))

LUSCclin_other <- LUSCclin %>% 
  dplyr::select(c(age_at_index, pack_years_smoked, cigarettes_per_day, shortest_dimension, intermediate_dimension, longest_dimension))

LUSCclin_other <- LUSCclin_other %>% mutate(volume=shortest_dimension*intermediate_dimension*longest_dimension)

LUSCclin_other <- LUSCclin_other %>% dplyr::select(-c(shortest_dimension,intermediate_dimension,longest_dimension))

LUSCclin_quantitative <- LUSCclin %>% 
  dplyr::select(c(ajcc_pathologic_stage, tissue_or_organ_of_origin, prior_treatment, prior_malignancy, ajcc_pathologic_n, ajcc_pathologic_m,ajcc_pathologic_t, race, gender, synchronous_malignancy, name))

#Getting rid of variable categories without enough examples
for (x in 1:nrow(LUSCclin_quantitative)) {
  if(LUSCclin_quantitative$tissue_or_organ_of_origin[x] != "Upper lobe, lung" & LUSCclin_quantitative$tissue_or_organ_of_origin[x] != "Lower lobe, lung") {
    LUSCclin_quantitative$tissue_or_organ_of_origin[x] <- "Lung"
  }
     if(LUSCclin_quantitative$prior_treatment[x]=="Not Reported") {
    LUSCclin_quantitative$prior_treatment[x] <- NA
     }
   if(LUSCclin_quantitative$prior_malignancy[x]=="not reported") {
    LUSCclin_quantitative$prior_malignancy[x] <- NA
   }
     if(LUSCclin_quantitative$synchronous_malignancy[x]=="Not Reported") {
    LUSCclin_quantitative$synchronous_malignancy[x] <- NA
     }
      if(!is.na(LUSCclin_quantitative$ajcc_pathologic_n[x]) & LUSCclin_quantitative$ajcc_pathologic_n[x]=="NX") {
    LUSCclin_quantitative$ajcc_pathologic_n[x] <- NA
      }
      if(!is.na(LUSCclin_quantitative$ajcc_pathologic_m[x]) &LUSCclin_quantitative$ajcc_pathologic_m[x]=="MX") {
    LUSCclin_quantitative$ajcc_pathologic_m[x] <- NA
      }
        if(LUSCclin_quantitative$race[x]=="american indian or alaska native") {
    LUSCclin_quantitative$race[x] <- NA
          }
        if(LUSCclin_quantitative$ajcc_pathologic_t[x]=="TX") {
            LUSCclin_quantitative$ajcc_pathologic_t[x] <- NA
        }

}

# Changing categorical variables to numeric before scaling
LUSCclin_quantitative <- sapply(LUSCclin_quantitative, function(x) if(class(x)=="character") { as.numeric(as.factor(x))  })
LUSCclin_quantitative <- as.data.frame(LUSCclin_quantitative)

LUSC_scaled <- cbind(LUSCclin_quantitative, LUSCclin_other)
LUSC_scaled <- scale(LUSC_scaled)
LUSC_scaled_together <- cbind(LUSC_scaled, LUSCclin_half)

#only taking patients who have tumors and not 'NT' which is normal tissue
LUSCclin_normal <- LUSC_scaled_together  %>%
  filter(shortLetterCode=='NT')
normalpatients1 <- LUSCclin_normal$barcode
LUSCclin_lung <- LUSC_scaled_together %>%
  filter(shortLetterCode=='TP')


#LIMITING OPTIONS TO ONLY THOSE patients in the multimodal analysis - people who have samples across all data types (732 in all)
barcodes_both <- LUSCclin_lung$barcode
load(file="~/desktop/Multimodal/multimodal_barcodes.Rdata")
LUSCclin_lung$barcode <- substring(LUSCclin_lung$barcode,1,19)
LUSCclin_lung <- LUSCclin_lung %>% filter(barcode %in% multimodal_barcodes)


#Getting censored survival times - getting rid of examples with no data
LUSCclin_lung$censor_time1 <- LUSCclin_lung$days_to_death
for (x in 1:nrow(LUSCclin_lung)) {
  if (is.na(LUSCclin_lung$censor_time1[x])) {
  LUSCclin_lung$censor_time1[x] <- LUSCclin_lung$days_to_last_follow_up[x]
  }
}

#Removing those with no data or censor time of zero
removal_na_indices <- which(is.na(LUSCclin_lung$censor_time1))
LUSCclin_lung <- LUSCclin_lung[-c(removal_na_indices), ]
LUSCclin_lung <- LUSCclin_lung %>% filter(censor_time1 > 1)

#Removing any duplicate patients
LUSCclin_lung <- distinct(LUSCclin_lung, barcode, .keep_all = TRUE)
LUSCclin_lung <- LUSCclin_lung[order(LUSCclin_lung$barcode),]
```


### Setting Unified Testing/Training and Run Info

```{r, echo=FALSE}
#Making sure that it's the same each time by using indices and setting a seed for reproducibility

train_set_size <- 585 #80/20 split
test_set_size <- 147

# File I took from the clinical data modality of all of the barcodes and disease type for reference to make sure I get an even split of each cancer type in test/train
load(file="~/desktop/Multimodal/barcode_disease_mapping.Rdata")
LUAD_indices <- which(barcode_disease_mapping$name=="Lung Adenocarcinoma") #408 LUAD
LUSC_indices <- which(barcode_disease_mapping$name=="Lung Squamous Cell Carcinoma") #324 LUSC

#Fraction of cases of each type
LUAD_fraction <- length(LUAD_indices)/(length(LUSC_indices)+length(LUAD_indices))
LUSC_fraction <- 1-LUAD_fraction

#Number of cases in testing and training dataset
LUAD_training_num <- round((LUAD_fraction*train_set_size),0)
LUSC_training_num <- round(LUSC_fraction*train_set_size,0)

#Sampling LUAD Dataset
set.seed(10)
LUAD_indices_sample <- sample(LUAD_indices)
LUAD_train_sample <- LUAD_indices_sample[1:LUAD_training_num]
LUAD_test_sample <- LUAD_indices_sample[(LUAD_training_num+1):length(LUAD_indices_sample)]

#Sampling LUSC Dataset
set.seed(10)
LUSC_indices_sample <- sample(LUSC_indices)
LUSC_train_sample <- LUSC_indices_sample[1:LUSC_training_num]
LUSC_test_sample <- LUSC_indices_sample[(LUSC_training_num+1):length(LUSC_indices_sample)]


#Combining Train and Test Indexes for Each
train_indices <- c(LUSC_train_sample, LUAD_train_sample)
test_indices <- c(LUSC_test_sample, LUAD_test_sample)

#Starting Early Stopping for Training to Prevent Overfitting
early_stop <- keras::callback_early_stopping(monitor = "val_loss", 
                                      min_delta = 0.001, 
                                      patience = 5,
                                      restore_best_weights = TRUE,
                                      verbose = 1)
```


# LNC RNA Modality

```{r, echo=FALSE}
#List of lncRNAs to extract downloaded from Li et al. (2020) - see dissertation
load(file="~/desktop/Multimodal/LNCRNAS/barcodes_lna.Rdata")
load(file="~/desktop/Multimodal/LNCRNAS/all_lncs.Rdata")

#Clinical Data to get Barcodes - extracted from clinical data modality
load(file="~/desktop/Multimodal/LUSC_Clin_Harm.RData")
load(file="~/desktop/Multimodal/LUAD_Clin_Harm.RData")

#Scaling the data
all_sig_features <- scale(all_sig_features)

#Making sure only the barcodes in the clinical data (already filtered) are in the LNC RNA data
all_sig_features <- as.data.frame(cbind(all_sig_features, barcodes_both))
colnames(all_sig_features)[ncol(all_sig_features)] <- "barcode"
all_sig_features <- merge(all_sig_features, LUSCclin_lung, by="barcode")
LUSC_lncRNA_process <- distinct(all_sig_features, barcode, .keep_all = TRUE)
```


### Splitting Training and Testing

```{r, echo=FALSE}
# Splitting into training and testing by the indices created above
LUSC_lncRNA_process_test <- LUSC_lncRNA_process[test_indices, ]
LUSC_lncRNA_process <- LUSC_lncRNA_process[train_indices,]

#Saving the barcodes from each
lncrna_barcode_train <- LUSC_lncRNA_process$barcode
lncrna_barcode_test <- LUSC_lncRNA_process_test$barcode

#Taking the output data
vital_vectorlncs <- LUSC_lncRNA_process$vital_status
censor_time1 <- LUSC_lncRNA_process$censor_time1

# only taking LNC RNA features (which start with E)
LUSC_lncRNA_process <- LUSC_lncRNA_process%>% dplyr::select(starts_with("E"))
LUSC_lncRNA_process_test<-LUSC_lncRNA_process_test %>% dplyr::select(starts_with("E"))

#Making sure they are type numeric
LUSC_lncRNA_process <- as.data.frame(sapply(LUSC_lncRNA_process, as.numeric))
LUSC_lncRNA_process_test <- as.data.frame(sapply(LUSC_lncRNA_process_test, as.numeric))
```

# Linear Feature Selection

```{r, echo=FALSE}
y <- (as.numeric(factor(vital_vectorlncs))-1)
association_mat <- as.matrix(LUSC_lncRNA_process)

#If linear_feature_selection is "yes" - then only use linear feature selection and take the top 30 features
if (linear_feature_selection == "yes") {
  set.seed(1)
  associat <- uni.selection(t.vec = censor_time1, d.vec = y, X.mat=association_mat,
                          P.value = 0.4, randomize=FALSE, K=5)
  associat$P <- associat$P[1:lnc_features]
  col_filtered <- rownames(as.data.frame(associat$P))
  LUSC_lncRNA_process <- LUSC_lncRNA_process %>% dplyr::select(all_of(col_filtered))
  LUSC_lncRNA_process_test <- LUSC_lncRNA_process_test %>% dplyr::select(all_of(col_filtered))
  
}

#If not, extract the top 500 features and utilize the autoencoder below
if (linear_feature_selection != "yes") {
  set.seed(1)
associat <- uni.selection(t.vec = censor_time1, d.vec = y, X.mat=association_mat,
                          P.value = 0.4, randomize=FALSE, K=5)
associat$P <- associat$P[1:n_lncs]
col_filtered <- rownames(as.data.frame(associat$P))
LUSC_lncRNA_process <- LUSC_lncRNA_process %>% dplyr::select(all_of(col_filtered))
LUSC_lncRNA_process_test <- LUSC_lncRNA_process_test %>% dplyr::select(all_of(col_filtered))
}

#Making sure no duplicate column names
LUSC_lncRNA_process <- LUSC_lncRNA_process[, !duplicated(colnames(LUSC_lncRNA_process))]
LUSC_lncRNA_process_test <- LUSC_lncRNA_process_test[, !duplicated(colnames(LUSC_lncRNA_process_test))]
LUSC_lncRNA_process_test <- as.matrix(LUSC_lncRNA_process_test)

#Extracting for altogether AE (if fullAE == "yes" in the beginning, if not this won't matter)
lncrna_fullAE <- LUSC_lncRNA_process
lncrna_fullAE_test <- LUSC_lncRNA_process_test

lncrna_fullAE$barcode <- lncrna_barcode_train
lncrna_fullAE_test <- as.data.frame(cbind(lncrna_fullAE_test, lncrna_barcode_test))
colnames(lncrna_fullAE_test)[ncol(lncrna_fullAE_test)] <- "barcode"
```


### Denoising Autoencoder (zeros)

```{r, echo=FALSE}
if (autoencoder_type == "stacked_denoising_zeros") {
  
if (n_lncs <= 502 & n_lncs>60) {

# Corruption by adding random zeros function
corrupt_with_ones <- function(x) {
  n_to_sample <- floor(length(x) * .3)
  elements_to_corrupt <- sample(seq_along(x), n_to_sample, replace = FALSE)
  x[elements_to_corrupt] <- 0
  return(x)
}

#Creating corrupted dataframe
inputs_currupted_ones <- LUSC_lncRNA_process %>%
  as.data.frame() %>%
  purrr::map_df(corrupt_with_ones)

#Creating matrices for training and testing
features <- as.matrix(LUSC_lncRNA_process)
inputs_currupted_ones <- as.matrix(inputs_currupted_ones)

#Creating the autoencoder model (with "tanh" activation function)
model1 <- keras_model_sequential()
model1 %>%
  layer_dense(units = lnc_features, activation = "tanh", name = "BottleNeck", input_shape = ncol(inputs_currupted_ones)) %>%
  layer_dense(units = ncol(inputs_currupted_ones))

#Mean squared error loss and adam optimizer (used for all autoencoder types)
model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

#Training with 100 epochs, 0.1 split, but early stopping if validation isn't improving
model1 %>% keras::fit(
  x = inputs_currupted_ones,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

#Evaluating performance on the training data
mse.ae1 <- evaluate(model1, features,inputs_currupted_ones)

#Getting values from the bottleneck layer
intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_lncRNA_auto <- predict(intermediate_layer_model1, inputs_currupted_ones)

}

}

#If linear feature selection is desired (or the number of features is too small to use an autoencoder), this skips the steps and renames the columns of the new matrix
if (n_lncs <= 50 | linear_feature_selection=="yes") {
  LUSC_lncRNA_auto <-  LUSC_lncRNA_process
  for (i in 1:ncol(LUSC_lncRNA_auto)) {
    colnames(LUSC_lncRNA_auto)[i] <- paste0("V",i)
  }
     LUSC_lncRNA_auto_output <- LUSC_lncRNA_process_test
      for (i in 1:ncol(LUSC_lncRNA_auto_output)) {
    colnames(LUSC_lncRNA_auto_output)[i] <- paste0("V",i)
  }
}
```

### Testing Different Autoencoder Types - Sparse, Denoising Gaussian, Variational (didn't have time to include in paper), Stacked Denoising Gaussian

```{r, echo=FALSE}
if (autoencoder_type == "sparse") {
  
features <- as.matrix(LUSC_lncRNA_process)

#Creating sparse model with regularization penalty and chosen activation functions
model1 <- keras_model_sequential()
model1 %>%
  layer_dense(units = lnc_features, activation = "relu", name = "BottleNeck", input_shape = ncol(features), activity_regularizer=regularizer_l1(10e-5)) %>%
  layer_dense(units = ncol(features), activation='sigmoid')

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

#Fitting the model using the same parameters
model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

#Evaluating and getting the bottleneck layer output, same as above
mse.ae1 <- evaluate(model1, features,features)
intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_lncRNA_auto <- predict(intermediate_layer_model1, features)
first_AE_RMSE <- evaluate(model1,LUSC_lncRNA_process_test, LUSC_lncRNA_process_test)

}

#Variational

if (autoencoder_type == "variational") {

features <- as.matrix(LUSC_lncRNA_process)
  
#Particular inputs
enc_input <- ncol(features)
latent_size <- lnc_features
epsilon_std <- 1.0

enc_input <- layer_input(shape = c(ncol(features)))
layer_one <- layer_dense(enc_input, units=256, activation = "relu")
z_mean <- layer_dense(layer_one, latent_size, name = "BottleNeck")
z_log_var <- layer_dense(layer_one, latent_size)
 
encoder <- keras_model(enc_input, z_mean)
summary(encoder)

sampling <- function(arg){
  z_mean <- arg[, 1:(latent_size)]
  z_log_var <- arg[, (latent_size + 1):(2 * latent_size)]
  epsilon <- k_random_normal(shape = c(k_shape(z_mean)[[1]]), mean=0)
  z_mean + k_exp(z_log_var/2)*epsilon
}

z <- layer_concatenate(list(z_mean, z_log_var)) %>% 
  layer_lambda(sampling)


decoder_layer <- layer_dense(units = 256, activation = "relu")
decoder_mean <- layer_dense(units = ncol(features), activation = "sigmoid")
h_decoded <- decoder_layer(z)
x_decoded_mean <- decoder_mean(h_decoded)
 
model1 <- keras_model(enc_input, x_decoded_mean)

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% fit(
  features, features, 
  epochs = 100, 
  validation_split=0.1
)


mse.ae1 <- evaluate(model1, features,features)


intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_lncRNA_auto <- predict(intermediate_layer_model1, features)
first_AE_RMSE <- evaluate(model1,LUSC_lncRNA_process_test, LUSC_lncRNA_process_test)

}

#Denoising (Gaussian Noise)

if (autoencoder_type=="denoising_gaussian") {

features <- as.matrix(LUSC_lncRNA_process)

model1 <- keras_model_sequential()
model1 %>%
  layer_gaussian_noise(stddev = 1, input_shape = ncol(features)) %>% #Adding noise
  layer_dense(units = lnc_features, activation = "tanh", name = "BottleNeck") %>%
  layer_dense(units = ncol(features))

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,features)
intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_lncRNA_auto <- predict(intermediate_layer_model1, features)

}


if (autoencoder_type=="basic") {

features <- as.matrix(LUSC_lncRNA_process)

model1 <- keras_model_sequential()
model1 %>%
  layer_dense(units = lnc_features, activation = "tanh", name = "BottleNeck", input_shape = ncol(features)) %>%
  layer_dense(units = ncol(features))

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,features)
intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_lncRNA_auto <- predict(intermediate_layer_model1, features)
}

if (autoencoder_type=="denoising_gaussian_stacked") {

features <- as.matrix(LUSC_lncRNA_process)

#Extra layers since it's a stacked version
model1 <- keras_model_sequential()
model1 %>%
  layer_gaussian_noise(stddev = 1, input_shape = ncol(features)) %>%
  layer_dense(units = 200, activation = "tanh") %>%
  layer_dense(units = lnc_features, activation = "tanh", name = "BottleNeck") %>%
  layer_dense(units = 200, activation = "tanh") %>%
  layer_dense(units = ncol(features))

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,features)

intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_lncRNA_auto <- predict(intermediate_layer_model1, features)
}
```

### Evaluating Autoencoders on Testing Data

```{r, echo=FALSE}
#only run if you ran an autoencoder in the first place
if (linear_feature_selection != "yes") {

#Making sure it's numeric
if (class(LUSC_lncRNA_process_test[1,1])=="character") {
LUSC_lncRNA_process_test <- as.data.frame(sapply(LUSC_lncRNA_process_test, as.numeric))
}

if (n_lncs <= 502 & n_lncs>60) {
first_AE_RMSE <- evaluate(model1,LUSC_lncRNA_process_test, LUSC_lncRNA_process_test)
LUSC_lncRNA_auto_output <-predict(intermediate_layer_model1,LUSC_lncRNA_process_test)
}

}

if (n_lncs <= 60) {
  LUSC_lncRNA_auto_output <-  LUSC_lncRNA_process_test
  for (i in 1:ncol(LUSC_lncRNA_auto_output)) {
    colnames(LUSC_lncRNA_auto_output)[i] <- paste0("V",i)
  }
}
```

### Comparing to PCA - Totally Optional (This analysis not included in manuscript)

```{r, echo=FALSE}
if (PCA == "yes") {

#Training Set
pc <- prcomp(features)
multimodal_pc <- as.data.frame(pc$x) %>% dplyr::select(PC1:PC30)

#Testing Set
pc_test <- predict(pc, newdata = LUSC_lncRNA_process_test)

multimodal_pc_test <- as.data.frame(pc_test) %>% dplyr::select(PC1:PC30)
  
pca.recon <- function(pca, x, k){
  mu <- matrix(rep(pca$center, nrow(pca$x)), nrow = nrow(pca$x), byrow = T)
  recon <- pca$x[,1:k] %*% t(pca$rotation[,1:k]) + mu
  rmse <- sqrt(mean((recon - x)^2))
  return(list(x = recon, rmse = rmse))
}

pca.recon_test <- function(pca, x){
  rmse <- sqrt(mean((pca - x)^2))
  return(list(x = pca, rmse = rmse))
}

PCA_RMSE_LNC <- pca.recon(pc, features, 30)$rmse
PCA_RMSE_LNC_TEST <- pca.recon_test(pc_test, LUSC_lncRNA_process_test)$rmse


#Turning into dataframes
multimodal_pc <- as.data.frame(multimodal_pc)
multimodal_pc_test <- as.data.frame(multimodal_pc_test)
LUSCclin_lung$barcode <- substring(LUSCclin_lung$barcode,1,19)

#Creating patient ID columns
multimodal_pc$barcode <- lncrna_barcode_train
multimodal_pc$barcode <- substring(multimodal_pc$barcode,1,19)
multimodal_pc <- multimodal_pc %>% mutate(TestTrain="Train")
multimodal_both <- merge(LUSCclin_lung, multimodal_pc, by = "barcode")

multimodal_pc_test <- multimodal_pc_test %>% mutate(TestTrain="Test")
multimodal_pc_test$barcode <- lncrna_barcode_test
multimodal_pc_test$barcode <- substring(multimodal_pc_test$barcode,1,19)
multimodal_both2 <- merge(LUSCclin_lung, multimodal_pc_test, by = "barcode")
multimodal_both <- rbind(multimodal_both, multimodal_both2)


multimodal_prediction <- multimodal_both %>%
  dplyr::select(-c(patient, treatments, days_to_last_follow_up, days_to_death,  shortLetterCode))

#IMPUTING NECESSARY DATA
testtrain_vector <- multimodal_prediction$TestTrain
vital_vector <- multimodal_prediction$vital_status
censor_time <- multimodal_prediction$censor_time1
barcode_lnc <- multimodal_prediction$barcode

multimodal_prediction <- multimodal_prediction %>%
  dplyr::select(-c(vital_status, censor_time1, TestTrain, barcode))


multimodal_prediction <- as.data.frame(lapply(multimodal_prediction, as.numeric))
multimodal_prediction <- scale(multimodal_prediction)
multimodal_prediction <- as.data.frame(multimodal_prediction)

#Factor Variable
multimodal_prediction$vital_status <- vital_vector
multimodal_prediction$censor_time <- censor_time
multimodal_prediction$vital_status <- as.factor(multimodal_prediction$vital_status)
multimodal_prediction$TestTrain <- testtrain_vector
multimodal_prediction$barcode <- barcode_lnc

multimodal_imputed <- multimodal_prediction
multimodal_imputed_train <- multimodal_imputed %>% filter(TestTrain=="Train")
multimodal_imputed_test <- multimodal_imputed %>% filter(TestTrain=="Test")

vital_filtered_train  <- multimodal_imputed_train$vital_status
vital_filtered_test <- multimodal_imputed_test$vital_status

censor_time_filtered_train <- multimodal_imputed_train$censor_time
censor_time_filtered_test <- multimodal_imputed_test$censor_time

multimodal_imputed_train <- multimodal_imputed_train %>% dplyr::select(starts_with("p")|starts_with("P"), barcode)

multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(c(starts_with("p")|starts_with("P"), barcode))

multimodal_imputed_train <- multimodal_imputed_train %>% dplyr::select(-c(prior_treatment, prior_malignancy, pack_years_smoked))
multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(-c(prior_treatment, prior_malignancy, pack_years_smoked))

multimodal_imputed_train$censor_time <- censor_time_filtered_train
multimodal_imputed_train$vital_status <- vital_filtered_train
multimodal_imputed_train$vital_status <- as.numeric(multimodal_imputed_train$vital_status)
multimodal_imputed <- multimodal_imputed_train

multimodal_imputed_test$censor_time <- censor_time_filtered_test
multimodal_imputed_test$vital_status <- vital_filtered_test
multimodal_imputed_test$vital_status <- as.numeric(multimodal_imputed_test$vital_status)


for (i in 1:((ncol(multimodal_imputed)-3))) {
  colnames(multimodal_imputed)[i] <- paste0("LNC", i)
}

for (i in 1:((ncol(multimodal_imputed_test)-3))) {
  colnames(multimodal_imputed_test)[i] <- paste0("LNC", i)
}

lnc_test_pca <- multimodal_imputed_test
lnc_train_pca <- multimodal_imputed

multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(c(-barcode))
multimodal_imputed <- multimodal_imputed %>% dplyr::select(c(-barcode))

multimodal_imputed_test <- as.data.frame(sapply(multimodal_imputed_test, as.numeric))
multimodal_imputed <- as.data.frame(sapply(multimodal_imputed, as.numeric))



library(glmnet)


#Overall Model
cox <- coxph(Surv(censor_time, vital_status) ~ ., data=multimodal_imputed)
cox_fit <- survfit(cox)
cox_overfit_preds <- predict(cox, multimodal_imputed_test)
cox_testingdata <- Cindex(cox_overfit_preds, Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

multimodal_imputed$vital_status <- multimodal_imputed$vital_status - 1
multimodal_imputed_test$vital_status <- multimodal_imputed_test$vital_status - 1

rfsrc.pbc_plot <- rfsrc(Surv(censor_time, vital_status) ~ ., # survival object
                  data = multimodal_imputed, # data
                  nsplit = 1, # number of random splits to consider
                  tree.err = FALSE
                  
                  )
rfsrc.pbc.testplot <- predict(rfsrc.pbc_plot,multimodal_imputed_test)
rf2_cindex_testingdata <- Cindex(rfsrc.pbc.testplot$predicted, Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))




multimodal_cv <-as.matrix(multimodal_imputed[1:nrow(multimodal_imputed), 1:(ncol(multimodal_imputed)-2)])
multimodal_cv_test <-as.matrix(multimodal_imputed_test[1:nrow(multimodal_imputed_test), 1:(ncol(multimodal_imputed_test)-2)])

cv.fit = cv.glmnet(multimodal_cv,  Surv(multimodal_imputed$censor_time,multimodal_imputed$vital_status),
                   alpha =0.5, # lasso: alpha = 1; ridge: alpha=0
                   family = "cox", maxit = 10000, type.measure = "C")
#plot(cv.fit)
est.coef = coef(cv.fit, s = cv.fit$lambda.min) # returns the p length coefficient vector
active.k = which(est.coef != 0)

glm_pred <- predict(cv.fit, multimodal_cv_test)
glm_c_testingdata <- Cindex(glm_pred,Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

library(survXgboost)
library(xgboost)

multimodal_xg <- multimodal_imputed
multimodal_xg$vital_status <- as.integer(multimodal_xg$vital_status)
label <- ifelse(multimodal_xg$vital_status == 1, multimodal_xg$censor_time, -multimodal_xg$censor_time)

matrix_xg <- as.matrix(multimodal_xg[1:nrow(multimodal_xg), 1:(ncol(multimodal_xg)-2)])

prediction_df <- multimodal_xg[1:nrow(multimodal_imputed),
                               (ncol(multimodal_imputed)-1):ncol(multimodal_imputed)]

#Overall for Plot:
val_ind <- sample.int(nrow(multimodal_xg), 0.1 * nrow(multimodal_xg))
x_train <- as.matrix(multimodal_xg [-val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")])

x_label <- label[-val_ind]
x_val <- xgb.DMatrix(as.matrix(multimodal_xg[val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")]), label = label[val_ind])

#Testing Data
multimodal_xg_test <- multimodal_imputed_test
multimodal_xg_test$vital_status <- as.integer(multimodal_xg_test$vital_status)
test_label <- ifelse(multimodal_xg_test$vital_status == 1, multimodal_xg_test$censor_time, -multimodal_xg_test$censor_time)
multimodal_test_xgboost <- multimodal_xg_test %>% dplyr::select(-c(vital_status,censor_time))

surv_xgboost_model_all <- xgb.train.surv(
  params = list(
    objective = "survival:cox",
    eval_metric = "cox-nloglik",
    eta = 0.001, subsample=.5,
    #,learning_rate=0.001
               max_depth=2
  ), data = matrix_xg, label = label,
  watchlist = list(val3 = x_val),
  nrounds = 50
  , early_stopping_rounds = 15

)

# predict survival curves
xg_df = predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost))
C_xg_overfit <- Cindex(xg_df, Surv(multimodal_xg_test$censor_time, multimodal_xg_test$vital_status))

#Examining risk scores (should be somewhat spread out)
risk_scores <- predict(object = surv_xgboost_model_all, newdata = as.matrix(multimodal_test_xgboost), type = "risk")
hist(risk_scores)

#Not overfit
elastic_res <- glm_c_testingdata
cox_res <- cox_testingdata
xg_res <- C_xg_overfit
rsf_res <- rf2_cindex_testingdata

Cindex_Test <- c(xg_res, cox_res, elastic_res, rsf_res)
Models <- c("XGBoost", "Cox", "ElasticNet", "RSF")
Data_Types <- rep("LNC RNAs", 4)
results.df <- as.data.frame(cbind(Models, Cindex_Test))
Dimension_Reduction <- rep("PCA", 4)
colnames(results.df)[2] <- "Cindex_Test"
results.df <- cbind(results.df, Data_Types)
results.df <- cbind(results.df, Dimension_Reduction)
results.df$Cindex_Test <- as.numeric(results.df$Cindex_Test)
results.df$Cindex_Test <- round(results.df$Cindex_Test, 3)
results.df_lncs_pca <- results.df
}
```


### Putting Together the Training Dataframe

```{r, echo=FALSE}
#Turning into dataframes
LUSC_lncRNA_auto <- as.data.frame(LUSC_lncRNA_auto)
LUSCclin_lung$barcode <- substring(LUSCclin_lung$barcode,1,19)

#Creating patient ID columns
LUSC_lncRNA_auto$barcode <- lncrna_barcode_train
LUSC_lncRNA_auto$barcode <- substring(LUSC_lncRNA_auto$barcode,1,19)
LUSC_lncRNA_auto <- LUSC_lncRNA_auto %>% mutate(TestTrain="Train")

#Creating dataframes with miRNA, mRNA and both (can then test all of them)
multimodal_both <- merge(LUSCclin_lung, LUSC_lncRNA_auto, by = "barcode") 
```

# Testing Dataframe

```{r,echo=FALSE}
LUSC_lncRNA_auto_output <- as.data.frame(LUSC_lncRNA_auto_output)
LUSC_lncRNA_test <- LUSC_lncRNA_auto_output %>% mutate(TestTrain="Test")
LUSC_lncRNA_test$barcode <- lncrna_barcode_test
LUSC_lncRNA_test$barcode <- substring(LUSC_lncRNA_test$barcode,1,19)
multimodal_both2 <- merge(LUSCclin_lung, LUSC_lncRNA_test, by = "barcode")
multimodal_both <- rbind(multimodal_both, multimodal_both2)
```


### Multimodal Cox PH Regression and Survival Models

### Setting up the Model

```{r, echo=FALSE}
#Only taking chosen clinical variables
multimodal_prediction <- multimodal_both %>%
  dplyr::select(-c(patient, treatments, days_to_last_follow_up, days_to_death,  shortLetterCode))

#Saving vectors for later use
testtrain_vector <- multimodal_prediction$TestTrain
vital_vector <- multimodal_prediction$vital_status
censor_time <- multimodal_prediction$censor_time1
barcode_lnc <- multimodal_prediction$barcode
multimodal_prediction <- multimodal_prediction %>%
  dplyr::select(-c(vital_status, censor_time1, TestTrain, barcode))

#Scaling the data
multimodal_prediction <- as.data.frame(lapply(multimodal_prediction, as.numeric))
multimodal_prediction <- scale(multimodal_prediction)
multimodal_prediction <- as.data.frame(multimodal_prediction)

#Imputation loop
for (i in 1:ncol(multimodal_prediction)) {
  multimodal_prediction[ , i][is.na(multimodal_prediction[ , i])] <- median(multimodal_prediction[ , i], na.rm=TRUE)
}

#Factor Variables
multimodal_prediction$vital_status <- vital_vector
multimodal_prediction$censor_time <- censor_time
multimodal_prediction$vital_status <- as.factor(multimodal_prediction$vital_status)
multimodal_prediction$TestTrain <- testtrain_vector
multimodal_prediction$barcode <- barcode_lnc
```

#Resplitting the data and creating the final feature space

```{r, echo=FALSE}
#Resplitting the data - same testing and training

#Splitting test and train again
multimodal_imputed <- multimodal_prediction
multimodal_imputed_train <- multimodal_imputed %>% filter(TestTrain=="Train")
multimodal_imputed_test <- multimodal_imputed %>% filter(TestTrain=="Test")
censor_time_filtered_train <- multimodal_imputed_train$censor_time
censor_time_filtered_test <- multimodal_imputed_test$censor_time

#only taking the autoencoder features that are from the LNC data (which start with V)
multimodal_imputed_train <- multimodal_imputed_train %>% dplyr::select(starts_with("v")|starts_with("V"), barcode)
multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(c(starts_with("v")|starts_with("V"), barcode))

#Removing volume variable (starts with V)
multimodal_imputed_train <- multimodal_imputed_train %>% dplyr::select(-c(volume))
multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(-c(volume))

#Saving outcome variables
multimodal_imputed_train$censor_time <- censor_time_filtered_train
multimodal_imputed_train$vital_status <- as.numeric(multimodal_imputed_train$vital_status)
multimodal_imputed <- multimodal_imputed_train

multimodal_imputed_test$censor_time <- censor_time_filtered_test
multimodal_imputed_test$vital_status <- as.numeric(multimodal_imputed_test$vital_status)
```

# Taking dataframe for multimodal section (to concatenate later on)

```{r, echo=FALSE}
for (i in 1:((ncol(multimodal_imputed)-3))) {
  colnames(multimodal_imputed)[i] <- paste0("LNC", i)
}

for (i in 1:((ncol(multimodal_imputed_test)-3))) {
  colnames(multimodal_imputed_test)[i] <- paste0("LNC", i)
}

lnc_test <- multimodal_imputed_test
lnc_train <- multimodal_imputed


lnc_LUSC_indices <- merge(multimodal_imputed, barcode_disease_mapping, by="barcode")
lnc_LUAD_indices_train <- which(lnc_LUSC_indices$name=="Lung Adenocarcinoma") 
lnc_LUSC_indices_train <- which(lnc_LUSC_indices$name=="Lung Squamous Cell Carcinoma")

lnc_LUSC_indices_test <- merge(multimodal_imputed_test, barcode_disease_mapping, by="barcode")
lnc_LUAD_indices_test <- which(lnc_LUSC_indices_test$name=="Lung Adenocarcinoma") 
lnc_LUSC_indices_test <- which(lnc_LUSC_indices_test$name=="Lung Squamous Cell Carcinoma")

multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(c(-barcode))
multimodal_imputed <- multimodal_imputed %>% dplyr::select(c(-barcode))

multimodal_imputed_test <- as.data.frame(sapply(multimodal_imputed_test, as.numeric))
multimodal_imputed <- as.data.frame(sapply(multimodal_imputed, as.numeric))
```


### Cox PH Model 

```{r,echo=FALSE}
library(glmnet) # needed for linear survival modeling


#Overall Model
cox <- coxph(Surv(censor_time, vital_status) ~ ., data=multimodal_imputed)
cox_fit <- survfit(cox)
cox_overfit_preds <- predict(cox, multimodal_imputed_test)
cox_testingdata <- Cindex(cox_overfit_preds, Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

#Cancer Specific - LUSC
multimodal_test_LUSC <- multimodal_imputed_test[c(lnc_LUSC_indices_test), ]
cox_overfit_preds_LUSC <- predict(cox, multimodal_test_LUSC)
cox_testingdata_LUSC <- Cindex(cox_overfit_preds_LUSC, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
multimodal_test_LUAD <- multimodal_imputed_test[c(lnc_LUAD_indices_test), ]
cox_overfit_preds_LUAD <- predict(cox, multimodal_test_LUAD)
cox_testingdata_LUAD <- Cindex(cox_overfit_preds_LUAD, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))
```


### Random Survival Forests Model

```{r, echo=FALSE}
library(randomForestSRC) # package for random forest
library(ggRandomForests)

#Adjusting output variable
multimodal_imputed$vital_status <- multimodal_imputed$vital_status - 1
multimodal_imputed_test$vital_status <- multimodal_imputed_test$vital_status - 1

rfsrc.pbc_plot <- rfsrc(Surv(censor_time, vital_status) ~ ., # survival object
                  data = multimodal_imputed, # data
                  nsplit = 1, # number of random splits to consider
                  tree.err = FALSE
                  )
rfsrc.pbc.testplot <- predict(rfsrc.pbc_plot,multimodal_imputed_test)
rf2_cindex_testingdata <- Cindex(rfsrc.pbc.testplot$predicted, Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

multimodal_test_LUAD$vital_status <- multimodal_test_LUAD$vital_status - 1
multimodal_test_LUSC$vital_status <- multimodal_test_LUSC$vital_status - 1

#Cancer Specific - LUSC
rf_overfit_preds_LUSC <- predict(rfsrc.pbc_plot, multimodal_test_LUSC)
rf_testingdata_LUSC <- Cindex(rf_overfit_preds_LUSC$predicted, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
rf_overfit_preds_LUAD <- predict(rfsrc.pbc_plot, multimodal_test_LUAD)
rf_testingdata_LUAD <- Cindex(rf_overfit_preds_LUAD$predicted, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))
```

### Elastic Net Model (used in the paper)

```{r, echo=FALSE}
multimodal_cv <-as.matrix(multimodal_imputed[1:nrow(multimodal_imputed), 1:(ncol(multimodal_imputed)-2)])
multimodal_cv_test <-as.matrix(multimodal_imputed_test[1:nrow(multimodal_imputed_test), 1:(ncol(multimodal_imputed_test)-2)])

cv.fit = cv.glmnet(multimodal_cv,  Surv(multimodal_imputed$censor_time,multimodal_imputed$vital_status),
                   alpha =1, # lasso: alpha = 1; ridge: alpha=0
                   family = "cox", maxit = 10000, type.measure = "C")
est.coef = coef(cv.fit, s = cv.fit$lambda.min) # returns the p length coefficient vector

#will give you the active non-zero coefficients used to make predictions
active.k = which(est.coef != 0)
glm_pred <- predict(cv.fit, multimodal_cv_test)
glm_c_testingdata <- Cindex(glm_pred,Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

#Individual Cancer Types Analysis
multimodal_test_LUSC_cv <-as.matrix(multimodal_test_LUSC[1:nrow(multimodal_test_LUSC ), 1:(ncol(multimodal_test_LUSC)-2)])
multimodal_test_LUAD_cv <-as.matrix(multimodal_test_LUAD[1:nrow(multimodal_test_LUAD ), 1:(ncol(multimodal_test_LUAD)-2)])

#Cancer Specific - LUSC
elastic_overfit_preds_LUSC <- predict(cv.fit, multimodal_test_LUSC_cv)
elastic_testingdata_LUSC <- Cindex(elastic_overfit_preds_LUSC, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
elastic_overfit_preds_LUAD <- predict(cv.fit, multimodal_test_LUAD_cv)
elastic_testingdata_LUAD <- Cindex(elastic_overfit_preds_LUAD, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))
```


### XG Boost

```{r, echo=FALSE}
library(survXgboost)
library(xgboost)

multimodal_xg <- multimodal_imputed
multimodal_xg$vital_status <- as.integer(multimodal_xg$vital_status)
label <- ifelse(multimodal_xg$vital_status == 1, multimodal_xg$censor_time, -multimodal_xg$censor_time)

matrix_xg <- as.matrix(multimodal_xg[1:nrow(multimodal_xg), 1:(ncol(multimodal_xg)-2)])

prediction_df <- multimodal_xg[1:nrow(multimodal_imputed),
                               (ncol(multimodal_imputed)-1):ncol(multimodal_imputed)]

#Overall for Plot:
val_ind <- sample.int(nrow(multimodal_xg), 0.1 * nrow(multimodal_xg))
x_train <- as.matrix(multimodal_xg [-val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")])

x_label <- label[-val_ind]
x_val <- xgb.DMatrix(as.matrix(multimodal_xg[val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")]), label = label[val_ind])

#Testing Data
multimodal_xg_test <- multimodal_imputed_test
multimodal_xg_test$vital_status <- as.integer(multimodal_xg_test$vital_status)
test_label <- ifelse(multimodal_xg_test$vital_status == 1, multimodal_xg_test$censor_time, -multimodal_xg_test$censor_time)
multimodal_test_xgboost <- multimodal_xg_test %>% dplyr::select(-c(vital_status,censor_time))

surv_xgboost_model_all <- xgb.train.surv(
  params = list(
    objective = "survival:cox",
    eval_metric = "cox-nloglik",
    eta = 0.001, subsample=.5,
    #,learning_rate=0.001
               max_depth=2
  ), data = matrix_xg, label = label,
  watchlist = list(val3 = x_val),
  nrounds = 50
  , early_stopping_rounds = 15

)

# predict survival curves
xg_df = predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost))
C_xg_overfit <- Cindex(xg_df, Surv(multimodal_xg_test$censor_time, multimodal_xg_test$vital_status))

#Examining risk scores (should be somewhat spread out)
risk_scores <- predict(object = surv_xgboost_model_all, newdata = as.matrix(multimodal_test_xgboost), type = "risk")
hist(risk_scores)

#Testing Data - Individual Cancer Types
multimodal_xg_test_LUSC <- multimodal_test_LUSC
multimodal_xg_test_LUSC$vital_status <- as.integer(multimodal_xg_test_LUSC$vital_status)
test_label <- ifelse(multimodal_xg_test_LUSC$vital_status == 1, multimodal_xg_test_LUSC$censor_time, -multimodal_xg_test_LUSC$censor_time)
multimodal_test_xgboost_LUSC <- multimodal_xg_test_LUSC %>% dplyr::select(-c(vital_status,censor_time))

#Testing Data
multimodal_xg_test_LUAD <- multimodal_test_LUAD
multimodal_xg_test_LUAD$vital_status <- as.integer(multimodal_xg_test_LUAD$vital_status)
test_label <- ifelse(multimodal_xg_test_LUAD$vital_status == 1, multimodal_xg_test_LUAD$censor_time, -multimodal_xg_test_LUAD$censor_time)
multimodal_test_xgboost_LUAD <- multimodal_xg_test_LUAD %>% dplyr::select(-c(vital_status,censor_time))

#Cancer Specific - LUSC
xg_overfit_preds_LUSC <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUSC))
xg_testingdata_LUSC <- Cindex(xg_overfit_preds_LUSC, Surv(multimodal_xg_test_LUSC$censor_time,multimodal_xg_test_LUSC$vital_status))

#LUAD
xg_overfit_preds_LUAD <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUAD))
xg_testingdata_LUAD <- Cindex(xg_overfit_preds_LUAD, Surv(multimodal_xg_test_LUAD$censor_time,multimodal_xg_test_LUAD$vital_status))
```



### Results 

```{r, echo=FALSE}
#Saving all results into a table which will be saved later
elastic_res <- glm_c_testingdata
cox_res <- cox_testingdata
xg_res <- C_xg_overfit
rsf_res <- rf2_cindex_testingdata

Cindex_Test <- c(xg_res, cox_res, elastic_res, rsf_res)
Models <- c("XGBoost", "Cox", "ElasticNet", "RSF")
Data_Types <- rep("LNC RNAs", 4)

results.df <- as.data.frame(cbind(Models, Cindex_Test))

Dimension_Reduction <- rep(autoencoder_type, 4)
results.df <- cbind(results.df, Dimension_Reduction)
colnames(results.df)[2] <- "Cindex_Test"
results.df <- cbind(results.df, Data_Types)

#individual modalities
Cindex_LUSC <- c(xg_testingdata_LUSC, cox_testingdata_LUSC, elastic_testingdata_LUSC, rf_testingdata_LUSC) 
Cindex_LUAD <- c(xg_testingdata_LUAD, cox_testingdata_LUAD, elastic_testingdata_LUAD, rf_testingdata_LUAD)
results.df <- as.data.frame(cbind(results.df, Cindex_LUSC))
results.df <- as.data.frame(cbind(results.df, Cindex_LUAD))

results.df$Cindex_Test <- as.numeric(results.df$Cindex_Test)

results.df$Cindex_Test <- round(results.df$Cindex_Test, 3)
results.df_lncs <- results.df #results.df_lncs is the full dataframe with all results
```


# VISUALIZATIONS of the LNC RNA Modality

## Stratified by Survival Status - 5 Year Survival PCA

```{r, echo=FALSE}
X <- multimodal_prediction

X <- merge(X, barcode_disease_mapping, by="barcode")

X_deceased <- X %>% filter(vital_status=="Dead" & censor_time < 1825)
X_alive_5years <- X %>% filter(censor_time > 1825)
X_alive_5years$vital_status <- "Alive"

X <- rbind(X_deceased, X_alive_5years)

disease_vital <- NULL
for (i in 1:nrow(X)) {
  if (X$vital_status[i]=="Dead" & X$name.y[i]=="Lung Squamous Cell Carcinoma") {
  disease_vital[i] <- "LUSC - Dead"
}
 if (X$vital_status[i]=="Alive" & X$name.y[i]=="Lung Squamous Cell Carcinoma") {
  disease_vital[i] <- "LUSC - Alive"
}
 if (X$vital_status[i]=="Dead" & X$name.y[i]=="Lung Adenocarcinoma") {
  disease_vital[i] <- "LUAD - Dead"
}
 if (X$vital_status[i]=="Alive" & X$name.y[i]=="Lung Adenocarcinoma") {
  disease_vital[i] <- "LUAD - Alive"
}
}


X <- as.data.frame(cbind(X, disease_vital))

vital_status_vector_LNC <- X$vital_status
disease_vitalstatus <- X$disease_vital

#median(X$censor_time) # Median time to death is 651, average is 938 days

X <- X %>%
  dplyr::select(-c(vital_status, censor_time, TestTrain, barcode, disease_vital, name.y, volume))

X <- X %>%
  dplyr::select(starts_with("V"))

LUSC_PCA_LNC <- prcomp(X)

ggplot(as.data.frame(LUSC_PCA_LNC$x), aes(x = PC1, y = PC2, col = factor(disease_vitalstatus))) + geom_point() + scale_colour_manual(name = "Survival Status", values = c("#EE3377", "dark red", "cadetblue1", "blue4"))  + labs(title="LNC RNA") +theme_bw()
```

## T-SNE 5 Years

```{r, echo=FALSE}
set.seed(1234) # reproducibility
tsne <- Rtsne(X, dims = 2, perplexity = 10, check_duplicates=FALSE)
t.df_LNC <- as.data.frame(tsne$Y)
colnames(t.df_LNC) <- c("V1", "V2")
ggplot(t.df_LNC, aes(x = V1, y = V2, color = factor(disease_vitalstatus))) + geom_point()+ labs(title="LNC RNAs - tSNE") + scale_color_discrete(name = "Survival Status")
```


### Autoencoder Performance

```{r,echo=FALSE}
if (linear_feature_selection!="yes") {
  
if (n_lncs <= 502 & n_lncs>60) {
print(paste0("First AE Training:", mse.ae1))
print(paste0("First AE Testing:", first_AE_RMSE))

}  

AE_RMSE <- NULL #vectors to store autoencoder performance in RMSE
AE_RMSE_train <- NULL

AE_RMSE[1] <- first_AE_RMSE
AE_RMSE_train[1] <- mse.ae1

}
```


# GENE EXPRESSION MODALITY - Code is very similar to the LNC RNA code above

### EXTRACTING LNC RNAs

```{r, echo=FALSE, eval=FALSE}
#Data Downloaded from the download script 
#This chunk is set to eval=False, because it took took too long to rerun the processing with such a huge number of genes, so I ran it these two chunks once and saved the result before loading the result two chunks below
#This section is included for reference - note: same procedure was used for the methylation data since it is too large to process more than once

load(file="~/desktop/Multimodal/RNASeq/LUSC_RNASeq_Harmonized.RData")
load(file="~/desktop/Multimodal/RNASeq/LUAD_RNASeq_Harmonized.RData")
load(file="~/desktop/Multimodal/all_sig_lncrna.Rdata") #Taken from li et al (2020)

#Removing all of the LNCRNAs in the list
remove_all <- intersect(all_sig_lncrnas, colnames(LUAD_RNAseq))
LUSC_RNAseq <- as.data.frame(LUSC_RNAseq)
LUAD_RNAseq <- as.data.frame(LUAD_RNAseq)
LUSC_RNAseq <- LUSC_RNAseq %>% dplyr::select(-c(all_of(remove_all)))
LUAD_RNAseq <- LUAD_RNAseq %>% dplyr::select(-c(all_of(remove_all)))
LUAD_RNAseq <- as.matrix(LUAD_RNAseq)
LUSC_RNAseq <- as.matrix(LUSC_RNAseq)

#Saving the data with LNC RNAs removed
#save(LUSC_RNAseq, file="~/desktop/Multimodal/RNASeq/LUSC_RNASeq_Harmonized_LNCSRemoved.RData")
#save(LUAD_RNAseq, file="~/desktop/Multimodal/RNASeq/LUAD_RNASeq_Harmonized_LNCSRemoved.RData")
```

## Processing mRNA data

```{r,echo=FALSE,eval=FALSE}
#Loading mrna data
barcodes_both <- LUSCclin_lung$barcode
load(file="~/desktop/Multimodal/RNASeq/LUAD_RNASeq_Harmonized_LNCSRemoved.RData")
load(file="~/desktop/Multimodal/RNASeq/LUSC_RNASeq_Harmonized_LNCSRemoved.RData")

#Clinical Data to get Barcodes
#load(file="~/desktop/Multimodal/LUSC_Clin_Harm.RData")
#load(file="~/desktop/Multimodal/LUAD_Clin_Harm.RData")

#Getting rid of tumor samples
normalortumor <- c(LUSCclin$shortLetterCode)
LUSC_RNAseq <- rbind(LUSC_RNAseq, LUAD_RNAseq)
rownames(LUSC_RNAseq) <- normalortumor
LUSC_RNAseq  <- as.data.frame(subset(LUSC_RNAseq, rownames(LUSC_RNAseq)=="TP"))

#Removing genes with more than 20% zeros and then scaling
LUSC_RNA_filt <- LUSC_RNAseq[, which(colMeans(LUSC_RNAseq==0) < 0.2)]
LUSC_RNA_zeros <- base::scale(LUSC_RNA_filt)
LUSC_RNA_zeros <- as.data.frame(cbind(LUSC_RNA_zeros, barcodes_both))
colnames(LUSC_RNA_zeros)[ncol(LUSC_RNA_zeros)] <- "barcode"

#Taking only the multimodal patients
LUSC_RNA_merge <- merge(LUSC_RNA_zeros, LUSCclin_lung, by="barcode")
LUSC_RNA_process <- distinct(LUSC_RNA_merge, barcode, .keep_all = TRUE)

#Saving the file before using it below
save(LUSC_RNA_process,file='~/desktop/Multimodal/RNASeq/RNASeq_Processed.RData')
```

### Splitting Training and Testing
```{r, echo=FALSE}
# Splitting into training and testing using file formed above
load(file='~/desktop/Multimodal/RNASeq/RNASeq_Processed.RData')

LUSC_RNA_process_test <- LUSC_RNA_process[test_indices, ]
LUSC_RNA_process <- LUSC_RNA_process[train_indices,]

mrna_barcode_train <- LUSC_RNA_process$barcode
mrna_barcode_test <- LUSC_RNA_process_test$barcode

vital_vectormrnas <- LUSC_RNA_process$vital_status
censor_time1 <- LUSC_RNA_process$censor_time1

LUSC_RNA_process <- LUSC_RNA_process %>% dplyr::select(starts_with("E"))
LUSC_RNA_process_test <- LUSC_RNA_process_test %>% dplyr::select(starts_with("E"))

LUSC_RNA_process <- as.data.frame(sapply(LUSC_RNA_process, as.numeric))
LUSC_RNA_process_test <- as.data.frame(sapply(LUSC_RNA_process_test, as.numeric))
```

# Linear Feature Selection

```{r, echo=FALSE}
y <- (as.numeric(factor(vital_vectormrnas))-1)
association_mat <- as.matrix(LUSC_RNA_process)

if (linear_feature_selection == "yes") {
  set.seed(1)
  associat <- uni.selection(t.vec = censor_time1, d.vec = y, X.mat=association_mat,
                          P.value = 0.4, randomize=FALSE, K=5)
  associat$P <- associat$P[1:gene_features]
  col_filtered <- rownames(as.data.frame(associat$P))
  LUSC_RNA_process <- LUSC_RNA_process %>% dplyr::select(all_of(col_filtered))
  LUSC_RNA_process_test <- LUSC_RNA_process_test %>% dplyr::select(all_of(col_filtered))
}

if (linear_feature_selection != "yes") {
  
#This also took very long to run so I saved the result - but code is below for reference than can be uncommented
#associat <- uni.selection(t.vec = censor_time1, d.vec = y, X.mat=association_mat,
                         #P.value = 0.4, randomize=FALSE, K=5)
#save(associat, file="~/desktop/Multimodal/featureselection_mrna.Rdata"#)
load(file="~/desktop/Multimodal/featureselection_mrna.Rdata")
associat$P <- associat$P[1:ngenes]
col_filtered <- rownames(as.data.frame(associat$P))
LUSC_RNA_process <- LUSC_RNA_process %>% dplyr::select(all_of(col_filtered))
LUSC_RNA_process_test <- LUSC_RNA_process_test %>% dplyr::select(all_of(col_filtered))
}


#Making sure no duplicate column names
LUSC_RNA_process <- LUSC_RNA_process[, !duplicated(colnames(LUSC_RNA_process))]
LUSC_RNA_process_test <- LUSC_RNA_process_test[, !duplicated(colnames(LUSC_RNA_process_test))]
LUSC_RNA_process_test <- as.matrix(LUSC_RNA_process_test)

#Extracting for altogether AE
mrna_fullAE <- LUSC_RNA_process
mrna_fullAE_test <- LUSC_RNA_process_test
mrna_fullAE$barcode <- mrna_barcode_train
mrna_fullAE_test <- as.data.frame(cbind(mrna_fullAE_test, mrna_barcode_test))
colnames(mrna_fullAE_test)[ncol(mrna_fullAE_test)] <- "barcode"
```


### Denoising Autoencoder (Zeros)

```{r, echo=FALSE}
if (autoencoder_type == "stacked_denoising_zeros") {
set.seed(123)

if (ngenes <= 520 & ngenes>60) {
# one-of corruption
corrupt_with_ones <- function(x) {
  n_to_sample <- floor(length(x) * .3)
  elements_to_corrupt <- sample(seq_along(x), n_to_sample, replace = FALSE)
  x[elements_to_corrupt] <- 0
  return(x)
}

inputs_currupted_ones <- LUSC_RNA_process %>%
  as.data.frame() %>%
  purrr::map_df(corrupt_with_ones)

features <- as.matrix(LUSC_RNA_process)
inputs_currupted_ones <- as.matrix(inputs_currupted_ones)

model1 <- keras_model_sequential()
model1 %>%
  layer_dense(units = gene_features, activation = "tanh", name = "BottleNeck", input_shape = ncol(inputs_currupted_ones)) %>%
  layer_dense(units = ncol(inputs_currupted_ones))

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = inputs_currupted_ones,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,inputs_currupted_ones)

intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_RNA_auto <- predict(intermediate_layer_model1, inputs_currupted_ones)
}
}

if (ngenes <= 60 | linear_feature_selection=="yes") {
  LUSC_RNA_auto <-  LUSC_RNA_process
  for (i in 1:ncol(LUSC_RNA_auto)) {
    colnames(LUSC_RNA_auto)[i] <- paste0("V",i)
  }
  
       LUSC_RNA_auto_output <- LUSC_RNA_process_test
      for (i in 1:ncol(LUSC_RNA_auto_output)) {
    colnames(LUSC_RNA_auto_output)[i] <- paste0("V",i)
  }
}
```


### Testing Different Autoencoder Types

```{r, echo=FALSE}
if (autoencoder_type == "sparse") {
  
features <- as.matrix(LUSC_RNA_process)

model1 <- keras_model_sequential()
model1 %>%
  layer_dense(units = gene_features, activation = "relu", name = "BottleNeck", input_shape = ncol(features), activity_regularizer=regularizer_l1(10e-5)) %>%
  layer_dense(units = ncol(features), activation='sigmoid')


model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,features)

intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_RNA_auto <- predict(intermediate_layer_model1, features)
first_AE_RMSE <- evaluate(model1,LUSC_RNA_process_test, LUSC_RNA_process_test)

}

if (autoencoder_type == "variational") {

features <- as.matrix(LUSC_RNA_process)
  
enc_input <- ncol(features)
latent_size <- gene_features
epsilon_std <- 1.0

enc_input <- layer_input(shape = c(ncol(features)))
layer_one <- layer_dense(enc_input, units=256, activation = "relu")
z_mean <- layer_dense(layer_one, latent_size, name = "BottleNeck")
z_log_var <- layer_dense(layer_one, latent_size)
 
encoder <- keras_model(enc_input, z_mean)
summary(encoder)


sampling <- function(arg){
  z_mean <- arg[, 1:(latent_size)]
  z_log_var <- arg[, (latent_size + 1):(2 * latent_size)]
  epsilon <- k_random_normal(shape = c(k_shape(z_mean)[[1]]), mean=0)
  z_mean + k_exp(z_log_var/2)*epsilon
}

z <- layer_concatenate(list(z_mean, z_log_var)) %>% 
  layer_lambda(sampling)


decoder_layer <- layer_dense(units = 256, activation = "relu")
decoder_mean <- layer_dense(units = ncol(features), activation = "sigmoid")
h_decoded <- decoder_layer(z)
x_decoded_mean <- decoder_mean(h_decoded)
 
model1 <- keras_model(enc_input, x_decoded_mean)

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% fit(
  features, features, 
  epochs = 100, 
  validation_split=0.1
)


mse.ae1 <- evaluate(model1, features,features)


intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_RNA_auto <- predict(intermediate_layer_model1, features)
first_AE_RMSE <- evaluate(model1,LUSC_RNA_process_test, LUSC_RNA_process_test)

}

if (autoencoder_type=="denoising_gaussian") {

features <- as.matrix(LUSC_RNA_process)

model1 <- keras_model_sequential()
model1 %>%
  layer_gaussian_noise(stddev = 1, input_shape = ncol(features)) %>%
  layer_dense(units = gene_features, activation = "tanh", name = "BottleNeck") %>%
  layer_dense(units = ncol(features))


model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,features)

intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_RNA_auto <- predict(intermediate_layer_model1, features)

}

if (autoencoder_type=="basic") {

features <- as.matrix(LUSC_RNA_process)

model1 <- keras_model_sequential()
model1 %>%
  layer_dense(units = gene_features, activation = "tanh", name = "BottleNeck", input_shape = ncol(features)) %>%
  layer_dense(units = ncol(features))

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,features)
intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_RNA_auto <- predict(intermediate_layer_model1, features)
}

```

### Evaluating Testing Data

```{r, echo=FALSE}
if(linear_feature_selection != "yes") {

if (class(LUSC_RNA_process_test[1,1])=="character") {
LUSC_RNA_process_test <- as.data.frame(sapply(LUSC_RNA_process_test, as.numeric))
}

if (ngenes <= 520 & ngenes>60) {
first_AE_RMSE <- evaluate(model1,LUSC_RNA_process_test, LUSC_RNA_process_test)
LUSC_RNA_auto_output <-predict(intermediate_layer_model1,LUSC_RNA_process_test)
  
}
}
```

### Comparing to PCA

```{r, echo=FALSE}
if (PCA == "yes") {

#Training Set
pc <- prcomp(features) 
multimodal_pc <- as.data.frame(pc$x) %>% dplyr::select(PC1:PC30)

#Testing Set
pc_test <- predict(pc, newdata = LUSC_RNA_process_test)
multimodal_pc_test <- as.data.frame(pc_test) %>% dplyr::select(PC1:PC30)
  
pca.recon <- function(pca, x, k){
  mu <- matrix(rep(pca$center, nrow(pca$x)), nrow = nrow(pca$x), byrow = T)
  recon <- pca$x[,1:k] %*% t(pca$rotation[,1:k]) + mu
  rmse <- sqrt(mean((recon - x)^2))
  return(list(x = recon, rmse = rmse))
}

pca.recon_test <- function(pca, x){
  rmse <- sqrt(mean((pca - x)^2))
  return(list(x = pca, rmse = rmse))
}

PCA_RMSE_GENE <- pca.recon(pc, features, 30)$rmse
PCA_RMSE_GENE_TEST <- pca.recon_test(pc_test, LUSC_RNA_process_test)$rmse

#Turning into dataframes
multimodal_pc <- as.data.frame(multimodal_pc)
multimodal_pc_test <- as.data.frame(multimodal_pc_test)
LUSCclin_lung$barcode <- substring(LUSCclin_lung$barcode,1,19)

#Creating patient ID columns
multimodal_pc$barcode <- mrna_barcode_train
multimodal_pc$barcode <- substring(multimodal_pc$barcode,1,19)
multimodal_pc <- multimodal_pc %>% mutate(TestTrain="Train")
multimodal_both <- merge(LUSCclin_lung, multimodal_pc, by = "barcode")

multimodal_pc_test <- multimodal_pc_test %>% mutate(TestTrain="Test")
multimodal_pc_test$barcode <- lncrna_barcode_test
multimodal_pc_test$barcode <- substring(multimodal_pc_test$barcode,1,19)
multimodal_both2 <- merge(LUSCclin_lung, multimodal_pc_test, by = "barcode")
multimodal_both <- rbind(multimodal_both, multimodal_both2)

multimodal_prediction <- multimodal_both %>%
  dplyr::select(-c(patient, treatments, days_to_last_follow_up, days_to_death,  shortLetterCode))

#IMPUTING NECESSARY DATA
testtrain_vector <- multimodal_prediction$TestTrain
vital_vector <- multimodal_prediction$vital_status
censor_time <- multimodal_prediction$censor_time1
barcode_lnc <- multimodal_prediction$barcode

multimodal_prediction <- multimodal_prediction %>%
  dplyr::select(-c(vital_status, censor_time1, TestTrain, barcode))


multimodal_prediction <- as.data.frame(lapply(multimodal_prediction, as.numeric))
multimodal_prediction <- scale(multimodal_prediction)
multimodal_prediction <- as.data.frame(multimodal_prediction)

#Factor Variable
multimodal_prediction$vital_status <- vital_vector
multimodal_prediction$censor_time <- censor_time
multimodal_prediction$vital_status <- as.factor(multimodal_prediction$vital_status)
multimodal_prediction$TestTrain <- testtrain_vector
multimodal_prediction$barcode <- barcode_lnc

multimodal_imputed <- multimodal_prediction
multimodal_imputed_train <- multimodal_imputed %>% filter(TestTrain=="Train")
multimodal_imputed_test <- multimodal_imputed %>% filter(TestTrain=="Test")

vital_filtered_train  <- multimodal_imputed_train$vital_status
vital_filtered_test <- multimodal_imputed_test$vital_status

censor_time_filtered_train <- multimodal_imputed_train$censor_time
censor_time_filtered_test <- multimodal_imputed_test$censor_time

multimodal_imputed_train <- multimodal_imputed_train %>% dplyr::select(starts_with("p")|starts_with("P"), barcode)

multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(c(starts_with("p")|starts_with("P"), barcode))

multimodal_imputed_train <- multimodal_imputed_train %>% dplyr::select(-c(prior_treatment, prior_malignancy, pack_years_smoked))
multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(-c(prior_treatment, prior_malignancy, pack_years_smoked))

multimodal_imputed_train$censor_time <- censor_time_filtered_train
multimodal_imputed_train$vital_status <- vital_filtered_train
multimodal_imputed_train$vital_status <- as.numeric(multimodal_imputed_train$vital_status)
multimodal_imputed <- multimodal_imputed_train

multimodal_imputed_test$censor_time <- censor_time_filtered_test
multimodal_imputed_test$vital_status <- vital_filtered_test
multimodal_imputed_test$vital_status <- as.numeric(multimodal_imputed_test$vital_status)


for (i in 1:((ncol(multimodal_imputed)-3))) {
  colnames(multimodal_imputed)[i] <- paste0("Gene", i)
}

for (i in 1:((ncol(multimodal_imputed_test)-3))) {
  colnames(multimodal_imputed_test)[i] <- paste0("Gene", i)
}

gene_test_pca <- multimodal_imputed_test
gene_train_pca <- multimodal_imputed

multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(c(-barcode))
multimodal_imputed <- multimodal_imputed %>% dplyr::select(c(-barcode))

multimodal_imputed_test <- as.data.frame(sapply(multimodal_imputed_test, as.numeric))
multimodal_imputed <- as.data.frame(sapply(multimodal_imputed, as.numeric))

#Overall Model
cox <- coxph(Surv(censor_time, vital_status) ~ ., data=multimodal_imputed)
cox_fit <- survfit(cox)
cox_overfit_preds <- predict(cox, multimodal_imputed_test)
cox_testingdata <- Cindex(cox_overfit_preds, Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

multimodal_imputed$vital_status <- multimodal_imputed$vital_status - 1
multimodal_imputed_test$vital_status <- multimodal_imputed_test$vital_status - 1

rfsrc.pbc_plot <- rfsrc(Surv(censor_time, vital_status) ~ ., # survival object
                  data = multimodal_imputed, # data
                  nsplit = 1, # number of random splits to consider
                  tree.err = FALSE
                  
                  )
rfsrc.pbc.testplot <- predict(rfsrc.pbc_plot,multimodal_imputed_test)
rf2_cindex_testingdata <- Cindex(rfsrc.pbc.testplot$predicted, Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

multimodal_cv <-as.matrix(multimodal_imputed[1:nrow(multimodal_imputed), 1:(ncol(multimodal_imputed)-2)])
multimodal_cv_test <-as.matrix(multimodal_imputed_test[1:nrow(multimodal_imputed_test), 1:(ncol(multimodal_imputed_test)-2)])

cv.fit = cv.glmnet(multimodal_cv,  Surv(multimodal_imputed$censor_time,multimodal_imputed$vital_status),
                   alpha =0.5, # lasso: alpha = 1; ridge: alpha=0
                   family = "cox", maxit = 10000, type.measure = "C")
#plot(cv.fit)
est.coef = coef(cv.fit, s = cv.fit$lambda.min) # returns the p length coefficient vector
active.k = which(est.coef != 0)

glm_pred <- predict(cv.fit, multimodal_cv_test)
glm_c_testingdata <- Cindex(glm_pred,Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))


library(survXgboost)
library(xgboost)

multimodal_xg <- multimodal_imputed
multimodal_xg$vital_status <- as.integer(multimodal_xg$vital_status)
label <- ifelse(multimodal_xg$vital_status == 1, multimodal_xg$censor_time, -multimodal_xg$censor_time)

matrix_xg <- as.matrix(multimodal_xg[1:nrow(multimodal_xg), 1:(ncol(multimodal_xg)-2)])

prediction_df <- multimodal_xg[1:nrow(multimodal_imputed),
                               (ncol(multimodal_imputed)-1):ncol(multimodal_imputed)]

#Overall for Plot:
val_ind <- sample.int(nrow(multimodal_xg), 0.1 * nrow(multimodal_xg))
x_train <- as.matrix(multimodal_xg [-val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")])

x_label <- label[-val_ind]
x_val <- xgb.DMatrix(as.matrix(multimodal_xg[val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")]), label = label[val_ind])

#Testing Data
multimodal_xg_test <- multimodal_imputed_test
multimodal_xg_test$vital_status <- as.integer(multimodal_xg_test$vital_status)
test_label <- ifelse(multimodal_xg_test$vital_status == 1, multimodal_xg_test$censor_time, -multimodal_xg_test$censor_time)
multimodal_test_xgboost <- multimodal_xg_test %>% dplyr::select(-c(vital_status,censor_time))

surv_xgboost_model_all <- xgb.train.surv(
  params = list(
    objective = "survival:cox",
    eval_metric = "cox-nloglik",
    eta = 0.001, subsample=.5,
    #,learning_rate=0.001
               max_depth=2
  ), data = matrix_xg, label = label,
  watchlist = list(val3 = x_val),
  nrounds = 50
  , early_stopping_rounds = 15

)

# predict survival curves
xg_df = predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost))
C_xg_overfit <- Cindex(xg_df, Surv(multimodal_xg_test$censor_time, multimodal_xg_test$vital_status))

#Examining risk scores (should be somewhat spread out)
risk_scores <- predict(object = surv_xgboost_model_all, newdata = as.matrix(multimodal_test_xgboost), type = "risk")
hist(risk_scores)

#Not overfit
elastic_res <- glm_c_testingdata
cox_res <- cox_testingdata
xg_res <- C_xg_overfit
rsf_res <- rf2_cindex_testingdata

Cindex_Test <- c(xg_res, cox_res, elastic_res, rsf_res)
Models <- c("XGBoost", "Cox", "ElasticNet", "RSF")
Data_Types <- rep("Gene Expression", 4)
results.df <- as.data.frame(cbind(Models, Cindex_Test))
Dimension_Reduction <- rep("PCA", 4)
colnames(results.df)[2] <- "Cindex_Test"
results.df <- cbind(results.df, Data_Types)
results.df <- cbind(results.df, Dimension_Reduction)
results.df$Cindex_Test <- as.numeric(results.df$Cindex_Test)
results.df$Cindex_Test <- round(results.df$Cindex_Test, 3)
results.df_gene_pca <- results.df
}
```


### Putting Together the Dataframe

```{r, echo=FALSE}
#Turning into dataframes
LUSC_RNA_auto <- as.data.frame(LUSC_RNA_auto)

#Creating patient ID columns
LUSC_RNA_auto$barcode <- mrna_barcode_train
LUSC_RNA_auto$barcode <- substring(LUSC_RNA_auto$barcode,1,19)
LUSC_RNA_auto <- LUSC_RNA_auto %>% mutate(TestTrain="Train")
multimodal_both <- merge(LUSCclin_lung, LUSC_RNA_auto, by = "barcode") 
```

# Testing Data

```{r,echo=FALSE}
LUSC_RNA_auto_output <- as.data.frame(LUSC_RNA_auto_output)
LUSC_RNA_test <- LUSC_RNA_auto_output %>% mutate(TestTrain="Test")
LUSC_RNA_test$barcode <- mrna_barcode_test
LUSC_RNA_test$barcode <- substring(LUSC_RNA_test$barcode,1,19)
multimodal_both2 <- merge(LUSCclin_lung, LUSC_RNA_test, by = "barcode")
multimodal_both <- rbind(multimodal_both, multimodal_both2)
```


### Multimodal Cox PH Regression and Survival Models

### Setting up the Model

```{r, echo=FALSE}
multimodal_prediction <- multimodal_both %>%
  dplyr::select(-c(patient, treatments, days_to_last_follow_up, days_to_death, shortLetterCode))

#IMPUTING NECESSARY DATA
testtrain_vector <- multimodal_prediction$TestTrain
vital_vector <- multimodal_prediction$vital_status
censor_time <- multimodal_prediction$censor_time1
barcode_gene <- multimodal_prediction$barcode

multimodal_prediction <- multimodal_prediction %>%
  dplyr::select(-c(vital_status, censor_time1, TestTrain, barcode))


multimodal_prediction <- as.data.frame(sapply(multimodal_prediction, as.numeric))
multimodal_prediction <- scale(multimodal_prediction)
multimodal_prediction <- as.data.frame(multimodal_prediction)
#Imputation loop
for (i in 1:ncol(multimodal_prediction)) {
  multimodal_prediction[ , i][is.na(multimodal_prediction[ , i])] <- median(multimodal_prediction[ , i], na.rm=TRUE)
}

#Factor Variable
multimodal_prediction$vital_status <- vital_vector
multimodal_prediction$censor_time <- censor_time
multimodal_prediction$vital_status <- as.factor(multimodal_prediction$vital_status)
multimodal_prediction$TestTrain <- testtrain_vector
multimodal_prediction$barcode <- barcode_gene

paste0("Final dataset has ", nrow(multimodal_prediction), " Patients")
```

```{r, echo=FALSE}
#Resplitting the data - same testing and training

multimodal_imputed <- multimodal_prediction
multimodal_imputed_train <- multimodal_imputed %>% filter(TestTrain=="Train")
multimodal_imputed_test <- multimodal_imputed %>% filter(TestTrain=="Test")

censor_time_filtered_train <- multimodal_imputed_train$censor_time
censor_time_filtered_test <- multimodal_imputed_test$censor_time

multimodal_imputed_train <- multimodal_imputed_train %>% dplyr::select(c(starts_with("v")|starts_with("V"), barcode))

multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(c(starts_with("v")|starts_with("V"), barcode))

multimodal_imputed_train <- multimodal_imputed_train %>% dplyr::select(-c(volume))
multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(-c(volume))

multimodal_imputed_train$censor_time <- censor_time_filtered_train
multimodal_imputed_train$vital_status <- as.numeric(multimodal_imputed_train$vital_status)
multimodal_imputed <- multimodal_imputed_train

multimodal_imputed_test$censor_time <- censor_time_filtered_test
multimodal_imputed_test$vital_status <- as.numeric(multimodal_imputed_test$vital_status)
```

# Taking dataframe for multimodal section to merge later on

```{r, echo=FALSE}
for (i in 1:((ncol(multimodal_imputed)-3))) {
  colnames(multimodal_imputed)[i] <- paste0("Gene Expression", i)
}

for (i in 1:((ncol(multimodal_imputed_test)-3))) {
  colnames(multimodal_imputed_test)[i] <- paste0("Gene Expression", i)
}

gene_test <- multimodal_imputed_test
gene_train <- multimodal_imputed

gene_LUSC_indices <- merge(multimodal_imputed, barcode_disease_mapping, by="barcode")
gene_LUAD_indices_train <- which(gene_LUSC_indices$name=="Lung Adenocarcinoma") 
gene_LUSC_indices_train <- which(gene_LUSC_indices$name=="Lung Squamous Cell Carcinoma")

gene_LUSC_indices_test <- merge(multimodal_imputed_test, barcode_disease_mapping, by="barcode")
gene_LUAD_indices_test <- which(gene_LUSC_indices_test$name=="Lung Adenocarcinoma") 
gene_LUSC_indices_test <- which(gene_LUSC_indices_test$name=="Lung Squamous Cell Carcinoma")

multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(c(-barcode))
multimodal_imputed <- multimodal_imputed %>% dplyr::select(c(-barcode))

multimodal_imputed_test <- as.data.frame(sapply(multimodal_imputed_test, as.numeric))
multimodal_imputed <- as.data.frame(sapply(multimodal_imputed, as.numeric))
```

### Cox PH Model

```{r,echo=FALSE}
#Overall Model
cox <- coxph(Surv(censor_time, vital_status) ~ ., data=multimodal_imputed)
cox_fit <- survfit(cox)
cox_overfit_preds <- predict(cox, multimodal_imputed_test)
cox_testingdata <- Cindex(cox_overfit_preds, Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

#Cancer Specific - LUSC
multimodal_test_LUSC <- multimodal_imputed_test[c(gene_LUSC_indices_test), ]
cox_overfit_preds_LUSC <- predict(cox, multimodal_test_LUSC)
cox_testingdata_LUSC <- Cindex(cox_overfit_preds_LUSC, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
multimodal_test_LUAD <- multimodal_imputed_test[c(gene_LUAD_indices_test), ]
cox_overfit_preds_LUAD <- predict(cox, multimodal_test_LUAD)
cox_testingdata_LUAD <- Cindex(cox_overfit_preds_LUAD, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))
```



### Random Survival Forests

```{r, echo=FALSE}
library(randomForestSRC) # package for random forest
library(ggRandomForests)
multimodal_imputed$vital_status <- multimodal_imputed$vital_status - 1
multimodal_imputed_test$vital_status <- multimodal_imputed_test$vital_status - 1

rfsrc.pbc_plot <- rfsrc(Surv(censor_time, vital_status) ~ ., # survival object
                  data = multimodal_imputed, # data
                  nsplit = 1, # number of random splits to consider
                  tree.err = FALSE
                 
                  )
rfsrc.pbc.testplot <- predict(rfsrc.pbc_plot,multimodal_imputed_test)
rf2_cindex_testingdata <- Cindex(rfsrc.pbc.testplot$predicted, Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

multimodal_test_LUAD$vital_status <- multimodal_test_LUAD$vital_status - 1
multimodal_test_LUSC$vital_status <- multimodal_test_LUSC$vital_status - 1

#Cancer Specific - LUSC
rf_overfit_preds_LUSC <- predict(rfsrc.pbc_plot, multimodal_test_LUSC)
rf_testingdata_LUSC <- Cindex(rf_overfit_preds_LUSC$predicted, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
rf_overfit_preds_LUAD <- predict(rfsrc.pbc_plot, multimodal_test_LUAD)
rf_testingdata_LUAD <- Cindex(rf_overfit_preds_LUAD$predicted, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))
```

### Elastic Net

```{r, echo=FALSE}
multimodal_cv <-as.matrix(multimodal_imputed[1:nrow(multimodal_imputed), 1:(ncol(multimodal_imputed)-2)])
multimodal_cv_test <-as.matrix(multimodal_imputed_test[1:nrow(multimodal_imputed_test), 1:(ncol(multimodal_imputed_test)-2)])

cv.fit = cv.glmnet(multimodal_cv,  Surv(multimodal_imputed$censor_time,multimodal_imputed$vital_status),
                   alpha =0.5, # lasso: alpha = 1; ridge: alpha=0
                   family = "cox", maxit = 10000, type.measure = "C")
#plot(cv.fit)
est.coef = coef(cv.fit, s = cv.fit$lambda.min) # returns the p length coefficient vector
active.k = which(est.coef != 0)

glm_pred <- predict(cv.fit, multimodal_cv_test)
glm_c_testingdata <- Cindex(glm_pred,Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

multimodal_test_LUSC_cv <-as.matrix(multimodal_test_LUSC[1:nrow(multimodal_test_LUSC ), 1:(ncol(multimodal_test_LUSC)-2)])
multimodal_test_LUAD_cv <-as.matrix(multimodal_test_LUAD[1:nrow(multimodal_test_LUAD ), 1:(ncol(multimodal_test_LUAD)-2)])


#Cancer Specific - LUSC
elastic_overfit_preds_LUSC <- predict(cv.fit, multimodal_test_LUSC_cv)
elastic_testingdata_LUSC <- Cindex(elastic_overfit_preds_LUSC, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
elastic_overfit_preds_LUAD <- predict(cv.fit, multimodal_test_LUAD_cv)
elastic_testingdata_LUAD <- Cindex(elastic_overfit_preds_LUAD, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))
```


### XG Boost

```{r, echo=FALSE}
library(survXgboost)
library(xgboost)

multimodal_xg <- multimodal_imputed
multimodal_xg$vital_status <- as.integer(multimodal_xg$vital_status)
label <- ifelse(multimodal_xg$vital_status == 1, multimodal_xg$censor_time, -multimodal_xg$censor_time)

matrix_xg <- as.matrix(multimodal_xg[1:nrow(multimodal_xg), 1:(ncol(multimodal_xg)-2)])

prediction_df <- multimodal_xg[1:nrow(multimodal_imputed),
                               (ncol(multimodal_imputed)-1):ncol(multimodal_imputed)]

#Overall for Plot:
val_ind <- sample.int(nrow(multimodal_xg), 0.1 * nrow(multimodal_xg))
x_train <- as.matrix(multimodal_xg [-val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")])

x_label <- label[-val_ind]
x_val <- xgb.DMatrix(as.matrix(multimodal_xg[val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")]), label = label[val_ind])

#Testing Data
multimodal_xg_test <- multimodal_imputed_test
multimodal_xg_test$vital_status <- as.integer(multimodal_xg_test$vital_status)
test_label <- ifelse(multimodal_xg_test$vital_status == 1, multimodal_xg_test$censor_time, -multimodal_xg_test$censor_time)
multimodal_test_xgboost <- multimodal_xg_test %>% dplyr::select(-c(vital_status,censor_time))

surv_xgboost_model_all <- xgb.train.surv(
  params = list(
    objective = "survival:cox",
    eval_metric = "cox-nloglik",
    eta = 0.001, subsample=.5,
    #,learning_rate=0.001
               max_depth=2
  ), data = matrix_xg, label = label,
  watchlist = list(val3 = x_val),
  nrounds = 50
  , early_stopping_rounds = 15

)

# predict survival curves
xg_df = predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost))
C_xg_overfit <- Cindex(xg_df, Surv(multimodal_xg_test$censor_time, multimodal_xg_test$vital_status))

#Examining risk scores (should be somewhat spread out)
risk_scores <- predict(object = surv_xgboost_model_all, newdata = as.matrix(multimodal_test_xgboost), type = "risk")
hist(risk_scores)

#Testing Data - Individual Cancer Types
multimodal_xg_test_LUSC <- multimodal_test_LUSC
multimodal_xg_test_LUSC$vital_status <- as.integer(multimodal_xg_test_LUSC$vital_status)
test_label <- ifelse(multimodal_xg_test_LUSC$vital_status == 1, multimodal_xg_test_LUSC$censor_time, -multimodal_xg_test_LUSC$censor_time)
multimodal_test_xgboost_LUSC <- multimodal_xg_test_LUSC %>% dplyr::select(-c(vital_status,censor_time))

#Testing Data
multimodal_xg_test_LUAD <- multimodal_test_LUAD
multimodal_xg_test_LUAD$vital_status <- as.integer(multimodal_xg_test_LUAD$vital_status)
test_label <- ifelse(multimodal_xg_test_LUAD$vital_status == 1, multimodal_xg_test_LUAD$censor_time, -multimodal_xg_test_LUAD$censor_time)
multimodal_test_xgboost_LUAD <- multimodal_xg_test_LUAD %>% dplyr::select(-c(vital_status,censor_time))

#Cancer Specific - LUSC
xg_overfit_preds_LUSC <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUSC))
xg_testingdata_LUSC <- Cindex(xg_overfit_preds_LUSC, Surv(multimodal_xg_test_LUSC$censor_time,multimodal_xg_test_LUSC$vital_status))

#LUAD
xg_overfit_preds_LUAD <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUAD))
xg_testingdata_LUAD <- Cindex(xg_overfit_preds_LUAD, Surv(multimodal_xg_test_LUAD$censor_time,multimodal_xg_test_LUAD$vital_status))
```



### Results 

```{r, echo=FALSE}
#Not overfit
elastic_res <- glm_c_testingdata
cox_res <- cox_testingdata
xg_res <- C_xg_overfit
rsf_res <- rf2_cindex_testingdata

Cindex_Test <- c(xg_res, cox_res, elastic_res, rsf_res)
Models <- c("XGBoost", "Cox", "ElasticNet", "RSF")
Data_Types <- rep("Gene Expression", 4)
results.df <- as.data.frame(cbind(Models, Cindex_Test))
Dimension_Reduction <- rep(autoencoder_type, 4)
results.df <- cbind(results.df, Dimension_Reduction)
colnames(results.df)[2] <- "Cindex_Test"
results.df <- cbind(results.df, Data_Types)
results.df$Cindex_Test <- as.numeric(results.df$Cindex_Test)

Cindex_LUSC <- c(xg_testingdata_LUSC, cox_testingdata_LUSC, elastic_testingdata_LUSC, rf_testingdata_LUSC) 
Cindex_LUAD <- c(xg_testingdata_LUAD, cox_testingdata_LUAD, elastic_testingdata_LUAD, rf_testingdata_LUAD)
results.df <- as.data.frame(cbind(results.df, Cindex_LUSC))
results.df <- as.data.frame(cbind(results.df, Cindex_LUAD))


results.df$Cindex_Test <-round(results.df$Cindex_Test, 3)
results.df_gene <- results.df
```


# VISUALIZATIONS

## Stratified by Survival Status - 5 Year Survival

```{r, echo=FALSE}
X <- multimodal_prediction

X <- merge(X, barcode_disease_mapping, by="barcode")

X_deceased <- X %>% filter(vital_status=="Dead" & censor_time < 1825)
X_alive_5years <- X %>% filter(censor_time > 1825)
X_alive_5years$vital_status <- "Alive"

X <- rbind(X_deceased, X_alive_5years)

disease_vital <- NULL
for (i in 1:nrow(X)) {
  if (X$vital_status[i]=="Dead" & X$name.y[i]=="Lung Squamous Cell Carcinoma") {
  disease_vital[i] <- "LUSC - Dead"
}
 if (X$vital_status[i]=="Alive" & X$name.y[i]=="Lung Squamous Cell Carcinoma") {
  disease_vital[i] <- "LUSC - Alive"
}
 if (X$vital_status[i]=="Dead" & X$name.y[i]=="Lung Adenocarcinoma") {
  disease_vital[i] <- "LUAD - Dead"
}
 if (X$vital_status[i]=="Alive" & X$name.y[i]=="Lung Adenocarcinoma") {
  disease_vital[i] <- "LUAD - Alive"
}
}

X <- as.data.frame(cbind(X, disease_vital))

vital_status_vector_GENE <- X$vital_status
disease_vitalstatus <- X$disease_vital

#median(X$censor_time) # Median time to death is 651, average is 938 days

X <- X %>%
  dplyr::select(-c(vital_status, censor_time, TestTrain, barcode, disease_vital, name.y, volume))

X <- X %>%
  dplyr::select(starts_with("V"))

LUSC_PCA_GENE <- prcomp(X)

ggplot(as.data.frame(LUSC_PCA_GENE$x), aes(x = PC1, y = PC2, col = factor(disease_vitalstatus))) + geom_point() + scale_colour_manual(name = "Survival Status", values = c("#EE3377", "dark red", "cadetblue1", "blue4"))  + labs(title="Gene Expression") +theme_bw()
```

## T-SNE 5 Years

```{r, echo=FALSE}
set.seed(1234) # reproducibility
tsne <- Rtsne(X, dims = 2, perplexity = 10)
t.df_GENE <- as.data.frame(tsne$Y)
colnames(t.df_GENE) <- c("V1", "V2")
ggplot(t.df_GENE, aes(x = V1, y = V2, color = factor(disease_vitalstatus))) + geom_point()+ labs(title="Gene Expression - tSNE") + scale_color_discrete(name = "Survival Status")
```



### Autoencoder Performance

```{r,echo=FALSE}
if (linear_feature_selection != "yes") {
  
if (ngenes <= 520 & ngenes>60) {
print(paste0("First AE Training:", mse.ae1))
print(paste0("First AE Testing:", first_AE_RMSE))

}  
AE_RMSE[2] <- first_AE_RMSE
AE_RMSE_train[2] <- mse.ae1
}
```



# MIRNA MODALITY


### miRNA SEQ DATA

```{r, echo=FALSE}
#Downloaded using download script
load(file="~/desktop/Multimodal/mirna/LUSC_miRNA.RData")
load(file="~/desktop/Multimodal/mirna/LUAD_miRNA.RData")
LUSC_miRNA <- rbind(LUSC_miRNA, LUAD_miRNA) #combining the LUAD and LUSC data

#rownames contain  which I saved as another file
load(file="~/desktop/Multimodal/mirna/LUADtumorstatus_mirna.RData")
load(file="~/desktop/Multimodal/mirna/LUSCtumorstatus_mirna.RData")

#Clinical miRNA data - only taking tumor samples
LUSCtumorstatus_mirna <- c(LUSCtumorstatus_mirna, LUADtumorstatus_mirna)
mirna_tumorstatus_index <- which(LUSCtumorstatus_mirna == "Primary Tumor")
LUSC_miRNA <- LUSC_miRNA[c(mirna_tumorstatus_index), ]

#Getting rid of features with more than 20% zeros
LUSC_miRNA_zeros <- LUSC_miRNA[, which(colMeans(!LUSC_miRNA==0) > 0.2)]
LUSC_miRNA_zeros <- scale(LUSC_miRNA_zeros)

#filtering only patients with all types of data
LUSC_miRNA_zeros <- as.data.frame(LUSC_miRNA_zeros)
LUSC_miRNA_zeros <- rownames_to_column(LUSC_miRNA_zeros)
LUSC_miRNA_zeros$rowname <- str_split_fixed(LUSC_miRNA_zeros$rowname, "_", 6)[,6]
LUSC_miRNA_zeros$rowname <- substring(LUSC_miRNA_zeros$rowname,1,19)
colnames(LUSC_miRNA_zeros)[1] <- "barcode" 
LUSC_miRNA_zeros <- merge(LUSC_miRNA_zeros, LUSCclin_lung, by="barcode")
LUSC_miRNA_process <- distinct(LUSC_miRNA_zeros, barcode, .keep_all = TRUE)
```

### Splitting Training and Testing
```{r, echo=FALSE}
#using same indices as every modality
LUSC_miRNA_process_test <- LUSC_miRNA_process[test_indices, ]
LUSC_miRNA_process <- LUSC_miRNA_process[train_indices,]

miRNA_barcode_train <- LUSC_miRNA_process$barcode
miRNA_barcode_test <- LUSC_miRNA_process_test$barcode

vital_vectormirnas <- LUSC_miRNA_process$vital_status
censor_time1 <- LUSC_miRNA_process$censor_time1

LUSC_miRNA_process <- LUSC_miRNA_process %>% dplyr::select(starts_with("V"))
LUSC_miRNA_process_test <- LUSC_miRNA_process_test %>% dplyr::select(starts_with("V"))

LUSC_miRNA_process <- LUSC_miRNA_process %>% dplyr::select(-c(volume,vital_status))
LUSC_miRNA_process_test <- LUSC_miRNA_process_test %>% dplyr::select(-c(volume,vital_status))

LUSC_miRNA_process <- as.data.frame(sapply(LUSC_miRNA_process, as.numeric))
LUSC_miRNA_process_test <- as.data.frame(sapply(LUSC_miRNA_process_test, as.numeric))
```

# Linear Feature Selection

```{r, echo=FALSE}
y <- (as.numeric(factor(vital_vectormirnas))-1)
association_mat <- as.matrix(LUSC_miRNA_process)

if (linear_feature_selection == "yes") {
  set.seed(10)
  associat <- uni.selection(t.vec = censor_time1, d.vec = y, X.mat=association_mat,
                          P.value = 0.4, randomize=FALSE, K=5)
  associat$P <- associat$P[1:mirna_features]
  col_filtered <- rownames(as.data.frame(associat$P))
  LUSC_miRNA_process <- LUSC_miRNA_process %>% dplyr::select(all_of(col_filtered))
  LUSC_miRNA_process_test <- LUSC_miRNA_process_test %>% dplyr::select(all_of(col_filtered))
}

if (linear_feature_selection != "yes") {
set.seed(10)
associat <- uni.selection(t.vec = censor_time1, d.vec = y, X.mat=association_mat,
                          P.value = 0.7, randomize=FALSE, K=5)
associat$P <- associat$P[1:n_mirna]
col_filtered <- rownames(as.data.frame(associat$P))
LUSC_miRNA_process <- LUSC_miRNA_process %>% dplyr::select(all_of(col_filtered))
LUSC_miRNA_process_test <- LUSC_miRNA_process_test %>% dplyr::select(all_of(col_filtered))
}


#Making sure no duplicate column names
LUSC_miRNA_process <- LUSC_miRNA_process[, !duplicated(colnames(LUSC_miRNA_process))]
LUSC_miRNA_process_test <- LUSC_miRNA_process_test[, !duplicated(colnames(LUSC_miRNA_process_test))]
LUSC_miRNA_process_test <- as.matrix(LUSC_miRNA_process_test)

#Extracting for altogether AE
mirna_fullAE <- LUSC_miRNA_process
mirna_fullAE_test <- LUSC_miRNA_process_test

mirna_fullAE$barcode <- miRNA_barcode_train

mirna_fullAE_test <- as.data.frame(cbind(mirna_fullAE_test, miRNA_barcode_test))
colnames(mirna_fullAE_test)[ncol(mirna_fullAE_test)] <- "barcode"
```


### Denoising Autoencoder (Zeros) For miRNA

```{r, echo=FALSE}
if (autoencoder_type == "stacked_denoising_zeros") {
if (n_mirna >= 250) {

set.seed(123)

# one-of corruption
corrupt_with_ones <- function(x) {
  n_to_sample <- floor(length(x) * .3)
  elements_to_corrupt <- sample(seq_along(x), n_to_sample, replace = FALSE)
  x[elements_to_corrupt] <- 0
  return(x)
}

inputs_currupted_ones_mirna <- LUSC_miRNA_process %>%
  as.data.frame() %>%
  purrr::map_df(corrupt_with_ones)

LUSC_miRNA_process <- as.matrix(LUSC_miRNA_process)
inputs_currupted_ones_mirna <- as.matrix(inputs_currupted_ones_mirna)

model1 <- keras_model_sequential()
model1 %>%
  layer_dense(units = mirna_features, activation = "tanh", name = "BottleNeck", input_shape = ncol(inputs_currupted_ones_mirna)) %>%
  layer_dense(units = ncol(inputs_currupted_ones_mirna))

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = inputs_currupted_ones_mirna,
  y = LUSC_miRNA_process,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, LUSC_miRNA_process,inputs_currupted_ones_mirna)
intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_miRNA_auto <- predict(intermediate_layer_model1, inputs_currupted_ones_mirna)
}
}
  
if (n_mirna < 35 | linear_feature_selection == "yes") {
  LUSC_miRNA_auto <- LUSC_miRNA_process
  for (i in 1:ncol(LUSC_miRNA_auto)) {
    colnames(LUSC_miRNA_auto)[i] <- paste0("V",i)
  }
   LUSC_miRNA_auto_test_output <- LUSC_miRNA_process_test
      for (i in 1:ncol(LUSC_miRNA_auto_test_output)) {
    colnames(LUSC_miRNA_auto_test_output)[i] <- paste0("V",i)
  }
}
```

### Testing Different Autoencoder Types

```{r, echo=FALSE}
if (autoencoder_type == "sparse") {
  
features <- as.matrix(LUSC_miRNA_process)

model1 <- keras_model_sequential()
model1 %>%
  layer_dense(units = mirna_features, activation = "relu", name = "BottleNeck", input_shape = ncol(features), activity_regularizer=regularizer_l1(10e-5)) %>%
  layer_dense(units = ncol(features), activation='sigmoid')


model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,features)
intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_miRNA_auto <- predict(intermediate_layer_model1, features)
first_AE_RMSE <- evaluate(model1,LUSC_miRNA_process_test, LUSC_miRNA_process_test)

}


if (autoencoder_type == "variational") {

features <- as.matrix(LUSC_miRNA_process)
  
enc_input <- ncol(features)
latent_size <- mirna_features
epsilon_std <- 1.0

enc_input <- layer_input(shape = c(ncol(features)))
layer_one <- layer_dense(enc_input, units=256, activation = "relu")
z_mean <- layer_dense(layer_one, latent_size, name = "BottleNeck")
z_log_var <- layer_dense(layer_one, latent_size)
 
encoder <- keras_model(enc_input, z_mean)
summary(encoder)

sampling <- function(arg){
  z_mean <- arg[, 1:(latent_size)]
  z_log_var <- arg[, (latent_size + 1):(2 * latent_size)]
  epsilon <- k_random_normal(shape = c(k_shape(z_mean)[[1]]), mean=0)
  z_mean + k_exp(z_log_var/2)*epsilon
}

z <- layer_concatenate(list(z_mean, z_log_var)) %>% 
  layer_lambda(sampling)

decoder_layer <- layer_dense(units = 256, activation = "relu")
decoder_mean <- layer_dense(units = ncol(features), activation = "sigmoid")
h_decoded <- decoder_layer(z)
x_decoded_mean <- decoder_mean(h_decoded)
 
model1 <- keras_model(enc_input, x_decoded_mean)

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% fit(
  features, features, 
  epochs = 100, 
  validation_split=0.1
)


mse.ae1 <- evaluate(model1, features,features)
intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_miRNA_auto <- predict(intermediate_layer_model1, features)
first_AE_RMSE <- evaluate(model1,LUSC_miRNA_process_test, LUSC_miRNA_process_test)

}

if (autoencoder_type=="denoising_gaussian") {

features <- as.matrix(LUSC_miRNA_process)

model1 <- keras_model_sequential()
model1 %>%
  layer_gaussian_noise(stddev = 1, input_shape = ncol(features)) %>%
  layer_dense(units = mirna_features, activation = "tanh", name = "BottleNeck") %>%
  layer_dense(units = ncol(features))

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,features)
intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_miRNA_auto <- predict(intermediate_layer_model1, features)
}


if (autoencoder_type=="basic") {

features <- as.matrix(LUSC_miRNA_process)

model1 <- keras_model_sequential()
model1 %>%
  layer_dense(units = mirna_features, activation = "tanh", name = "BottleNeck", input_shape = ncol(features)) %>%
  layer_dense(units = ncol(features))

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,features)

intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_miRNA_auto <- predict(intermediate_layer_model1, features)

}

if (autoencoder_type=="denoising_gaussian_stacked") {

features <- as.matrix(LUSC_miRNA_process)

model1 <- keras_model_sequential()
model1 %>%
  layer_gaussian_noise(stddev = 1, input_shape = ncol(features)) %>%
  layer_dense(units = 200, activation = "tanh") %>%
  layer_dense(units = mirna_features, activation = "tanh", name = "BottleNeck") %>%
  layer_dense(units = 200, activation = "tanh") %>%
  layer_dense(units = ncol(features))


model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,features)

intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_miRNA_auto <- predict(intermediate_layer_model1, features)

}
```


### Evaluating Testing Data on Autoencoder

```{r, echo=FALSE}
if (linear_feature_selection != "yes") {
if (n_mirna >= 250) {

first_AE_RMSE <- evaluate(model1,LUSC_miRNA_process_test, LUSC_miRNA_process_test)
LUSC_miRNA_auto_test_output <-predict(intermediate_layer_model1,LUSC_miRNA_process_test)
}
}
```

### Comparing to PCA

```{r, echo=FALSE}
if (PCA == "yes") {

#Training Set
pc <- prcomp(LUSC_miRNA_process) 
multimodal_pc <- as.data.frame(pc$x) %>% dplyr::select(PC1:PC30)

#Testing Set
pc_test <- predict(pc, newdata = LUSC_miRNA_process_test)
multimodal_pc_test <- as.data.frame(pc_test) %>% dplyr::select(PC1:PC30)
  

pca.recon <- function(pca, x, k){
  mu <- matrix(rep(pca$center, nrow(pca$x)), nrow = nrow(pca$x), byrow = T)
  recon <- pca$x[,1:k] %*% t(pca$rotation[,1:k]) + mu
  rmse <- sqrt(mean((recon - x)^2))
  return(list(x = recon, rmse = rmse))
}

pca.recon_test <- function(pca, x){
  rmse <- sqrt(mean((pca - x)^2))
  return(list(x = pca, rmse = rmse))
}

PCA_RMSE_MIRNA <- pca.recon(pc, features, 30)$rmse
PCA_RMSE_MIRNA_TEST <- pca.recon_test(pc_test, LUSC_miRNA_process_test)$rmse

#Turning into dataframes
multimodal_pc <- as.data.frame(multimodal_pc)
multimodal_pc_test <- as.data.frame(multimodal_pc_test)
LUSCclin_lung$barcode <- substring(LUSCclin_lung$barcode,1,19)

#Creating patient ID columns
multimodal_pc$barcode <- lncrna_barcode_train
multimodal_pc$barcode <- substring(multimodal_pc$barcode,1,19)
multimodal_pc <- multimodal_pc %>% mutate(TestTrain="Train")
multimodal_both <- merge(LUSCclin_lung, multimodal_pc, by = "barcode")

multimodal_pc_test <- multimodal_pc_test %>% mutate(TestTrain="Test")
multimodal_pc_test$barcode <- lncrna_barcode_test
multimodal_pc_test$barcode <- substring(multimodal_pc_test$barcode,1,19)
multimodal_both2 <- merge(LUSCclin_lung, multimodal_pc_test, by = "barcode")
multimodal_both <- rbind(multimodal_both, multimodal_both2)


multimodal_prediction <- multimodal_both %>%
  dplyr::select(-c(patient, treatments, days_to_last_follow_up, days_to_death,  shortLetterCode))

#IMPUTING NECESSARY DATA
testtrain_vector <- multimodal_prediction$TestTrain
vital_vector <- multimodal_prediction$vital_status
censor_time <- multimodal_prediction$censor_time1
barcode_lnc <- multimodal_prediction$barcode

multimodal_prediction <- multimodal_prediction %>%
  dplyr::select(-c(vital_status, censor_time1, TestTrain, barcode))


multimodal_prediction <- as.data.frame(lapply(multimodal_prediction, as.numeric))
multimodal_prediction <- scale(multimodal_prediction)
multimodal_prediction <- as.data.frame(multimodal_prediction)

#Factor Variable
multimodal_prediction$vital_status <- vital_vector
multimodal_prediction$censor_time <- censor_time
multimodal_prediction$vital_status <- as.factor(multimodal_prediction$vital_status)
multimodal_prediction$TestTrain <- testtrain_vector
multimodal_prediction$barcode <- barcode_lnc

multimodal_imputed <- multimodal_prediction
multimodal_imputed_train <- multimodal_imputed %>% filter(TestTrain=="Train")
multimodal_imputed_test <- multimodal_imputed %>% filter(TestTrain=="Test")

vital_filtered_train  <- multimodal_imputed_train$vital_status
vital_filtered_test <- multimodal_imputed_test$vital_status

censor_time_filtered_train <- multimodal_imputed_train$censor_time
censor_time_filtered_test <- multimodal_imputed_test$censor_time

multimodal_imputed_train <- multimodal_imputed_train %>% dplyr::select(starts_with("p")|starts_with("P"), barcode)

multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(c(starts_with("p")|starts_with("P"), barcode))

multimodal_imputed_train <- multimodal_imputed_train %>% dplyr::select(-c(prior_treatment, prior_malignancy, pack_years_smoked))
multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(-c(prior_treatment, prior_malignancy, pack_years_smoked))

multimodal_imputed_train$censor_time <- censor_time_filtered_train
multimodal_imputed_train$vital_status <- vital_filtered_train
multimodal_imputed_train$vital_status <- as.numeric(multimodal_imputed_train$vital_status)
multimodal_imputed <- multimodal_imputed_train

multimodal_imputed_test$censor_time <- censor_time_filtered_test
multimodal_imputed_test$vital_status <- vital_filtered_test
multimodal_imputed_test$vital_status <- as.numeric(multimodal_imputed_test$vital_status)


for (i in 1:((ncol(multimodal_imputed)-3))) {
  colnames(multimodal_imputed)[i] <- paste0("miRNA", i)
}

for (i in 1:((ncol(multimodal_imputed_test)-3))) {
  colnames(multimodal_imputed_test)[i] <- paste0("miRNA", i)
}

mirna_test_pca <- multimodal_imputed_test
mirna_train_pca <- multimodal_imputed

multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(c(-barcode))
multimodal_imputed <- multimodal_imputed %>% dplyr::select(c(-barcode))

multimodal_imputed_test <- as.data.frame(sapply(multimodal_imputed_test, as.numeric))
multimodal_imputed <- as.data.frame(sapply(multimodal_imputed, as.numeric))

#Overall Model
cox <- coxph(Surv(censor_time, vital_status) ~ ., data=multimodal_imputed)
cox_fit <- survfit(cox)
cox_overfit_preds <- predict(cox, multimodal_imputed_test)
cox_testingdata <- Cindex(cox_overfit_preds, Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))


library(randomForestSRC) # package for random forest
library(ggRandomForests)
multimodal_imputed$vital_status <- multimodal_imputed$vital_status - 1
multimodal_imputed_test$vital_status <- multimodal_imputed_test$vital_status - 1

rfsrc.pbc_plot <- rfsrc(Surv(censor_time, vital_status) ~ ., # survival object
                  data = multimodal_imputed, # data
                  nsplit = 1, # number of random splits to consider
                  tree.err = FALSE
                  )
rfsrc.pbc.testplot <- predict(rfsrc.pbc_plot,multimodal_imputed_test)
rf2_cindex_testingdata <- Cindex(rfsrc.pbc.testplot$predicted, Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))




multimodal_cv <-as.matrix(multimodal_imputed[1:nrow(multimodal_imputed), 1:(ncol(multimodal_imputed)-2)])
multimodal_cv_test <-as.matrix(multimodal_imputed_test[1:nrow(multimodal_imputed_test), 1:(ncol(multimodal_imputed_test)-2)])

cv.fit = cv.glmnet(multimodal_cv,  Surv(multimodal_imputed$censor_time,multimodal_imputed$vital_status),
                   alpha =0.5, # lasso: alpha = 1; ridge: alpha=0
                   family = "cox", maxit = 10000, type.measure = "C")
#plot(cv.fit)
est.coef = coef(cv.fit, s = cv.fit$lambda.min) # returns the p length coefficient vector
active.k = which(est.coef != 0)

glm_pred <- predict(cv.fit, multimodal_cv_test)
glm_c_testingdata <- Cindex(glm_pred,Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

library(survXgboost)
library(xgboost)

multimodal_xg <- multimodal_imputed
multimodal_xg$vital_status <- as.integer(multimodal_xg$vital_status)
label <- ifelse(multimodal_xg$vital_status == 1, multimodal_xg$censor_time, -multimodal_xg$censor_time)

matrix_xg <- as.matrix(multimodal_xg[1:nrow(multimodal_xg), 1:(ncol(multimodal_xg)-2)])

prediction_df <- multimodal_xg[1:nrow(multimodal_imputed),
                               (ncol(multimodal_imputed)-1):ncol(multimodal_imputed)]

#Overall for Plot:
val_ind <- sample.int(nrow(multimodal_xg), 0.1 * nrow(multimodal_xg))
x_train <- as.matrix(multimodal_xg [-val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")])

x_label <- label[-val_ind]
x_val <- xgb.DMatrix(as.matrix(multimodal_xg[val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")]), label = label[val_ind])

#Testing Data
multimodal_xg_test <- multimodal_imputed_test
multimodal_xg_test$vital_status <- as.integer(multimodal_xg_test$vital_status)
test_label <- ifelse(multimodal_xg_test$vital_status == 1, multimodal_xg_test$censor_time, -multimodal_xg_test$censor_time)
multimodal_test_xgboost <- multimodal_xg_test %>% dplyr::select(-c(vital_status,censor_time))

surv_xgboost_model_all <- xgb.train.surv(
  params = list(
    objective = "survival:cox",
    eval_metric = "cox-nloglik",
    eta = 0.001, subsample=.5,
    #,learning_rate=0.001
               max_depth=2
  ), data = matrix_xg, label = label,
  watchlist = list(val3 = x_val),
  nrounds = 50
  , early_stopping_rounds = 15

)

# predict survival curves
xg_df = predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost))
C_xg_overfit <- Cindex(xg_df, Surv(multimodal_xg_test$censor_time, multimodal_xg_test$vital_status))

#Examining risk scores (should be somewhat spread out)
risk_scores <- predict(object = surv_xgboost_model_all, newdata = as.matrix(multimodal_test_xgboost), type = "risk")
hist(risk_scores)

#Not overfit
elastic_res <- glm_c_testingdata
cox_res <- cox_testingdata
xg_res <- C_xg_overfit
rsf_res <- rf2_cindex_testingdata

Cindex_Test <- c(xg_res, cox_res, elastic_res, rsf_res)
Models <- c("XGBoost", "Cox", "ElasticNet", "RSF")
Data_Types <- rep("miRNA", 4)
results.df <- as.data.frame(cbind(Models, Cindex_Test))
Dimension_Reduction <- rep("PCA", 4)
colnames(results.df)[2] <- "Cindex_Test"
results.df <- cbind(results.df, Data_Types)
results.df <- cbind(results.df, Dimension_Reduction)
results.df$Cindex_Test <- as.numeric(results.df$Cindex_Test)
results.df$Cindex_Test <- round(results.df$Cindex_Test, 3)
results.df_mirna_pca <- results.df
print("done")
}
```

### Putting Together the Training Dataframe

```{r, echo=FALSE}
#Turning into dataframes
LUSC_miRNA_auto <- as.data.frame(LUSC_miRNA_auto)

#Creating patient ID columns
LUSC_miRNA_auto$barcode <- miRNA_barcode_train
LUSC_miRNA_auto$barcode <- substring(LUSC_miRNA_auto$barcode,1,19)
LUSC_miRNA_auto <- LUSC_miRNA_auto %>% mutate(TestTrain="Train")
multimodal_both <- merge(LUSCclin_lung, LUSC_miRNA_auto, by = "barcode") 
```

# Testing Dataframe

```{r,echo=FALSE}
LUSC_miRNA_auto_output <- as.data.frame(LUSC_miRNA_auto_test_output)
LUSC_miRNA_test <- LUSC_miRNA_auto_output %>% mutate(TestTrain="Test")
LUSC_miRNA_test$barcode <- miRNA_barcode_test
multimodal_both2 <- merge(LUSCclin_lung, LUSC_miRNA_test, by = "barcode")
multimodal_both <- rbind(multimodal_both, multimodal_both2)
```


### Multimodal Cox PH Regression and Survival Models

### Setting up the Model

```{r, echo=FALSE}
multimodal_prediction <- multimodal_both %>%
  dplyr::select(-c(patient, treatments, days_to_last_follow_up, days_to_death, shortLetterCode))
testtrain_vector <- multimodal_prediction$TestTrain
vital_vector <- multimodal_prediction$vital_status
censor_time <- multimodal_prediction$censor_time1
barcode_of_mirna <- multimodal_prediction$barcode

multimodal_prediction <- multimodal_prediction %>%
  dplyr::select(-c(vital_status, censor_time1, TestTrain, barcode))


multimodal_prediction <- as.data.frame(sapply(multimodal_prediction, as.numeric))
multimodal_prediction <- scale(multimodal_prediction)
multimodal_prediction <- as.data.frame(multimodal_prediction)
#Imputation loop
for (i in 1:ncol(multimodal_prediction)) {
  multimodal_prediction[ , i][is.na(multimodal_prediction[ , i])] <- median(multimodal_prediction[ , i], na.rm=TRUE)
}

#Factor Variable
multimodal_prediction$vital_status <- vital_vector
multimodal_prediction$censor_time <- censor_time
multimodal_prediction$vital_status <- as.factor(multimodal_prediction$vital_status)
multimodal_prediction$TestTrain <- testtrain_vector
multimodal_prediction$barcode <- barcode_of_mirna

paste0("Final dataset has ", nrow(multimodal_prediction), " Patients")
```

```{r, echo=FALSE}
#Resplitting the data - same testing and training

multimodal_imputed <- multimodal_prediction
multimodal_imputed_train <- multimodal_imputed %>% filter(TestTrain=="Train")
multimodal_imputed_test <- multimodal_imputed %>% filter(TestTrain=="Test")

censor_time_filtered_train <- multimodal_imputed_train$censor_time
censor_time_filtered_test <- multimodal_imputed_test$censor_time

multimodal_imputed_train <- multimodal_imputed_train %>% dplyr::select(c(starts_with("v")|starts_with("V"), barcode))

multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(c(starts_with("v")|starts_with("V"), barcode))

multimodal_imputed_train <- multimodal_imputed_train %>% dplyr::select(-c(volume))
multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(-c(volume))

multimodal_imputed_train$censor_time <- censor_time_filtered_train
multimodal_imputed_train$vital_status <- as.numeric(multimodal_imputed_train$vital_status)
multimodal_imputed <- multimodal_imputed_train

multimodal_imputed_test$censor_time <- censor_time_filtered_test
multimodal_imputed_test$vital_status <- as.numeric(multimodal_imputed_test$vital_status)
```

# Taking miRNA dataframe for multimodal section

```{r, echo=FALSE}
for (i in 1:((ncol(multimodal_imputed)-3))) {
  colnames(multimodal_imputed)[i] <- paste0("miRNA", i)
}

for (i in 1:((ncol(multimodal_imputed_test)-3))) {
  colnames(multimodal_imputed_test)[i] <- paste0("miRNA", i)
}

mirna_test <- multimodal_imputed_test
mirna_train <- multimodal_imputed

mirna_LUSC_indices <- merge(multimodal_imputed, barcode_disease_mapping, by="barcode")
mirna_LUAD_indices_train <- which(mirna_LUSC_indices$name=="Lung Adenocarcinoma") 
mirna_LUSC_indices_train <- which(mirna_LUSC_indices$name=="Lung Squamous Cell Carcinoma")

mirna_LUSC_indices_test <- merge(multimodal_imputed_test, barcode_disease_mapping, by="barcode")
mirna_LUAD_indices_test <- which(mirna_LUSC_indices_test$name=="Lung Adenocarcinoma") 
mirna_LUSC_indices_test <- which(mirna_LUSC_indices_test$name=="Lung Squamous Cell Carcinoma")

multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(c(-barcode))
multimodal_imputed <- multimodal_imputed %>% dplyr::select(c(-barcode))

multimodal_imputed_test <- as.data.frame(sapply(multimodal_imputed_test, as.numeric))
multimodal_imputed <- as.data.frame(sapply(multimodal_imputed, as.numeric))
```

### Cox PH Model

```{r,echo=FALSE}
#Overall Model
cox <- coxph(Surv(censor_time, vital_status) ~ ., data=multimodal_imputed)
cox_fit <- survfit(cox)
cox_overfit_preds <- predict(cox, multimodal_imputed_test)
cox_testingdata <- Cindex(cox_overfit_preds, Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

#Cancer Specific - LUSC
multimodal_test_LUSC <- multimodal_imputed_test[c(mirna_LUSC_indices_test), ]
cox_overfit_preds_LUSC <- predict(cox, multimodal_test_LUSC)
cox_testingdata_LUSC <- Cindex(cox_overfit_preds_LUSC, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
multimodal_test_LUAD <- multimodal_imputed_test[c(mirna_LUAD_indices_test), ]
cox_overfit_preds_LUAD <- predict(cox, multimodal_test_LUAD)
cox_testingdata_LUAD <- Cindex(cox_overfit_preds_LUAD, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))
```


### Random Survival Forests

```{r, echo=FALSE}
multimodal_imputed$vital_status <- multimodal_imputed$vital_status - 1
multimodal_imputed_test$vital_status <- multimodal_imputed_test$vital_status - 1

rfsrc.pbc_plot <- rfsrc(Surv(censor_time, vital_status) ~ ., # survival object
                  data = multimodal_imputed, # data
                  nsplit = 1, # number of random splits to consider
                  tree.err = FALSE
                
                  )
rfsrc.pbc.testplot <- predict(rfsrc.pbc_plot,multimodal_imputed_test)
rf2_cindex_testingdata <- Cindex(rfsrc.pbc.testplot$predicted, Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

multimodal_test_LUAD$vital_status <- multimodal_test_LUAD$vital_status - 1
multimodal_test_LUSC$vital_status <- multimodal_test_LUSC$vital_status - 1

#Cancer Specific - LUSC
rf_overfit_preds_LUSC <- predict(rfsrc.pbc_plot, multimodal_test_LUSC)
rf_testingdata_LUSC <- Cindex(rf_overfit_preds_LUSC$predicted, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
rf_overfit_preds_LUAD <- predict(rfsrc.pbc_plot, multimodal_test_LUAD)
rf_testingdata_LUAD <- Cindex(rf_overfit_preds_LUAD$predicted, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))
```

### Elastic Net

```{r, echo=FALSE}
multimodal_cv <-as.matrix(multimodal_imputed[1:nrow(multimodal_imputed), 1:(ncol(multimodal_imputed)-2)])
multimodal_cv_test <-as.matrix(multimodal_imputed_test[1:nrow(multimodal_imputed_test), 1:(ncol(multimodal_imputed_test)-2)])

cv.fit = cv.glmnet(multimodal_cv,  Surv(multimodal_imputed$censor_time,multimodal_imputed$vital_status),
                   alpha =0.5, # lasso: alpha = 1; ridge: alpha=0
                   family = "cox", maxit = 10000, type.measure = "C")
est.coef = coef(cv.fit, s = cv.fit$lambda.min) # returns the p length coefficient vector
active.k = which(est.coef != 0)

glm_pred <- predict(cv.fit, multimodal_cv_test)
glm_c_testingdata <- Cindex(glm_pred,Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

#Individual Cancer Types
multimodal_test_LUSC_cv <-as.matrix(multimodal_test_LUSC[1:nrow(multimodal_test_LUSC ), 1:(ncol(multimodal_test_LUSC)-2)])
multimodal_test_LUAD_cv <-as.matrix(multimodal_test_LUAD[1:nrow(multimodal_test_LUAD ), 1:(ncol(multimodal_test_LUAD)-2)])


#Cancer Specific - LUSC
elastic_overfit_preds_LUSC <- predict(cv.fit, multimodal_test_LUSC_cv)
elastic_testingdata_LUSC <- Cindex(elastic_overfit_preds_LUSC, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
elastic_overfit_preds_LUAD <- predict(cv.fit, multimodal_test_LUAD_cv)
elastic_testingdata_LUAD <- Cindex(elastic_overfit_preds_LUAD, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))
```


### XG Boost

```{r, echo=FALSE}
multimodal_xg <- multimodal_imputed
multimodal_xg$vital_status <- as.integer(multimodal_xg$vital_status)
label <- ifelse(multimodal_xg$vital_status == 1, multimodal_xg$censor_time, -multimodal_xg$censor_time)

matrix_xg <- as.matrix(multimodal_xg[1:nrow(multimodal_xg), 1:(ncol(multimodal_xg)-2)])

prediction_df <- multimodal_xg[1:nrow(multimodal_imputed),
                               (ncol(multimodal_imputed)-1):ncol(multimodal_imputed)]
#Overall for Plot:
val_ind <- sample.int(nrow(multimodal_xg), 0.1 * nrow(multimodal_xg))
x_train <- as.matrix(multimodal_xg [-val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")])

x_label <- label[-val_ind]
x_val <- xgb.DMatrix(as.matrix(multimodal_xg[val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")]), label = label[val_ind])

#Testing Data
multimodal_xg_test <- multimodal_imputed_test
multimodal_xg_test$vital_status <- as.integer(multimodal_xg_test$vital_status)
test_label <- ifelse(multimodal_xg_test$vital_status == 1, multimodal_xg_test$censor_time, -multimodal_xg_test$censor_time)
multimodal_test_xgboost <- multimodal_xg_test %>% dplyr::select(-c(vital_status,censor_time))

surv_xgboost_model_all <- xgb.train.surv(
  params = list(
    objective = "survival:cox",
    eval_metric = "cox-nloglik",
    eta = 0.001, subsample=.5,
    #,learning_rate=0.001
               max_depth=2
  ), data = matrix_xg, label = label,
  watchlist = list(val3 = x_val),
  nrounds = 50
  , early_stopping_rounds = 15

)
# predict survival curves
xg_df = predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost))
C_xg_overfit <- Cindex(xg_df, Surv(multimodal_xg_test$censor_time, multimodal_xg_test$vital_status))

#Examining risk scores (should be somewhat spread out)
risk_scores <- predict(object = surv_xgboost_model_all, newdata = as.matrix(multimodal_test_xgboost), type = "risk")
hist(risk_scores)

#Testing Data - Individual Cancer Types
multimodal_xg_test_LUSC <- multimodal_test_LUSC
multimodal_xg_test_LUSC$vital_status <- as.integer(multimodal_xg_test_LUSC$vital_status)
test_label <- ifelse(multimodal_xg_test_LUSC$vital_status == 1, multimodal_xg_test_LUSC$censor_time, -multimodal_xg_test_LUSC$censor_time)
multimodal_test_xgboost_LUSC <- multimodal_xg_test_LUSC %>% dplyr::select(-c(vital_status,censor_time))

#Testing Data
multimodal_xg_test_LUAD <- multimodal_test_LUAD
multimodal_xg_test_LUAD$vital_status <- as.integer(multimodal_xg_test_LUAD$vital_status)
test_label <- ifelse(multimodal_xg_test_LUAD$vital_status == 1, multimodal_xg_test_LUAD$censor_time, -multimodal_xg_test_LUAD$censor_time)
multimodal_test_xgboost_LUAD <- multimodal_xg_test_LUAD %>% dplyr::select(-c(vital_status,censor_time))

#Cancer Specific - LUSC
xg_overfit_preds_LUSC <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUSC))
xg_testingdata_LUSC <- Cindex(xg_overfit_preds_LUSC, Surv(multimodal_xg_test_LUSC$censor_time,multimodal_xg_test_LUSC$vital_status))

#LUAD
xg_overfit_preds_LUAD <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUAD))
xg_testingdata_LUAD <- Cindex(xg_overfit_preds_LUAD, Surv(multimodal_xg_test_LUAD$censor_time,multimodal_xg_test_LUAD$vital_status))
```

### Results 

```{r, echo=FALSE}
#Not overfit
elastic_res <- glm_c_testingdata
cox_res <- cox_testingdata
xg_res <- C_xg_overfit
rsf_res <- rf2_cindex_testingdata

Cindex_Test <- c(xg_res, cox_res, elastic_res, rsf_res)
Models <- c("XGBoost", "Cox", "ElasticNet", "RSF")
Data_Types <- rep("miRNA", 4)
results.df <- as.data.frame(cbind(Models, Cindex_Test))
colnames(results.df)[2] <- "Cindex_Test"
results.df <- cbind(results.df, Data_Types)
results.df$Cindex_Test <- as.numeric(results.df$Cindex_Test)
Dimension_Reduction <- rep(autoencoder_type, 4)
results.df <- cbind(results.df, Dimension_Reduction)

Cindex_LUSC <- c(xg_testingdata_LUSC, cox_testingdata_LUSC, elastic_testingdata_LUSC, rf_testingdata_LUSC) 
Cindex_LUAD <- c(xg_testingdata_LUAD, cox_testingdata_LUAD, elastic_testingdata_LUAD, rf_testingdata_LUAD)
results.df <- as.data.frame(cbind(results.df, Cindex_LUSC))
results.df <- as.data.frame(cbind(results.df, Cindex_LUAD))

results.df$Cindex_Test <-round(results.df$Cindex_Test, 3)
results.df_mirna <- results.df
```


# VISUALIZATIONS

## Stratified by Survival Status - 5 Year Survival

```{r, echo=FALSE}
X <- multimodal_prediction
X <- merge(X, barcode_disease_mapping, by="barcode")

X_deceased <- X %>% filter(vital_status=="Dead" & censor_time < 1825)
X_alive_5years <- X %>% filter(censor_time > 1825)
X_alive_5years$vital_status <- "Alive"

X <- rbind(X_deceased, X_alive_5years)

disease_vital <- NULL
for (i in 1:nrow(X)) {
  if (X$vital_status[i]=="Dead" & X$name.y[i]=="Lung Squamous Cell Carcinoma") {
  disease_vital[i] <- "LUSC - Dead"
}
 if (X$vital_status[i]=="Alive" & X$name.y[i]=="Lung Squamous Cell Carcinoma") {
  disease_vital[i] <- "LUSC - Alive"
}
 if (X$vital_status[i]=="Dead" & X$name.y[i]=="Lung Adenocarcinoma") {
  disease_vital[i] <- "LUAD - Dead"
}
 if (X$vital_status[i]=="Alive" & X$name.y[i]=="Lung Adenocarcinoma") {
  disease_vital[i] <- "LUAD - Alive"
}
}

X <- as.data.frame(cbind(X, disease_vital))

vital_status_vector_MIRNA <- X$vital_status
disease_vitalstatus <- X$disease_vital

X <- X %>%
  dplyr::select(-c(vital_status, censor_time, TestTrain, barcode, disease_vital, name.y, volume))

X <- X %>%
  dplyr::select(starts_with("V"))

LUSC_PCA_MIRNA <- prcomp(X)

ggplot(as.data.frame(LUSC_PCA_MIRNA$x), aes(x = PC1, y = PC2, col = factor(disease_vitalstatus))) + geom_point() + scale_colour_manual(name = "Survival Status", values = c("#EE3377", "dark red", "cadetblue1", "blue4"))  + labs(title="miRNA") + theme_bw()
```


## T-SNE 5-Years

```{r, echo=FALSE}
set.seed(1234) # reproducibility
tsne <- Rtsne(X, dims = 2, perplexity = 10)
t.df_MIRNA <- as.data.frame(tsne$Y)
colnames(t.df_MIRNA) <- c("V1", "V2")
ggplot(t.df_MIRNA, aes(x = V1, y = V2, color = factor(disease_vitalstatus))) + geom_point()+ labs(title="miRNA - tSNE") + scale_color_discrete(name = "Survival Status")
```


# Autoencoder Performance

```{r,echo=FALSE}
if (linear_feature_selection != "yes") {
if (n_mirna > 250) {
print(paste0("First AE Training:", mse.ae1))
print(paste0("First AE Testing:", first_AE_RMSE))
}
AE_RMSE[3] <- first_AE_RMSE
AE_RMSE_train[3] <- mse.ae1
}
```

# DNA Methylation Modality

## Methylation Preprocessing

```{r,echo=FALSE, eval=FALSE}
#Similar to gene expression modality, these are 'eval=FALSE' because the data is way too large to download more than once (it's a 485,000 x 800) matrix
#My approach was to preprocess by continuing to save files after steps were completed in order to not have to perform that step again
#I included the code here for reference

#mapping probes to cpg islands - 129,648 probes
cpg_map <- read.csv("~/desktop/EPIC_annotations/illumina_probes.csv")
cpg_map <- cpg_map %>% dplyr::select(IlmnID, UCSC_CpG_Islands_Name, Relation_to_UCSC_CpG_Island)

probe_index2 <- which(cpg_map$Relation_to_UCSC_CpG_Island=="Island")
probe_id2 <- cpg_map$IlmnID[probe_index2]

load("~/desktop/Multimodal/methylation_results/processed_meth_fulldata.Rdata")
load("~/desktop/Multimodal/methylation_results/meth_barcodes.Rdata")
rownames(LUSC_meth) <- meth_barcodes

LUSC_meth_cpg <- LUSC_meth[ ,colnames(LUSC_meth) %in% probe_id2]

#Taking only patients in common with others - view code below 
LUSC_meth_cpg <- rownames_to_column(LUSC_meth_cpg)
LUSC_meth_cpg$rowname <- substring(LUSC_meth_cpg$rowname ,1,19)

#List of only barcodes of patients that have all modalities
load(file="~/desktop/Multimodal/multimodal_barcodes.Rdata")
LUSC_cpg <- LUSC_meth_cpg %>% filter(rowname %in% multimodal_barcodes)

cpg_barcodes <- LUSC_cpg$rowname
#save(cpg_barcodes,file= "~/desktop/Multimodal/methylation_results/CPG_islands_barcodes.Rdata")
LUSC_cpg <- LUSC_cpg[,-c(1)]

#save(LUSC_cpg, file="~/desktop/Multimodal/methylation_results/CPG_islands_only.Rdata")

#Taking top 25,000 probes based on variation 
load('~/desktop/Multimodal/methylation_results/CPG_islands_only.Rdata')

LUSC_cpg_trans <- t(LUSC_cpg)
LUSC_cpg_trans <- na.omit(LUSC_cpg_trans)
LUSC_cpg <- t(LUSC_cpg_trans)
LUSC_cpg <- as.matrix(LUSC_cpg)

ranks <- colVars(LUSC_cpg)
ranks_sort <- sort(ranks, index.return=TRUE, decreasing=TRUE)
#sapply(LUSC_cpg, function(i) sd(i)/abs(mean(i)))
top_indexes <- ranks_sort$ix[1:25000] #top 25,000 methylation probes by variance
LUSC_cpg_variance <- LUSC_cpg[,top_indexes]

save(LUSC_cpg_variance, file = '~/desktop/Multimodal/methylation_results/CPG_top25000byvariance.Rdata')
```

```{r, echo=FALSE, eval=FALSE}
# More data processing which starts by loading the file created at the last step

load('~/desktop/Multimodal/methylation_results/CPG_top25000byvariance.Rdata')

#Scaling and adding barcode data
LUSC_meth <- base::scale(LUSC_cpg_variance)
LUSC_meth <- cbind(LUSC_meth, cpg_barcodes)
colnames(LUSC_meth)[ncol(LUSC_meth)] <- "barcode"

#merging dataframes to get outcome value
LUSC_meth <- merge(LUSC_meth, LUSCclin_lung, by = "barcode")
LUSC_meth <- distinct(LUSC_meth, barcode,.keep_all= TRUE)

if (class(LUSC_meth)=="matrix") {
  LUSC_meth <- as.data.frame(LUSC_meth)
}

#Saving results (will load this file in the next chunk)
save(LUSC_meth, file="~/desktop/Multimodal/LUSC_meth_processed.Rdata")
```


### Splitting Training and Testing

```{r, echo=FALSE}
#Skipped to this step for timing reasons
load("~/desktop/Multimodal/methylation_results/CPG_islands_barcodes.Rdata")
load(file="~/desktop/Multimodal/LUSC_meth_processed.Rdata")
LUSC_meth_process <- merge(LUSC_meth, LUSCclin_lung, by="barcode")

LUSC_meth_process_test <- LUSC_meth_process[test_indices, ]
LUSC_meth_process <- LUSC_meth_process[train_indices,]

meth_barcode_train <- LUSC_meth_process$barcode
meth_barcode_test <- LUSC_meth_process_test$barcode

vital_vectormeth <- LUSC_meth_process$vital_status.y
censor_time1 <- LUSC_meth_process$censor_time1

#Taking only Methylation features (probe features start with c)
LUSC_meth_process <- LUSC_meth_process %>% dplyr::select(starts_with("c"))
LUSC_meth_process_test <- LUSC_meth_process_test %>% dplyr::select(starts_with("c"))

LUSC_meth_process <- LUSC_meth_process %>% dplyr::select(-c(cigarettes_per_day,censor_time1))

LUSC_meth_process_test <- LUSC_meth_process_test %>% dplyr::select(-c(cigarettes_per_day,censor_time1))

LUSC_meth_process <- as.data.frame(sapply(LUSC_meth_process, as.numeric))
LUSC_meth_process_test <- as.data.frame(sapply(LUSC_meth_process_test, as.numeric))
```

# Linear Feature Selection

```{r, echo=FALSE}
y <- (as.numeric(factor(vital_vectormeth))-1)
association_mat <- as.matrix(LUSC_meth_process)

if (linear_feature_selection == "yes") {
  set.seed(1)
  associat <- uni.selection(t.vec = censor_time1, d.vec = y, X.mat=association_mat,
                          P.value = 0.4, randomize=FALSE, K=5)
  associat$P <- associat$P[1:meth_features]
  col_filtered <- rownames(as.data.frame(associat$P))
  LUSC_meth_process <- LUSC_meth_process %>% dplyr::select(all_of(col_filtered))
  LUSC_meth_process_test <- LUSC_meth_process_test %>% dplyr::select(all_of(col_filtered))
}

if (linear_feature_selection != "yes") {
#associat <- uni.selection(t.vec = censor_time1, d.vec = y, X.mat=association_mat,
                         #P.value = 0.4, randomize=FALSE, K=5)
#save(associat, file="~/desktop/Multimodal/featureselection_meth.Rdata")
load(file="~/desktop/Multimodal/featureselection_meth.Rdata")
associat$P <- associat$P[1:nprobes]
col_filtered <- rownames(as.data.frame(associat$P))
LUSC_meth_process <- LUSC_meth_process %>% dplyr::select(all_of(col_filtered))
LUSC_meth_process_test <- LUSC_meth_process_test %>% dplyr::select(all_of(col_filtered))
}

#Making sure no duplicate column names
LUSC_meth_process <- LUSC_meth_process[, !duplicated(colnames(LUSC_meth_process))]
LUSC_meth_process_test <- LUSC_meth_process_test[, !duplicated(colnames(LUSC_meth_process_test))]
LUSC_meth_process_test <- as.matrix(LUSC_meth_process_test)

#Extracting for altogether AE
meth_fullAE <- LUSC_meth_process
meth_fullAE_test <- LUSC_meth_process_test

meth_fullAE$barcode <- meth_barcode_train

meth_fullAE_test <- as.data.frame(cbind(meth_fullAE_test, meth_barcode_test))
colnames(meth_fullAE_test)[ncol(meth_fullAE_test)] <- "barcode"
```

### Denoising (Zeros) Autoencoder

```{r, echo=FALSE}
if (nprobes <= 501 & nprobes>60) {
set.seed(123)
# one-of corruption
corrupt_with_ones <- function(x) {
  n_to_sample <- floor(length(x) * .3)
  elements_to_corrupt <- sample(seq_along(x), n_to_sample, replace = FALSE)
  x[elements_to_corrupt] <- 0
  return(x)
}

inputs_currupted_ones <- LUSC_meth_process %>%
  as.data.frame() %>%
  purrr::map_df(corrupt_with_ones)

features <- as.matrix(LUSC_meth_process)
inputs_currupted_ones <- as.matrix(inputs_currupted_ones)

model1 <- keras_model_sequential()
model1 %>%
  layer_dense(units = meth_features, activation = "tanh", name = "BottleNeck", input_shape = ncol(inputs_currupted_ones)) %>%
  layer_dense(units = ncol(inputs_currupted_ones))

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

history3 <- model1 %>% keras::fit(
  x = inputs_currupted_ones,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)
plot(history3)

mse.ae1 <- evaluate(model1, features,inputs_currupted_ones)

intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_meth_auto <- predict(intermediate_layer_model1, inputs_currupted_ones)

}

if (nprobes <= 60 | linear_feature_selection=="yes") {
  LUSC_meth_auto <-  LUSC_meth_process
  for (i in 1:ncol(LUSC_meth_auto)) {
    colnames(LUSC_meth_auto)[i] <- paste0("V",i)
  }
       LUSC_meth_auto_output <- LUSC_meth_process_test
      for (i in 1:ncol(LUSC_meth_auto_output)) {
    colnames(LUSC_meth_auto_output)[i] <- paste0("V",i)
  }
}
```

### Testing Different Autoencoder Types

```{r, echo=FALSE}
if (autoencoder_type == "sparse") {
  
features <- as.matrix(LUSC_meth_process)

model1 <- keras_model_sequential()
model1 %>%
  layer_dense(units = meth_features, activation = "relu", name = "BottleNeck", input_shape = ncol(features), activity_regularizer=regularizer_l1(10e-5)) %>%
  layer_dense(units = ncol(features), activation='sigmoid')

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,features)

intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_meth_auto <- predict(intermediate_layer_model1, features)
first_AE_RMSE <- evaluate(model1,LUSC_meth_process_test, LUSC_meth_process_test)
}

if (autoencoder_type == "variational") {

features <- as.matrix(LUSC_meth_process)
  
enc_input <- ncol(features)
latent_size <- meth_features
epsilon_std <- 1.0

enc_input <- layer_input(shape = c(ncol(features)))
layer_one <- layer_dense(enc_input, units=256, activation = "relu")
z_mean <- layer_dense(layer_one, latent_size, name = "BottleNeck")
z_log_var <- layer_dense(layer_one, latent_size)
 
encoder <- keras_model(enc_input, z_mean)
summary(encoder)

sampling <- function(arg){
  z_mean <- arg[, 1:(latent_size)]
  z_log_var <- arg[, (latent_size + 1):(2 * latent_size)]
  epsilon <- k_random_normal(shape = c(k_shape(z_mean)[[1]]), mean=0)
  z_mean + k_exp(z_log_var/2)*epsilon
}

z <- layer_concatenate(list(z_mean, z_log_var)) %>% 
  layer_lambda(sampling)

decoder_layer <- layer_dense(units = 256, activation = "relu")
decoder_mean <- layer_dense(units = ncol(features), activation = "sigmoid")
h_decoded <- decoder_layer(z)
x_decoded_mean <- decoder_mean(h_decoded)
 
model1 <- keras_model(enc_input, x_decoded_mean)

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% fit(
  features, features, 
  epochs = 100, 
  validation_split=0.1
)

mse.ae1 <- evaluate(model1, features,features)
intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_meth_auto <- predict(intermediate_layer_model1, features)
first_AE_RMSE <- evaluate(model1,LUSC_meth_process_test, LUSC_meth_process_test)

}

if (autoencoder_type=="denoising_gaussian") {

features <- as.matrix(LUSC_meth_process)

model1 <- keras_model_sequential()
model1 %>%
  layer_gaussian_noise(stddev = 1, input_shape = ncol(features)) %>%
  layer_dense(units = meth_features, activation = "tanh", name = "BottleNeck") %>%
  layer_dense(units = ncol(features))

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,features)
intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_meth_auto <- predict(intermediate_layer_model1, features)
}

if (autoencoder_type=="basic") {

features <- as.matrix(LUSC_meth_process)

model1 <- keras_model_sequential()
model1 %>%
  layer_dense(units = meth_features, activation = "tanh", name = "BottleNeck", input_shape = ncol(features)) %>%
  layer_dense(units = ncol(features))

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,features)
intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_meth_auto <- predict(intermediate_layer_model1, features)
}

if (autoencoder_type=="denoising_gaussian_stacked") {

features <- as.matrix(LUSC_meth_process)

model1 <- keras_model_sequential()
model1 %>%
  layer_gaussian_noise(stddev = 1, input_shape = ncol(features)) %>%
  layer_dense(units = 200, activation = "tanh") %>%
  layer_dense(units = meth_features, activation = "tanh", name = "BottleNeck") %>%
  layer_dense(units = 200, activation = "tanh") %>%
  layer_dense(units = ncol(features))


model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)
mse.ae1 <- evaluate(model1, features,features)
intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_meth_auto <- predict(intermediate_layer_model1, features)
}
```

### Evaluating Autoencoder on Testing Data

```{r, echo=FALSE}
if (linear_feature_selection!="yes") {

if (class(LUSC_meth_process_test[1,1])=="character") {
LUSC_meth_process_test <- as.data.frame(sapply(LUSC_meth_process_test, as.numeric))
}

if (nprobes <= 501 & nprobes>60) {
first_AE_RMSE <- evaluate(model1,LUSC_meth_process_test, LUSC_meth_process_test)
LUSC_meth_auto_output <-predict(intermediate_layer_model1,LUSC_meth_process_test)
  
}

if (nprobes <= 60) {
  LUSC_meth_auto_output <-  LUSC_meth_process_test
  for (i in 1:ncol(LUSC_meth_auto_output)) {
    colnames(LUSC_meth_auto_output)[i] <- paste0("V",i)
  }
}
}
```

### Comparing to PCA

```{r, echo=FALSE}
if (PCA == "yes") {

#Training Set
pc <- prcomp(features) 
multimodal_pc <- as.data.frame(pc$x) %>% dplyr::select(PC1:PC30)

#Testing Set
pc_test <- predict(pc, newdata = LUSC_meth_process_test)
multimodal_pc_test <- as.data.frame(pc_test) %>% dplyr::select(PC1:PC30)
  

pca.recon <- function(pca, x, k){
  mu <- matrix(rep(pca$center, nrow(pca$x)), nrow = nrow(pca$x), byrow = T)
  recon <- pca$x[,1:k] %*% t(pca$rotation[,1:k]) + mu
  rmse <- sqrt(mean((recon - x)^2))
  return(list(x = recon, rmse = rmse))
}

pca.recon_test <- function(pca, x){
  rmse <- sqrt(mean((pca - x)^2))
  return(list(x = pca, rmse = rmse))
}

PCA_RMSE_METH <- pca.recon(pc, features, 30)$rmse
PCA_RMSE_METH_TEST <- pca.recon_test(pc_test, LUSC_meth_process_test)$rmse

#Turning into dataframes
multimodal_pc <- as.data.frame(multimodal_pc)
multimodal_pc_test <- as.data.frame(multimodal_pc_test)
LUSCclin_lung$barcode <- substring(LUSCclin_lung$barcode,1,19)

#Creating patient ID columns
multimodal_pc$barcode <- lncrna_barcode_train
multimodal_pc$barcode <- substring(multimodal_pc$barcode,1,19)
multimodal_pc <- multimodal_pc %>% mutate(TestTrain="Train")
multimodal_both <- merge(LUSCclin_lung, multimodal_pc, by = "barcode")

multimodal_pc_test <- multimodal_pc_test %>% mutate(TestTrain="Test")
multimodal_pc_test$barcode <- lncrna_barcode_test
multimodal_pc_test$barcode <- substring(multimodal_pc_test$barcode,1,19)
multimodal_both2 <- merge(LUSCclin_lung, multimodal_pc_test, by = "barcode")
multimodal_both <- rbind(multimodal_both, multimodal_both2)


multimodal_prediction <- multimodal_both %>%
  dplyr::select(-c(patient, treatments, days_to_last_follow_up, days_to_death,  shortLetterCode))

#IMPUTING NECESSARY DATA
testtrain_vector <- multimodal_prediction$TestTrain
vital_vector <- multimodal_prediction$vital_status
censor_time <- multimodal_prediction$censor_time1
barcode_lnc <- multimodal_prediction$barcode

multimodal_prediction <- multimodal_prediction %>%
  dplyr::select(-c(vital_status, censor_time1, TestTrain, barcode))


multimodal_prediction <- as.data.frame(lapply(multimodal_prediction, as.numeric))
multimodal_prediction <- scale(multimodal_prediction)
multimodal_prediction <- as.data.frame(multimodal_prediction)

#Factor Variable
multimodal_prediction$vital_status <- vital_vector
multimodal_prediction$censor_time <- censor_time
multimodal_prediction$vital_status <- as.factor(multimodal_prediction$vital_status)
multimodal_prediction$TestTrain <- testtrain_vector
multimodal_prediction$barcode <- barcode_lnc

multimodal_imputed <- multimodal_prediction
multimodal_imputed_train <- multimodal_imputed %>% filter(TestTrain=="Train")
multimodal_imputed_test <- multimodal_imputed %>% filter(TestTrain=="Test")

vital_filtered_train  <- multimodal_imputed_train$vital_status
vital_filtered_test <- multimodal_imputed_test$vital_status

censor_time_filtered_train <- multimodal_imputed_train$censor_time
censor_time_filtered_test <- multimodal_imputed_test$censor_time

multimodal_imputed_train <- multimodal_imputed_train %>% dplyr::select(starts_with("p")|starts_with("P"), barcode)

multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(c(starts_with("p")|starts_with("P"), barcode))

multimodal_imputed_train <- multimodal_imputed_train %>% dplyr::select(-c(prior_treatment, prior_malignancy, pack_years_smoked))
multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(-c(prior_treatment, prior_malignancy, pack_years_smoked))

multimodal_imputed_train$censor_time <- censor_time_filtered_train
multimodal_imputed_train$vital_status <- vital_filtered_train
multimodal_imputed_train$vital_status <- as.numeric(multimodal_imputed_train$vital_status)
multimodal_imputed <- multimodal_imputed_train

multimodal_imputed_test$censor_time <- censor_time_filtered_test
multimodal_imputed_test$vital_status <- vital_filtered_test
multimodal_imputed_test$vital_status <- as.numeric(multimodal_imputed_test$vital_status)


for (i in 1:((ncol(multimodal_imputed)-3))) {
  colnames(multimodal_imputed)[i] <- paste0("Methylation", i)
}

for (i in 1:((ncol(multimodal_imputed_test)-3))) {
  colnames(multimodal_imputed_test)[i] <- paste0("Methylation", i)
}

meth_test_pca <- multimodal_imputed_test
meth_train_pca <- multimodal_imputed

multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(c(-barcode))
multimodal_imputed <- multimodal_imputed %>% dplyr::select(c(-barcode))

multimodal_imputed_test <- as.data.frame(sapply(multimodal_imputed_test, as.numeric))
multimodal_imputed <- as.data.frame(sapply(multimodal_imputed, as.numeric))


#Overall Model
cox <- coxph(Surv(censor_time, vital_status) ~ ., data=multimodal_imputed)
cox_fit <- survfit(cox)
cox_overfit_preds <- predict(cox, multimodal_imputed_test)
cox_testingdata <- Cindex(cox_overfit_preds, Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

multimodal_imputed$vital_status <- multimodal_imputed$vital_status - 1
multimodal_imputed_test$vital_status <- multimodal_imputed_test$vital_status - 1

rfsrc.pbc_plot <- rfsrc(Surv(censor_time, vital_status) ~ ., # survival object
                  data = multimodal_imputed, # data
                  nsplit = 1, # number of random splits to consider
                  tree.err = FALSE
                  )
rfsrc.pbc.testplot <- predict(rfsrc.pbc_plot,multimodal_imputed_test)
rf2_cindex_testingdata <- Cindex(rfsrc.pbc.testplot$predicted, Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

multimodal_cv <-as.matrix(multimodal_imputed[1:nrow(multimodal_imputed), 1:(ncol(multimodal_imputed)-2)])
multimodal_cv_test <-as.matrix(multimodal_imputed_test[1:nrow(multimodal_imputed_test), 1:(ncol(multimodal_imputed_test)-2)])

cv.fit = cv.glmnet(multimodal_cv,  Surv(multimodal_imputed$censor_time,multimodal_imputed$vital_status),
                   alpha =0.5, # lasso: alpha = 1; ridge: alpha=0
                   family = "cox", maxit = 10000, type.measure = "C")
#plot(cv.fit)
est.coef = coef(cv.fit, s = cv.fit$lambda.min) # returns the p length coefficient vector
active.k = which(est.coef != 0)

glm_pred <- predict(cv.fit, multimodal_cv_test)
glm_c_testingdata <- Cindex(glm_pred,Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

multimodal_xg <- multimodal_imputed
multimodal_xg$vital_status <- as.integer(multimodal_xg$vital_status)
label <- ifelse(multimodal_xg$vital_status == 1, multimodal_xg$censor_time, -multimodal_xg$censor_time)

matrix_xg <- as.matrix(multimodal_xg[1:nrow(multimodal_xg), 1:(ncol(multimodal_xg)-2)])

prediction_df <- multimodal_xg[1:nrow(multimodal_imputed),
                               (ncol(multimodal_imputed)-1):ncol(multimodal_imputed)]
#Overall for Plot:
val_ind <- sample.int(nrow(multimodal_xg), 0.1 * nrow(multimodal_xg))
x_train <- as.matrix(multimodal_xg [-val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")])

x_label <- label[-val_ind]
x_val <- xgb.DMatrix(as.matrix(multimodal_xg[val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")]), label = label[val_ind])

#Testing Data
multimodal_xg_test <- multimodal_imputed_test
multimodal_xg_test$vital_status <- as.integer(multimodal_xg_test$vital_status)
test_label <- ifelse(multimodal_xg_test$vital_status == 1, multimodal_xg_test$censor_time, -multimodal_xg_test$censor_time)
multimodal_test_xgboost <- multimodal_xg_test %>% dplyr::select(-c(vital_status,censor_time))

surv_xgboost_model_all <- xgb.train.surv(
  params = list(
    objective = "survival:cox",
    eval_metric = "cox-nloglik",
    eta = 0.001, subsample=.5,
    #,learning_rate=0.001
               max_depth=2
  ), data = matrix_xg, label = label,
  watchlist = list(val3 = x_val),
  nrounds = 50
  , early_stopping_rounds = 15

)

# predict survival curves
xg_df = predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost))
C_xg_overfit <- Cindex(xg_df, Surv(multimodal_xg_test$censor_time, multimodal_xg_test$vital_status))

#Examining risk scores (should be somewhat spread out)
risk_scores <- predict(object = surv_xgboost_model_all, newdata = as.matrix(multimodal_test_xgboost), type = "risk")
hist(risk_scores)

#Not overfit
elastic_res <- glm_c_testingdata
cox_res <- cox_testingdata
xg_res <- C_xg_overfit
rsf_res <- rf2_cindex_testingdata

Cindex_Test <- c(xg_res, cox_res, elastic_res, rsf_res)
Models <- c("XGBoost", "Cox", "ElasticNet", "RSF")
Data_Types <- rep("Methylation", 4)
results.df <- as.data.frame(cbind(Models, Cindex_Test))
Dimension_Reduction <- rep("PCA", 4)
colnames(results.df)[2] <- "Cindex_Test"
results.df <- cbind(results.df, Data_Types)
results.df <- cbind(results.df, Dimension_Reduction)
results.df$Cindex_Test <- as.numeric(results.df$Cindex_Test)
results.df$Cindex_Test <- round(results.df$Cindex_Test, 3)
results.df_meth_pca <- results.df
}
```


### Putting Together the Training Dataframe

```{r, echo=FALSE}
#Turning into dataframes
LUSC_meth_auto <- as.data.frame(LUSC_meth_auto)
LUSC_meth_auto$barcode <- meth_barcode_train
LUSC_meth_auto <- LUSC_meth_auto %>% mutate(TestTrain="Train")
multimodal_both <- merge(LUSCclin_lung, LUSC_meth_auto, by = "barcode") 
```

# Testing Dataframe

```{r,echo=FALSE}
LUSC_meth_auto_output <- as.data.frame(LUSC_meth_auto_output)
LUSC_meth_test <- LUSC_meth_auto_output %>% mutate(TestTrain="Test")
LUSC_meth_test$barcode <- meth_barcode_test
multimodal_both2 <- merge(LUSCclin_lung, LUSC_meth_test, by = "barcode")
multimodal_both <- rbind(multimodal_both, multimodal_both2)
```

### Multimodal Cox PH Regression and Survival Models

### Setting up the Model

```{r, echo=FALSE}
multimodal_prediction <- multimodal_both %>%
  dplyr::select(-c(patient, treatments, days_to_last_follow_up, days_to_death, shortLetterCode))

#IMPUTING NECESSARY DATA
testtrain_vector <- multimodal_prediction$TestTrain
vital_vector <- multimodal_prediction$vital_status
censor_time <- multimodal_prediction$censor_time1
barcodes_of_meth <- multimodal_prediction$barcode

multimodal_prediction <- multimodal_prediction %>%
  dplyr::select(-c(vital_status, censor_time1, TestTrain, barcode))


multimodal_prediction <- as.data.frame(sapply(multimodal_prediction, as.numeric))
multimodal_prediction <- scale(multimodal_prediction)
multimodal_prediction <- as.data.frame(multimodal_prediction)

#Imputation loop with the median value
for (i in 1:ncol(multimodal_prediction)) {
  multimodal_prediction[ , i][is.na(multimodal_prediction[ , i])] <- median(multimodal_prediction[ , i], na.rm=TRUE)
}

#Factor Variable
multimodal_prediction$vital_status <- vital_vector
multimodal_prediction$censor_time <- censor_time
multimodal_prediction$vital_status <- as.factor(multimodal_prediction$vital_status)
multimodal_prediction$TestTrain <- testtrain_vector
multimodal_prediction$barcode <- barcodes_of_meth

paste0("Final dataset has ", nrow(multimodal_prediction), " Patients")
```

```{r, echo=FALSE}
#Resplitting the data - same testing and training

multimodal_imputed <- multimodal_prediction
multimodal_imputed_train <- multimodal_imputed %>% filter(TestTrain=="Train")
multimodal_imputed_test <- multimodal_imputed %>% filter(TestTrain=="Test")

censor_time_filtered_train <- multimodal_imputed_train$censor_time
censor_time_filtered_test <- multimodal_imputed_test$censor_time

multimodal_imputed_train <- multimodal_imputed_train %>% dplyr::select(c(starts_with("v")|starts_with("V"), barcode))

multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(c(starts_with("v")|starts_with("V"), barcode))

multimodal_imputed_train <- multimodal_imputed_train %>% dplyr::select(-c(volume))
multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(-c(volume))

multimodal_imputed_train$censor_time <- censor_time_filtered_train
multimodal_imputed_train$vital_status <- as.numeric(multimodal_imputed_train$vital_status)
multimodal_imputed <- multimodal_imputed_train

multimodal_imputed_test$censor_time <- censor_time_filtered_test
multimodal_imputed_test$vital_status <- as.numeric(multimodal_imputed_test$vital_status)
```

# Taking dataframe for multimodal section

```{r, echo=FALSE}
for (i in 1:((ncol(multimodal_imputed)-3))) {
  colnames(multimodal_imputed)[i] <- paste0("Methylation", i)
}

for (i in 1:((ncol(multimodal_imputed_test)-3))) {
  colnames(multimodal_imputed_test)[i] <- paste0("Methylation", i)
}

meth_test <- multimodal_imputed_test
meth_train <- multimodal_imputed

meth_LUSC_indices <- merge(multimodal_imputed, barcode_disease_mapping, by="barcode")
meth_LUAD_indices_train <- which(meth_LUSC_indices$name=="Lung Adenocarcinoma") 
meth_LUSC_indices_train <- which(meth_LUSC_indices$name=="Lung Squamous Cell Carcinoma")

meth_LUSC_indices_test <- merge(multimodal_imputed_test, barcode_disease_mapping, by="barcode")
meth_LUAD_indices_test <- which(meth_LUSC_indices_test$name=="Lung Adenocarcinoma") 
meth_LUSC_indices_test <- which(meth_LUSC_indices_test$name=="Lung Squamous Cell Carcinoma")

multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(c(-barcode))
multimodal_imputed <- multimodal_imputed %>% dplyr::select(c(-barcode))

multimodal_imputed_test <- as.data.frame(sapply(multimodal_imputed_test, as.numeric))
multimodal_imputed <- as.data.frame(sapply(multimodal_imputed, as.numeric))
```

### Cox PH Model

```{r,echo=FALSE}
#Overall Model
cox <- coxph(Surv(censor_time, vital_status) ~ ., data=multimodal_imputed)
cox_fit <- survfit(cox)
cox_overfit_preds <- predict(cox, multimodal_imputed_test)
cox_testingdata <- Cindex(cox_overfit_preds, Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

#Cancer Specific - LUSC
multimodal_test_LUSC <- multimodal_imputed_test[c(meth_LUSC_indices_test), ]
cox_overfit_preds_LUSC <- predict(cox, multimodal_test_LUSC)
cox_testingdata_LUSC <- Cindex(cox_overfit_preds_LUSC, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
multimodal_test_LUAD <- multimodal_imputed_test[c(meth_LUAD_indices_test), ]
cox_overfit_preds_LUAD <- predict(cox, multimodal_test_LUAD)
cox_testingdata_LUAD <- Cindex(cox_overfit_preds_LUAD, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))
```

### Random Survival Forests

```{r, echo=FALSE}
multimodal_imputed$vital_status <- multimodal_imputed$vital_status - 1
multimodal_imputed_test$vital_status <- multimodal_imputed_test$vital_status - 1

rfsrc.pbc_plot <- rfsrc(Surv(censor_time, vital_status) ~ ., # survival object
                  data = multimodal_imputed, # data
                  nsplit = 1, # number of random splits to consider
                  tree.err = FALSE
            
                  )
rfsrc.pbc.testplot <- predict(rfsrc.pbc_plot,multimodal_imputed_test)
rf2_cindex_testingdata <- Cindex(rfsrc.pbc.testplot$predicted, Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

multimodal_test_LUAD$vital_status <- multimodal_test_LUAD$vital_status - 1
multimodal_test_LUSC$vital_status <- multimodal_test_LUSC$vital_status - 1

#Cancer Specific - LUSC
rf_overfit_preds_LUSC <- predict(rfsrc.pbc_plot, multimodal_test_LUSC)
rf_testingdata_LUSC <- Cindex(rf_overfit_preds_LUSC$predicted, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
rf_overfit_preds_LUAD <- predict(rfsrc.pbc_plot, multimodal_test_LUAD)
rf_testingdata_LUAD <- Cindex(rf_overfit_preds_LUAD$predicted, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))
```

### Elastic Net

```{r, echo=FALSE}
multimodal_cv <-as.matrix(multimodal_imputed[1:nrow(multimodal_imputed), 1:(ncol(multimodal_imputed)-2)])
multimodal_cv_test <-as.matrix(multimodal_imputed_test[1:nrow(multimodal_imputed_test), 1:(ncol(multimodal_imputed_test)-2)])

cv.fit = cv.glmnet(multimodal_cv,  Surv(multimodal_imputed$censor_time,multimodal_imputed$vital_status),
                   alpha =0.5, # lasso: alpha = 1; ridge: alpha=0
                   family = "cox", maxit = 10000, type.measure = "C")
#plot(cv.fit)
est.coef = coef(cv.fit, s = cv.fit$lambda.min) # returns the p length coefficient vector
active.k = which(est.coef != 0)

glm_pred <- predict(cv.fit, multimodal_cv_test)
glm_c_testingdata <- Cindex(glm_pred,Surv(multimodal_imputed_test$censor_time,multimodal_imputed_test$vital_status))

#Individual Cancer Types
multimodal_test_LUSC_cv <-as.matrix(multimodal_test_LUSC[1:nrow(multimodal_test_LUSC ), 1:(ncol(multimodal_test_LUSC)-2)])
multimodal_test_LUAD_cv <-as.matrix(multimodal_test_LUAD[1:nrow(multimodal_test_LUAD ), 1:(ncol(multimodal_test_LUAD)-2)])


#Cancer Specific - LUSC
elastic_overfit_preds_LUSC <- predict(cv.fit, multimodal_test_LUSC_cv)
elastic_testingdata_LUSC <- Cindex(elastic_overfit_preds_LUSC, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
elastic_overfit_preds_LUAD <- predict(cv.fit, multimodal_test_LUAD_cv)
elastic_testingdata_LUAD <- Cindex(elastic_overfit_preds_LUAD, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))
```


### XG Boost

```{r, echo=FALSE}
multimodal_xg <- multimodal_imputed
multimodal_xg$vital_status <- as.integer(multimodal_xg$vital_status)
label <- ifelse(multimodal_xg$vital_status == 1, multimodal_xg$censor_time, -multimodal_xg$censor_time)

matrix_xg <- as.matrix(multimodal_xg[1:nrow(multimodal_xg), 1:(ncol(multimodal_xg)-2)])

prediction_df <- multimodal_xg[1:nrow(multimodal_imputed),
                               (ncol(multimodal_imputed)-1):ncol(multimodal_imputed)]

#Overall for Plot:
val_ind <- sample.int(nrow(multimodal_xg), 0.1 * nrow(multimodal_xg))
x_train <- as.matrix(multimodal_xg [-val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")])

x_label <- label[-val_ind]
x_val <- xgb.DMatrix(as.matrix(multimodal_xg[val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")]), label = label[val_ind])

#Testing Data
multimodal_xg_test <- multimodal_imputed_test
multimodal_xg_test$vital_status <- as.integer(multimodal_xg_test$vital_status)
test_label <- ifelse(multimodal_xg_test$vital_status == 1, multimodal_xg_test$censor_time, -multimodal_xg_test$censor_time)
multimodal_test_xgboost <- multimodal_xg_test %>% dplyr::select(-c(vital_status,censor_time))

surv_xgboost_model_all <- xgb.train.surv(
  params = list(
    objective = "survival:cox",
    eval_metric = "cox-nloglik",
    eta = 0.001, subsample=.5,
    #,learning_rate=0.001
               max_depth=2
  ), data = matrix_xg, label = label,
  watchlist = list(val3 = x_val),
  nrounds = 50
  , early_stopping_rounds = 15

)

# predict survival curves
xg_df = predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost))
C_xg_overfit <- Cindex(xg_df, Surv(multimodal_xg_test$censor_time, multimodal_xg_test$vital_status))

#Examining risk scores (should be somewhat spread out)
risk_scores <- predict(object = surv_xgboost_model_all, newdata = as.matrix(multimodal_test_xgboost), type = "risk")
hist(risk_scores)

#Testing Data - Individual Cancer Types
multimodal_xg_test_LUSC <- multimodal_test_LUSC
multimodal_xg_test_LUSC$vital_status <- as.integer(multimodal_xg_test_LUSC$vital_status)
test_label <- ifelse(multimodal_xg_test_LUSC$vital_status == 1, multimodal_xg_test_LUSC$censor_time, -multimodal_xg_test_LUSC$censor_time)
multimodal_test_xgboost_LUSC <- multimodal_xg_test_LUSC %>% dplyr::select(-c(vital_status,censor_time))

#Testing Data
multimodal_xg_test_LUAD <- multimodal_test_LUAD
multimodal_xg_test_LUAD$vital_status <- as.integer(multimodal_xg_test_LUAD$vital_status)
test_label <- ifelse(multimodal_xg_test_LUAD$vital_status == 1, multimodal_xg_test_LUAD$censor_time, -multimodal_xg_test_LUAD$censor_time)
multimodal_test_xgboost_LUAD <- multimodal_xg_test_LUAD %>% dplyr::select(-c(vital_status,censor_time))


#Cancer Specific - LUSC
xg_overfit_preds_LUSC <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUSC))
xg_testingdata_LUSC <- Cindex(xg_overfit_preds_LUSC, Surv(multimodal_xg_test_LUSC$censor_time,multimodal_xg_test_LUSC$vital_status))

#LUAD
xg_overfit_preds_LUAD <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUAD))
xg_testingdata_LUAD <- Cindex(xg_overfit_preds_LUAD, Surv(multimodal_xg_test_LUAD$censor_time,multimodal_xg_test_LUAD$vital_status))
```



### Results 

```{r, echo=FALSE}
#Not overfit
elastic_res <- glm_c_testingdata
cox_res <- cox_testingdata
xg_res <- C_xg_overfit
rsf_res <- rf2_cindex_testingdata

Cindex_Test <- c(xg_res, cox_res, elastic_res, rsf_res)
Models <- c("XGBoost", "Cox", "ElasticNet", "RSF")
Data_Types <- rep("Methylation", 4)
results.df <- as.data.frame(cbind(Models, Cindex_Test))
colnames(results.df)[2] <- "Cindex_Test"
results.df <- cbind(results.df, Data_Types)
results.df$Cindex_Test <- as.numeric(results.df$Cindex_Test)
Dimension_Reduction <- rep(autoencoder_type, 4)
results.df <- cbind(results.df, Dimension_Reduction)

Cindex_LUSC <- c(xg_testingdata_LUSC, cox_testingdata_LUSC, elastic_testingdata_LUSC, rf_testingdata_LUSC) 
Cindex_LUAD <- c(xg_testingdata_LUAD, cox_testingdata_LUAD, elastic_testingdata_LUAD, rf_testingdata_LUAD)
results.df <- as.data.frame(cbind(results.df, Cindex_LUSC))
results.df <- as.data.frame(cbind(results.df, Cindex_LUAD))

results.df$Cindex_Test <-round(results.df$Cindex_Test, 3)
results.df_meth <- results.df
```


# VISUALIZATIONS

## Stratified by Survival Status - 5 Year Survival

```{r, echo=FALSE}
library(Rtsne)

X <- multimodal_prediction

X <- merge(X, barcode_disease_mapping, by="barcode")

X_deceased <- X %>% filter(vital_status=="Dead" & censor_time < 1825)
X_alive_5years <- X %>% filter(censor_time > 1825)
X_alive_5years$vital_status <- "Alive"

X <- rbind(X_deceased, X_alive_5years)

disease_vital <- NULL
for (i in 1:nrow(X)) {
  if (X$vital_status[i]=="Dead" & X$name.y[i]=="Lung Squamous Cell Carcinoma") {
  disease_vital[i] <- "LUSC - Dead"
}
 if (X$vital_status[i]=="Alive" & X$name.y[i]=="Lung Squamous Cell Carcinoma") {
  disease_vital[i] <- "LUSC - Alive"
}
 if (X$vital_status[i]=="Dead" & X$name.y[i]=="Lung Adenocarcinoma") {
  disease_vital[i] <- "LUAD - Dead"
}
 if (X$vital_status[i]=="Alive" & X$name.y[i]=="Lung Adenocarcinoma") {
  disease_vital[i] <- "LUAD - Alive"
}
}
X <- as.data.frame(cbind(X, disease_vital))
vital_status_vector_METH <- X$vital_status
disease_vitalstatus <- X$disease_vital

X <- X %>%
  dplyr::select(-c(vital_status, censor_time, TestTrain, barcode, disease_vital, name.y, volume))

X <- X %>%
  dplyr::select(starts_with("V"))

LUSC_PCA_METH <- prcomp(X)

ggplot(as.data.frame(LUSC_PCA_METH$x), aes(x = PC1, y = PC2, col = factor(disease_vitalstatus))) + geom_point() + scale_colour_manual(name = "Survival Status", values = c("#EE3377", "dark red", "cadetblue1", "blue4"))  + labs(title="Methylation") +theme_bw()
```

## T-SNE 5 Years

```{r, echo=FALSE}
set.seed(1234) # reproducibility
tsne <- Rtsne(X, dims = 2, perplexity = 10, check_duplicates=FALSE)
t.df_METH <- as.data.frame(tsne$Y)
colnames(t.df_METH) <- c("V1", "V2")
ggplot(t.df_METH, aes(x = V1, y = V2, color = factor(disease_vitalstatus))) + geom_point()+ labs(title="Methylation - tSNE") + scale_color_discrete(name = "Survival Status")
```



# Autoencoder Performance

```{r,echo=FALSE}
if (linear_feature_selection != "yes") {
if (nprobes <= 501 & nprobes>60) {
print(paste0("First AE Training:", mse.ae1))
print(paste0("First AE Testing:", first_AE_RMSE))
}  

AE_RMSE[4] <- first_AE_RMSE
AE_RMSE_train[4] <- mse.ae1
}
```


# Clinical Modality


# Splitting into training and testing

```{r, echo=FALSE}
LUSCclin_lung <- LUSCclin_lung[order(LUSCclin_lung$barcode),]
LUSC_process_test <- LUSCclin_lung[test_indices, ]
LUSC_process <- LUSCclin_lung[train_indices,]

#LUSC_process <- LUSC_process %>% dplyr::select(-c(barcode))
#LUSC_process_test <- LUSC_process_test %>% dplyr::select(-c(barcode))

LUSC_process <- LUSC_process %>% mutate(TestTrain="Train")
LUSC_process_test <- LUSC_process_test %>% mutate(TestTrain="Test")
multimodal_both <- as.data.frame(rbind(LUSC_process, LUSC_process_test))
```


### Multimodal Cox PH Regression and Survival Models


### Setting up the Model

```{r, echo=FALSE}
multimodal_prediction <- multimodal_both %>%
  dplyr::select(-c(patient, treatments, days_to_last_follow_up, days_to_death,  shortLetterCode))

testtrain_vector <- multimodal_prediction$TestTrain
vital_vector <- multimodal_prediction$vital_status
clinbarcodes <- multimodal_prediction$barcode
censor_clin <- multimodal_prediction$censor_time1

multimodal_prediction <- multimodal_prediction %>%
  dplyr::select(-c(censor_time1, vital_status,TestTrain, barcode))
multimodal_prediction <- as.data.frame(sapply(multimodal_prediction, as.numeric))
multimodal_prediction <- scale(multimodal_prediction)
multimodal_prediction <- as.data.frame(multimodal_prediction)

#Imputation loop - imputing the median
for (i in 1:ncol(multimodal_prediction)) {
  multimodal_prediction[ , i][is.na(multimodal_prediction[ , i])] <- median(multimodal_prediction[ , i], na.rm=TRUE)
}

multimodal_prediction$vital_status <- vital_vector
multimodal_prediction$TestTrain <- testtrain_vector
multimodal_prediction$barcode <- clinbarcodes
multimodal_prediction$censor_time1 <- censor_clin

paste0("Final dataset has ", nrow(multimodal_prediction), " Patients")
```

```{r, echo=FALSE}
multimodal_imputed <- multimodal_prediction
multimodal_imputed_train <- multimodal_imputed %>% filter(TestTrain=="Train")
multimodal_imputed_test <- multimodal_imputed %>% filter(TestTrain=="Test")

multimodal_imputed <- multimodal_imputed_train
multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(-c(TestTrain))
multimodal_imputed <- multimodal_imputed %>% dplyr::select(-c(TestTrain))

#Putting into numeric form
multimodal_imputed$vital_status <- as.numeric(as.factor(multimodal_imputed$vital_status))
multimodal_imputed_test$vital_status <- as.numeric(as.factor(multimodal_imputed_test$vital_status))
```

# Taking dataframe for multimodal section

```{r, echo=FALSE}
clin_test <- multimodal_imputed_test
clin_train <- multimodal_imputed

clin_LUSC_indices <- merge(multimodal_imputed, barcode_disease_mapping, by="barcode")
clin_LUAD_indices_train <- which(clin_LUSC_indices$name.y=="Lung Adenocarcinoma") 
clin_LUSC_indices_train <- which(clin_LUSC_indices$name.y=="Lung Squamous Cell Carcinoma")

clin_LUSC_indices_test <- merge(multimodal_imputed_test, barcode_disease_mapping, by="barcode")
clin_LUAD_indices_test <- which(clin_LUSC_indices_test$name.y=="Lung Adenocarcinoma") 
clin_LUSC_indices_test <- which(clin_LUSC_indices_test$name.y=="Lung Squamous Cell Carcinoma")


multimodal_imputed_test <- multimodal_imputed_test %>% dplyr::select(c(-barcode))
multimodal_imputed <- multimodal_imputed %>% dplyr::select(c(-barcode))

multimodal_imputed_test <- as.data.frame(sapply(multimodal_imputed_test, as.numeric))
multimodal_imputed <- as.data.frame(sapply(multimodal_imputed, as.numeric))
```

### Multimodal Cox PH Regression and Survival Models

### Cox PH Model

```{r,echo=FALSE}
#Overall Model
cox <- coxph(Surv(censor_time1, vital_status) ~ ., data=multimodal_imputed)
cox_fit <- survfit(cox)
cox_overfit_preds <- predict(cox, multimodal_imputed_test)
cox_testingdata <- Cindex(cox_overfit_preds, Surv(multimodal_imputed_test$censor_time1,multimodal_imputed_test$vital_status))

#Cancer Specific - LUSC
multimodal_test_LUSC <- multimodal_imputed_test[clin_LUSC_indices_test, ]
cox_overfit_preds_LUSC <- predict(cox, multimodal_test_LUSC)
cox_testingdata_LUSC <- Cindex(cox_overfit_preds_LUSC, Surv(multimodal_test_LUSC$censor_time1,multimodal_test_LUSC$vital_status))

#LUAD
multimodal_test_LUAD <- multimodal_imputed_test[c(clin_LUAD_indices_test), ]
cox_overfit_preds_LUAD <- predict(cox, multimodal_test_LUAD)
cox_testingdata_LUAD <- Cindex(cox_overfit_preds_LUAD, Surv(multimodal_test_LUAD$censor_time1,multimodal_test_LUAD$vital_status))
```

### Random Survival Forests

```{r, echo=FALSE}
multimodal_imputed$vital_status <- multimodal_imputed$vital_status - 1
multimodal_imputed_test$vital_status <- multimodal_imputed_test$vital_status - 1

rfsrc.pbc_plot <- rfsrc(Surv(censor_time1, vital_status) ~ ., # survival object
                  data = multimodal_imputed, # data
                  nsplit = 1, # number of random splits to consider
                  tree.err = FALSE
                  ,importance = TRUE
                  )
rfsrc.pbc.testplot <- predict(rfsrc.pbc_plot,multimodal_imputed_test)
rf2_cindex_testingdata <- Cindex(rfsrc.pbc.testplot$predicted, Surv(multimodal_imputed_test$censor_time1,multimodal_imputed_test$vital_status))

#Top Ten Plot
imp_plot <- gg_vimp(rfsrc.pbc_plot)
imp_plot <- imp_plot[1:10, 1:ncol(imp_plot)]
imp_plot$vars <- as.character(imp_plot$vars)
imp_plot <- imp_plot[order(imp_plot$vars),]

imp_plot %>%
  ggplot(aes(x=vimp, y=reorder(vars, vimp))) + geom_col(fill="blue") + labs(y=NULL, x="Variable Importance", title = "Top 10 Most Important Variables")

multimodal_test_LUAD$vital_status <- multimodal_test_LUAD$vital_status - 1
multimodal_test_LUSC$vital_status <- multimodal_test_LUSC$vital_status - 1

#Cancer Specific - LUSC
rf_overfit_preds_LUSC <- predict(rfsrc.pbc_plot, multimodal_test_LUSC)
rf_testingdata_LUSC <- Cindex(rf_overfit_preds_LUSC$predicted, Surv(multimodal_test_LUSC$censor_time1,multimodal_test_LUSC$vital_status))

#LUAD
rf_overfit_preds_LUAD <- predict(rfsrc.pbc_plot, multimodal_test_LUAD)
rf_testingdata_LUAD <- Cindex(rf_overfit_preds_LUAD$predicted, Surv(multimodal_test_LUAD$censor_time1,multimodal_test_LUAD$vital_status))
```

### Elastic Net

```{r, echo=FALSE}
multimodal_cv <-as.matrix(multimodal_imputed[1:nrow(multimodal_imputed), 1:(ncol(multimodal_imputed)-2)])
multimodal_cv_test <-as.matrix(multimodal_imputed_test[1:nrow(multimodal_imputed_test), 1:(ncol(multimodal_imputed_test)-2)])

cv.fit = cv.glmnet(multimodal_cv,  Surv(multimodal_imputed$censor_time1,multimodal_imputed$vital_status),
                   alpha =0.5, # lasso: alpha = 1; ridge: alpha=0
                   family = "cox", maxit = 10000, type.measure = "C")
est.coef = coef(cv.fit, s = cv.fit$lambda.min) # returns the p length coefficient vector
active.k = which(est.coef != 0)

glm_pred <- predict(cv.fit, multimodal_cv_test)
glm_c_testingdata <- Cindex(glm_pred,Surv(multimodal_imputed_test$censor_time1,multimodal_imputed_test$vital_status))

#Individual Cancer Types
multimodal_test_LUSC_cv <-as.matrix(multimodal_test_LUSC[1:nrow(multimodal_test_LUSC ), 1:(ncol(multimodal_test_LUSC)-2)])
multimodal_test_LUAD_cv <-as.matrix(multimodal_test_LUAD[1:nrow(multimodal_test_LUAD ), 1:(ncol(multimodal_test_LUAD)-2)])

#Cancer Specific - LUSC
elastic_overfit_preds_LUSC <- predict(cv.fit, multimodal_test_LUSC_cv)
elastic_testingdata_LUSC <- Cindex(elastic_overfit_preds_LUSC, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
elastic_overfit_preds_LUAD <- predict(cv.fit, multimodal_test_LUAD_cv)
elastic_testingdata_LUAD <- Cindex(elastic_overfit_preds_LUAD, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))
```

### XG Boost

```{r, echo=FALSE}
library(survXgboost)
library(xgboost)

multimodal_xg <- multimodal_imputed
multimodal_xg$vital_status <- as.integer(multimodal_xg$vital_status)
label <- ifelse(multimodal_xg$vital_status == 1, multimodal_xg$censor_time1, -multimodal_xg$censor_time1)

matrix_xg <- as.matrix(multimodal_xg[1:nrow(multimodal_xg), 1:(ncol(multimodal_xg)-2)])

prediction_df <- multimodal_xg[1:nrow(multimodal_imputed),
                               (ncol(multimodal_imputed)-1):ncol(multimodal_imputed)]
#Overall for Plot:
val_ind <- sample.int(nrow(multimodal_xg), 0.1 * nrow(multimodal_xg))
x_train <- as.matrix(multimodal_xg [-val_ind, !names(multimodal_xg) %in% c("censor_time1", "vital_status")])

x_label <- label[-val_ind]
x_val <- xgb.DMatrix(as.matrix(multimodal_xg[val_ind, !names(multimodal_xg) %in% c("censor_time1", "vital_status")]), label = label[val_ind])

#Testing Data
multimodal_xg_test <- multimodal_imputed_test
multimodal_xg_test$vital_status <- as.integer(multimodal_xg_test$vital_status)
test_label <- ifelse(multimodal_xg_test$vital_status == 1, multimodal_xg_test$censor_time1, -multimodal_xg_test$censor_time1)
multimodal_test_xgboost <- multimodal_xg_test %>% dplyr::select(-c(vital_status,censor_time1))

surv_xgboost_model_all <- xgb.train.surv(
  params = list(
    objective = "survival:cox",
    eval_metric = "cox-nloglik",
    eta = 0.001, subsample=.5,
    #,learning_rate=0.001
               max_depth=2
  ), data = matrix_xg, label = label,
  watchlist = list(val3 = x_val),
  nrounds = 50
  , early_stopping_rounds = 15

)

# predict survival curves
xg_df = predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost))
C_xg_overfit <- Cindex(xg_df, Surv(multimodal_xg_test$censor_time1, multimodal_xg_test$vital_status))

#Examining risk scores (should be somewhat spread out)
risk_scores <- predict(object = surv_xgboost_model_all, newdata = as.matrix(multimodal_test_xgboost), type = "risk")
hist(risk_scores)

#Testing Data - Individual Cancer Types
multimodal_xg_test_LUSC <- multimodal_test_LUSC
multimodal_xg_test_LUSC$vital_status <- as.integer(multimodal_xg_test_LUSC$vital_status)
test_label <- ifelse(multimodal_xg_test_LUSC$vital_status == 1, multimodal_xg_test_LUSC$censor_time, -multimodal_xg_test_LUSC$censor_time)
multimodal_test_xgboost_LUSC <- multimodal_xg_test_LUSC %>% dplyr::select(-c(vital_status,censor_time1))

#Testing Data
multimodal_xg_test_LUAD <- multimodal_test_LUAD
multimodal_xg_test_LUAD$vital_status <- as.integer(multimodal_xg_test_LUAD$vital_status)
test_label <- ifelse(multimodal_xg_test_LUAD$vital_status == 1, multimodal_xg_test_LUAD$censor_time, -multimodal_xg_test_LUAD$censor_time)
multimodal_test_xgboost_LUAD <- multimodal_xg_test_LUAD %>% dplyr::select(-c(vital_status,censor_time1))


#Cancer Specific - LUSC
xg_overfit_preds_LUSC <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUSC))
xg_testingdata_LUSC <- Cindex(xg_overfit_preds_LUSC, Surv(multimodal_xg_test_LUSC$censor_time1,multimodal_xg_test_LUSC$vital_status))

#LUAD
xg_overfit_preds_LUAD <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUAD))
xg_testingdata_LUAD <- Cindex(xg_overfit_preds_LUAD, Surv(multimodal_xg_test_LUAD$censor_time1,multimodal_xg_test_LUAD$vital_status))
```



### Results 

```{r, echo=FALSE}
#Not overfit
elastic_res <- glm_c_testingdata
cox_res <- cox_testingdata
xg_res <- C_xg_overfit
rsf_res <- rf2_cindex_testingdata

Cindex_Test <- c(xg_res, cox_res, elastic_res, rsf_res)
Models <- c("XGBoost", "Cox", "ElasticNet", "RSF")
Data_Types <- rep("Clinical", 4)
results.df <- as.data.frame(cbind(Models, Cindex_Test))
colnames(results.df)[2] <- "Cindex_Test"
results.df <- cbind(results.df, Data_Types)
results.df$Cindex_Test <- as.numeric(results.df$Cindex_Test)
Dimension_Reduction <- rep(autoencoder_type, 4)
results.df <- cbind(results.df, Dimension_Reduction)

Cindex_LUSC <- c(xg_testingdata_LUSC, cox_testingdata_LUSC, elastic_testingdata_LUSC, rf_testingdata_LUSC) 
Cindex_LUAD <- c(xg_testingdata_LUAD, cox_testingdata_LUAD, elastic_testingdata_LUAD, rf_testingdata_LUAD)
results.df <- as.data.frame(cbind(results.df, Cindex_LUSC))
results.df <- as.data.frame(cbind(results.df, Cindex_LUAD))


results.df$Cindex_Test <-round(results.df$Cindex_Test, 3)
results.df_clin <- results.df
```

# VISUALIZATIONS

## Stratified by Survival Status - 5 Year Survival

```{r, echo=FALSE}
library(Rtsne)

X <- multimodal_prediction

X <- merge(X, barcode_disease_mapping, by="barcode")

X_deceased <- X %>% filter(vital_status=="Dead" & censor_time < 1825)
X_alive_5years <- X %>% filter(censor_time > 1825)
X_alive_5years$vital_status <- "Alive"

X <- rbind(X_deceased, X_alive_5years)



disease_vital <- NULL
for (i in 1:nrow(X)) {
  if (X$vital_status[i]=="Dead" & X$name.y[i]=="Lung Squamous Cell Carcinoma") {
  disease_vital[i] <- "LUSC - Dead"
}
 if (X$vital_status[i]=="Alive" & X$name.y[i]=="Lung Squamous Cell Carcinoma") {
  disease_vital[i] <- "LUSC - Alive"
}
 if (X$vital_status[i]=="Dead" & X$name.y[i]=="Lung Adenocarcinoma") {
  disease_vital[i] <- "LUAD - Dead"
}
 if (X$vital_status[i]=="Alive" & X$name.y[i]=="Lung Adenocarcinoma") {
  disease_vital[i] <- "LUAD - Alive"
}
}

X <- as.data.frame(cbind(X, disease_vital))

vital_status_vector_CLIN <- X$vital_status
disease_vitalstatus <- X$disease_vital

X <- X %>%
  dplyr::select(-c(vital_status, censor_time1, TestTrain, barcode, disease_vital, name.y, name.x))
X <- scale(X)


LUSC_PCA_CLIN <- prcomp(X)

ggplot(as.data.frame(LUSC_PCA_CLIN$x), aes(x = PC1, y = PC2, col = factor(disease_vitalstatus))) + geom_point() + scale_colour_manual(name = "Survival Status", values = c("#EE3377", "dark red", "cadetblue1", "blue4"))  + labs(title="Clinical") +theme_bw()
```


## T-SNE 5 Years

```{r, echo=FALSE}
set.seed(1234) # reproducibility

X <- X[, !duplicated(colnames(X))]

tsne <- Rtsne(X, dims = 2, perplexity = 10, check_duplicates=FALSE)
t.df_CLIN <- as.data.frame(tsne$Y)
colnames(t.df_CLIN) <- c("V1", "V2")
ggplot(t.df_CLIN, aes(x = V1, y = V2, color = factor(disease_vitalstatus))) + geom_point()+ labs(title="Clinical - tSNE") + scale_color_discrete(name = "Survival Status")
```

# Multimodal Analysis 

```{r, echo=FALSE}
#Setting up the dataframe - all of these were saved at previous points in the pipeline
mirna_train <- mirna_train %>% dplyr::select(-c(vital_status, censor_time))
gene_train <- gene_train %>% dplyr::select(-c(vital_status, censor_time))
lnc_train <- lnc_train %>% dplyr::select(-c(vital_status, censor_time))
meth_train <- meth_train %>% dplyr::select(-c(vital_status, censor_time))

mirna_test <- mirna_test %>% dplyr::select(-c(vital_status, censor_time))
gene_test <- gene_test %>% dplyr::select(-c(vital_status, censor_time))
lnc_test <- lnc_test %>% dplyr::select(-c(vital_status, censor_time))
meth_test <- meth_test %>% dplyr::select(-c(vital_status, censor_time))

#Training Data Combination
index <- which(colnames(clin_train)=="censor_time1")
colnames(clin_train)[index] <- "censor_time"
  
df2 <- merge(lnc_train, mirna_train, by="barcode")
df2 <- merge(df2, gene_train, by="barcode")
df2 <- merge(df2, meth_train, by="barcode")
multimodal_train <- merge(df2,clin_train, by="barcode")

#Testing Data Combination
index <- which(colnames(clin_test)=="censor_time1")
colnames(clin_test)[index] <- "censor_time"

df3 <- merge(lnc_test, mirna_test, by ="barcode")
df3 <- merge(df3, gene_test, by="barcode")
df3 <- merge(df3, meth_test, by="barcode")
multimodal_test <- merge(df3, clin_test,by="barcode")

#Making sure the correct test and train sets being generated
if (nrow(multimodal_train)!=585) {
  stop()
}

if (nrow(multimodal_test)!=147) {
  stop()
}

multimodal_LUSC_indices <- merge(multimodal_train, barcode_disease_mapping, by="barcode")
multimodal_LUAD_indices_train <- which(multimodal_LUSC_indices$name.y=="Lung Adenocarcinoma") 
multimodal_LUSC_indices_train <- which(multimodal_LUSC_indices$name.y=="Lung Squamous Cell Carcinoma")

multimodal_LUSC_indices_test_multi <- merge(multimodal_test, barcode_disease_mapping, by="barcode")
multimodal_LUAD_indices_test <- which(multimodal_LUSC_indices_test_multi$name.y=="Lung Adenocarcinoma") 
multimodal_LUSC_indices_test <- which(multimodal_LUSC_indices_test_multi$name.y=="Lung Squamous Cell Carcinoma")

multimodal_test <- multimodal_test %>% dplyr::select(-c(barcode))
multimodal_train <- multimodal_train %>% dplyr::select(-c(barcode))

multimodal_test <- as.data.frame(sapply(multimodal_test, as.numeric))
multimodal_train <- as.data.frame(sapply(multimodal_train, as.numeric))
```

### Cox PH Model

```{r,echo=FALSE}
#Overall Model
cox <- coxph(Surv(censor_time, vital_status) ~ ., data=multimodal_train)
cox_fit <- survfit(cox)
cox_overfit_preds <- predict(cox, multimodal_test)
cox_testingdata <- Cindex(cox_overfit_preds, Surv(multimodal_test$censor_time,multimodal_test$vital_status))

#Cancer Specific - LUSC
multimodal_test_LUSC <- multimodal_test[c(multimodal_LUSC_indices_test), ]
cox_overfit_preds_LUSC <- predict(cox, multimodal_test_LUSC)
cox_testingdata_LUSC <- Cindex(cox_overfit_preds_LUSC, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
multimodal_test_LUAD <- multimodal_test[c(multimodal_LUAD_indices_test), ]
cox_overfit_preds_LUAD <- predict(cox, multimodal_test_LUAD)
cox_testingdata_LUAD <- Cindex(cox_overfit_preds_LUAD, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))
```

### Random Survival Forests

```{r, echo=FALSE}
multimodal_train$vital_status <- multimodal_train$vital_status - 1
multimodal_test$vital_status <- multimodal_test$vital_status-1

rfsrc.pbc_plot <- rfsrc(Surv(censor_time, vital_status) ~ ., # survival object
                  data = multimodal_train, # data
                  nsplit = 1, # number of random splits to consider
                  tree.err = FALSE
                  ,importance = TRUE
                  )
rfsrc.pbc.testplot <- predict(rfsrc.pbc_plot,multimodal_test)
rf2_cindex_testingdata <- Cindex(rfsrc.pbc.testplot$predicted, Surv(multimodal_test$censor_time,multimodal_test$vital_status))

multimodal_test_LUAD$vital_status <- multimodal_test_LUAD$vital_status - 1
multimodal_test_LUSC$vital_status <- multimodal_test_LUSC$vital_status - 1

#Cancer Specific - LUSC
rf_overfit_preds_LUSC <- predict(rfsrc.pbc_plot, multimodal_test_LUSC)
rf_testingdata_LUSC <- Cindex(rf_overfit_preds_LUSC$predicted, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
rf_overfit_preds_LUAD <- predict(rfsrc.pbc_plot, multimodal_test_LUAD)
rf_testingdata_LUAD <- Cindex(rf_overfit_preds_LUAD$predicted, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))
```

### Multimodal Variable Importance Plot

```{r, echo=FALSE}
#Top Ten Plot
imp_plot <- gg_vimp(rfsrc.pbc_plot)
imp_plot <- imp_plot[1:30, 1:ncol(imp_plot)]
imp_plot$vars <- as.character(imp_plot$vars)
imp_plot <- imp_plot[order(imp_plot$vars),]

variable_importance <- imp_plot %>%
  ggplot(aes(x=vimp, y=reorder(vars, vimp))) + geom_col(fill="blue") + labs(y=NULL, x="Variable Importance", title = "Top 30 Most Important Variables")
variable_importance
```

### Elastic Net

```{r, echo=FALSE}
multimodal_cv <-as.matrix(multimodal_train[1:nrow(multimodal_train), 1:(ncol(multimodal_train)-2)])
multimodal_cv_test <-as.matrix(multimodal_test[1:nrow(multimodal_test), 1:(ncol(multimodal_test)-2)])

cv.fit = cv.glmnet(multimodal_cv,  Surv(multimodal_train$censor_time,multimodal_train$vital_status),
                   alpha =0.5, # lasso: alpha = 1; ridge: alpha=0
                   family = "cox", maxit = 10000, type.measure = "C")
#plot(cv.fit)
est.coef = coef(cv.fit, s = cv.fit$lambda.min) # returns the p length coefficient vector
active.k = which(est.coef != 0)

glm_pred <- predict(cv.fit, multimodal_cv_test)
glm_c_testingdata <- Cindex(glm_pred,Surv(multimodal_test$censor_time,multimodal_test$vital_status))

#Individual Cancer Types
multimodal_test_LUSC_cv <-as.matrix(multimodal_test_LUSC[1:nrow(multimodal_test_LUSC ), 1:(ncol(multimodal_test_LUSC)-2)])
multimodal_test_LUAD_cv <-as.matrix(multimodal_test_LUAD[1:nrow(multimodal_test_LUAD ), 1:(ncol(multimodal_test_LUAD)-2)])


#Cancer Specific - LUSC
elastic_overfit_preds_LUSC <- predict(cv.fit, multimodal_test_LUSC_cv)
elastic_testingdata_LUSC <- Cindex(elastic_overfit_preds_LUSC, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
elastic_overfit_preds_LUAD <- predict(cv.fit, multimodal_test_LUAD_cv)
elastic_testingdata_LUAD <- Cindex(elastic_overfit_preds_LUAD, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))
```

### XG Boost

```{r, echo=FALSE}
multimodal_xg <- multimodal_train
multimodal_xg$vital_status <- as.integer(multimodal_xg$vital_status)
label <- ifelse(multimodal_xg$vital_status == 1, multimodal_xg$censor_time, -multimodal_xg$censor_time)

matrix_xg <- as.matrix(multimodal_xg[1:nrow(multimodal_xg), 1:(ncol(multimodal_xg)-2)])

prediction_df <- multimodal_xg[1:nrow(multimodal_train),
                               (ncol(multimodal_train)-1):ncol(multimodal_train)]
#Overall for Plot:
val_ind <- sample.int(nrow(multimodal_xg), 0.1 * nrow(multimodal_xg))
x_train <- as.matrix(multimodal_xg [-val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")])

x_label <- label[-val_ind]
x_val <- xgb.DMatrix(as.matrix(multimodal_xg[val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")]), label = label[val_ind])

#Testing Data
multimodal_xg_test <- multimodal_test
multimodal_xg_test$vital_status <- as.integer(multimodal_xg_test$vital_status)
test_label <- ifelse(multimodal_xg_test$vital_status == 1, multimodal_xg_test$censor_time, -multimodal_xg_test$censor_time)
multimodal_test_xgboost <- multimodal_xg_test %>% dplyr::select(-c(vital_status,censor_time))

surv_xgboost_model_all <- xgb.train.surv(
  params = list(
    objective = "survival:cox",
    eval_metric = "cox-nloglik",
    eta = 0.001, subsample=.5,
    #,learning_rate=0.001
               max_depth=2
  ), data = matrix_xg, label = label,
  watchlist = list(val3 = x_val),
  nrounds = 50
  , early_stopping_rounds = 15

)

# predict survival curves
xg_df = predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost))
C_xg_overfit <- Cindex(xg_df, Surv(multimodal_xg_test$censor_time, multimodal_xg_test$vital_status))

#Examining risk scores (should be somewhat spread out)
risk_scores <- predict(object = surv_xgboost_model_all, newdata = as.matrix(multimodal_test_xgboost), type = "risk")
hist(risk_scores)

#Testing Data - Individual Cancer Types
multimodal_xg_test_LUSC <- multimodal_test_LUSC
multimodal_xg_test_LUSC$vital_status <- as.integer(multimodal_xg_test_LUSC$vital_status)
test_label <- ifelse(multimodal_xg_test_LUSC$vital_status == 1, multimodal_xg_test_LUSC$censor_time, -multimodal_xg_test_LUSC$censor_time)
multimodal_test_xgboost_LUSC <- multimodal_xg_test_LUSC %>% dplyr::select(-c(vital_status,censor_time))

#Testing Data
multimodal_xg_test_LUAD <- multimodal_test_LUAD
multimodal_xg_test_LUAD$vital_status <- as.integer(multimodal_xg_test_LUAD$vital_status)
test_label <- ifelse(multimodal_xg_test_LUAD$vital_status == 1, multimodal_xg_test_LUAD$censor_time, -multimodal_xg_test_LUAD$censor_time)
multimodal_test_xgboost_LUAD <- multimodal_xg_test_LUAD %>% dplyr::select(-c(vital_status,censor_time))


#Cancer Specific - LUSC
xg_overfit_preds_LUSC <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUSC))
xg_testingdata_LUSC <- Cindex(xg_overfit_preds_LUSC, Surv(multimodal_xg_test_LUSC$censor_time,multimodal_xg_test_LUSC$vital_status))

#LUAD
xg_overfit_preds_LUAD <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUAD))
xg_testingdata_LUAD <- Cindex(xg_overfit_preds_LUAD, Surv(multimodal_xg_test_LUAD$censor_time,multimodal_xg_test_LUAD$vital_status))
```



### Results 

```{r, echo=FALSE}
#Not overfit
elastic_res <- glm_c_testingdata
cox_res <- cox_testingdata
xg_res <- C_xg_overfit
rsf_res <- rf2_cindex_testingdata

Cindex_Test <- c(xg_res, cox_res, elastic_res, rsf_res)
Models <- c("XGBoost", "Cox", "ElasticNet", "RSF")
Data_Types <- rep("Multimodal", 4)
results.df <- as.data.frame(cbind(Models, Cindex_Test))
colnames(results.df)[2] <- "Cindex_Test"
results.df <- cbind(results.df, Data_Types)
results.df$Cindex_Test <- as.numeric(results.df$Cindex_Test)
Dimension_Reduction <- rep(autoencoder_type, 4)
results.df <- cbind(results.df, Dimension_Reduction)

Cindex_LUSC <- c(xg_testingdata_LUSC, cox_testingdata_LUSC, elastic_testingdata_LUSC, rf_testingdata_LUSC) 
Cindex_LUAD <- c(xg_testingdata_LUAD, cox_testingdata_LUAD, elastic_testingdata_LUAD, rf_testingdata_LUAD)
results.df <- as.data.frame(cbind(results.df, Cindex_LUSC))
results.df <- as.data.frame(cbind(results.df, Cindex_LUAD))

results.df$Cindex_Test <-round(results.df$Cindex_Test, 3)
results.df_multimodal <- results.df 
results.df_multimodal
```

# VISUALIZATIONS

## Stratified by Survival Status - 5 Year Survival

```{r, echo=FALSE}
multimodal_prediction <- rbind(multimodal_LUSC_indices, multimodal_LUSC_indices_test_multi)

X <- multimodal_prediction
X <- merge(X, barcode_disease_mapping, by="barcode")

X$vital_status <- as.factor(X$vital_status)
levels(X$vital_status) <- c("Alive", "Dead")

X_deceased <- X %>% filter(vital_status=="Dead" & censor_time < 1825)
X_alive_5years <- X %>% filter(censor_time > 1825)
X_alive_5years$vital_status <- "Alive"

X <- rbind(X_deceased, X_alive_5years)


disease_vital <- NULL
for (i in 1:nrow(X)) {
  if (X$vital_status[i]=="Dead" & X$name.y[i]=="Lung Squamous Cell Carcinoma") {
  disease_vital[i] <- "LUSC - Dead"
}
 if (X$vital_status[i]=="Alive" & X$name.y[i]=="Lung Squamous Cell Carcinoma") {
  disease_vital[i] <- "LUSC - Alive"
}
 if (X$vital_status[i]=="Dead" & X$name.y[i]=="Lung Adenocarcinoma") {
  disease_vital[i] <- "LUAD - Dead"
}
 if (X$vital_status[i]=="Alive" & X$name.y[i]=="Lung Adenocarcinoma") {
  disease_vital[i] <- "LUAD - Alive"
}
}

X <- as.data.frame(cbind(X, disease_vital))
vital_status_vector_MULTI <- X$vital_status
disease_vitalstatus <- X$disease_vital

X <- X %>%
  dplyr::select(-c(vital_status, censor_time, barcode, disease_vital, name.y, name.x, name))

LUSC_PCA_MULTI <- prcomp(X)

ggplot(as.data.frame(LUSC_PCA_MULTI$x), aes(x = PC1, y = PC2, col = factor(disease_vitalstatus))) + geom_point() + scale_colour_manual(name = "Survival Status", values = c("#EE3377", "dark red", "cadetblue1", "blue4"))  + labs(title="Multimodal") +theme_bw()
```

## T-SNE 5 Years

```{r, echo=FALSE}
set.seed(1234) # reproducibility
tsne <- Rtsne(X, dims = 2, perplexity = 10, check_duplicates=FALSE)
t.df_MULTI <- as.data.frame(tsne$Y)
colnames(t.df_MULTI) <- c("V1", "V2")
ggplot(t.df_MULTI, aes(x = V1, y = V2, color = factor(disease_vitalstatus))) + geom_point()+ labs(title="Multimodal - tSNE") + scale_color_discrete(name = "Survival Status")
```

## Multimodal WITH No Clinical Data Analysis

```{r, echo=FALSE}
#Getting rid of clinical features, but keeping all of the biological ones from above
multinoclin_train <- multimodal_train %>% dplyr::select(c(starts_with("L"),starts_with("M"),starts_with("G"),starts_with("L"),starts_with("m"), vital_status, censor_time))
multinoclin_train <- multinoclin_train %>% dplyr::select(-c(gender))

multinoclin_test <- multimodal_test %>% dplyr::select(c(starts_with("L"),starts_with("M"),starts_with("G"),starts_with("L"),starts_with("m"), vital_status, censor_time))
multinoclin_test <- multinoclin_test %>% dplyr::select(-c(gender))
```

### Cox PH Model

```{r,echo=FALSE}
#Overall Model
cox <- coxph(Surv(censor_time, vital_status) ~ ., data=multinoclin_train)
cox_fit <- survfit(cox)
cox_overfit_preds <- predict(cox, multinoclin_test)
cox_testingdata <- Cindex(cox_overfit_preds, Surv(multinoclin_test$censor_time,multinoclin_test$vital_status))

#Cancer Specific - LUSC
multinoclin_test_LUSC <- multinoclin_test[c(multimodal_LUSC_indices_test), ]
cox_overfit_preds_LUSC <- predict(cox, multinoclin_test_LUSC)
cox_testingdata_LUSC <- Cindex(cox_overfit_preds_LUSC, Surv(multinoclin_test_LUSC$censor_time,multinoclin_test_LUSC$vital_status))

#LUAD
multinoclin_test_LUAD <- multinoclin_test[c(multimodal_LUAD_indices_test), ]
cox_overfit_preds_LUAD <- predict(cox, multinoclin_test_LUAD)
cox_testingdata_LUAD <- Cindex(cox_overfit_preds_LUAD, Surv(multinoclin_test_LUAD$censor_time,multinoclin_test_LUAD$vital_status))
```

### Random Survival Forests

```{r, echo=FALSE}
rfsrc.pbc_plot <- rfsrc(Surv(censor_time, vital_status) ~ ., # survival object
                  data = multinoclin_train, # data
                  nsplit = 1, # number of random splits to consider
                  tree.err = FALSE
                  ,importance = TRUE
                  )
rfsrc.pbc.testplot <- predict(rfsrc.pbc_plot,multinoclin_test)
rf2_cindex_testingdata <- Cindex(rfsrc.pbc.testplot$predicted, Surv(multinoclin_test$censor_time,multinoclin_test$vital_status))

#Cancer Specific - LUSC
rf_overfit_preds_LUSC <- predict(rfsrc.pbc_plot, multinoclin_test_LUSC)
rf_testingdata_LUSC <- Cindex(rf_overfit_preds_LUSC$predicted, Surv(multinoclin_test_LUSC$censor_time,multinoclin_test_LUSC$vital_status))

#LUAD
rf_overfit_preds_LUAD <- predict(rfsrc.pbc_plot, multinoclin_test_LUAD)
rf_testingdata_LUAD <- Cindex(rf_overfit_preds_LUAD$predicted, Surv(multinoclin_test_LUAD$censor_time,multinoclin_test_LUAD$vital_status))
```

### Elastic Net

```{r, echo=FALSE}
multimodal_cv <-as.matrix(multinoclin_train[1:nrow(multinoclin_train), 1:(ncol(multinoclin_train)-2)])
multimodal_cv_test <-as.matrix(multinoclin_test[1:nrow(multinoclin_test), 1:(ncol(multinoclin_test)-2)])

cv.fit = cv.glmnet(multimodal_cv,  Surv(multinoclin_train$censor_time,multinoclin_train$vital_status),
                   alpha =0.5, # lasso: alpha = 1; ridge: alpha=0
                   family = "cox", maxit = 10000, type.measure = "C")
#plot(cv.fit)
est.coef = coef(cv.fit, s = cv.fit$lambda.min) # returns the p length coefficient vector
active.k = which(est.coef != 0)

glm_pred <- predict(cv.fit, multimodal_cv_test)
glm_c_testingdata <- Cindex(glm_pred,Surv(multinoclin_test$censor_time,multinoclin_test$vital_status))

#Individual Cancer Types
multimodal_test_LUSC_cv <-as.matrix(multinoclin_test_LUSC[1:nrow(multinoclin_test_LUSC ), 1:(ncol(multinoclin_test_LUSC)-2)])
multimodal_test_LUAD_cv <-as.matrix(multinoclin_test_LUAD[1:nrow(multinoclin_test_LUAD ), 1:(ncol(multinoclin_test_LUAD)-2)])


#Cancer Specific - LUSC
elastic_overfit_preds_LUSC <- predict(cv.fit, multimodal_test_LUSC_cv)
elastic_testingdata_LUSC <- Cindex(elastic_overfit_preds_LUSC, Surv(multinoclin_test_LUSC$censor_time,multinoclin_test_LUSC$vital_status))

#LUAD
elastic_overfit_preds_LUAD <- predict(cv.fit, multimodal_test_LUAD_cv)
elastic_testingdata_LUAD <- Cindex(elastic_overfit_preds_LUAD, Surv(multinoclin_test_LUAD$censor_time,multinoclin_test_LUAD$vital_status))
```


### XG Boost

```{r, echo=FALSE}
multimodal_xg <- multinoclin_train
multimodal_xg$vital_status <- as.integer(multimodal_xg$vital_status)
label <- ifelse(multimodal_xg$vital_status == 1, multimodal_xg$censor_time, -multimodal_xg$censor_time)

matrix_xg <- as.matrix(multimodal_xg[1:nrow(multimodal_xg), 1:(ncol(multimodal_xg)-2)])

prediction_df <- multimodal_xg[1:nrow(multinoclin_train),
                               (ncol(multinoclin_train)-1):ncol(multinoclin_train)]

#Overall for Plot:
val_ind <- sample.int(nrow(multimodal_xg), 0.1 * nrow(multimodal_xg))
x_train <- as.matrix(multimodal_xg [-val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")])

x_label <- label[-val_ind]
x_val <- xgb.DMatrix(as.matrix(multimodal_xg[val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")]), label = label[val_ind])

#Testing Data
multimodal_xg_test <- multinoclin_test
multimodal_xg_test$vital_status <- as.integer(multimodal_xg_test$vital_status)
test_label <- ifelse(multimodal_xg_test$vital_status == 1, multimodal_xg_test$censor_time, -multimodal_xg_test$censor_time)
multimodal_test_xgboost <- multimodal_xg_test %>% dplyr::select(-c(vital_status,censor_time))

surv_xgboost_model_all <- xgb.train.surv(
  params = list(
    objective = "survival:cox",
    eval_metric = "cox-nloglik",
    eta = 0.001, subsample=.5,
               max_depth=2
  ), data = matrix_xg, label = label,
  watchlist = list(val3 = x_val),
  nrounds = 50
  , early_stopping_rounds = 15

)

# predict survival curves
xg_df = predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost))
C_xg_overfit <- Cindex(xg_df, Surv(multimodal_xg_test$censor_time, multimodal_xg_test$vital_status))

#Examining risk scores (should be somewhat spread out)
risk_scores <- predict(object = surv_xgboost_model_all, newdata = as.matrix(multimodal_test_xgboost), type = "risk")
hist(risk_scores)

#Testing Data - Individual Cancer Types
multimodal_xg_test_LUSC <- multinoclin_test_LUSC
multimodal_xg_test_LUSC$vital_status <- as.integer(multimodal_xg_test_LUSC$vital_status)
test_label <- ifelse(multimodal_xg_test_LUSC$vital_status == 1, multimodal_xg_test_LUSC$censor_time, -multimodal_xg_test_LUSC$censor_time)
multimodal_test_xgboost_LUSC <- multimodal_xg_test_LUSC %>% dplyr::select(-c(vital_status,censor_time))

#Testing Data
multimodal_xg_test_LUAD <- multinoclin_test_LUAD
multimodal_xg_test_LUAD$vital_status <- as.integer(multimodal_xg_test_LUAD$vital_status)
test_label <- ifelse(multimodal_xg_test_LUAD$vital_status == 1, multimodal_xg_test_LUAD$censor_time, -multimodal_xg_test_LUAD$censor_time)
multimodal_test_xgboost_LUAD <- multimodal_xg_test_LUAD %>% dplyr::select(-c(vital_status,censor_time))

#Cancer Specific - LUSC
xg_overfit_preds_LUSC <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUSC))
xg_testingdata_LUSC <- Cindex(xg_overfit_preds_LUSC, Surv(multimodal_xg_test_LUSC$censor_time,multimodal_xg_test_LUSC$vital_status))

#LUAD
xg_overfit_preds_LUAD <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUAD))
xg_testingdata_LUAD <- Cindex(xg_overfit_preds_LUAD, Surv(multimodal_xg_test_LUAD$censor_time,multimodal_xg_test_LUAD$vital_status))
```



### Results 

```{r, echo=FALSE}
#Not overfit
elastic_res <- glm_c_testingdata
cox_res <- cox_testingdata
xg_res <- C_xg_overfit
rsf_res <- rf2_cindex_testingdata

Cindex_Test <- c(xg_res, cox_res, elastic_res, rsf_res)
Models <- c("XGBoost", "Cox", "ElasticNet", "RSF")
Data_Types <- rep("Multi (No Clin)", 4)
results.df <- as.data.frame(cbind(Models, Cindex_Test))
colnames(results.df)[2] <- "Cindex_Test"
results.df <- cbind(results.df, Data_Types)
results.df$Cindex_Test <- as.numeric(results.df$Cindex_Test)
Dimension_Reduction <- rep(autoencoder_type, 4)
results.df <- cbind(results.df, Dimension_Reduction)

Cindex_LUSC <- c(xg_testingdata_LUSC, cox_testingdata_LUSC, elastic_testingdata_LUSC, rf_testingdata_LUSC) 
Cindex_LUAD <- c(xg_testingdata_LUAD, cox_testingdata_LUAD, elastic_testingdata_LUAD, rf_testingdata_LUAD)
results.df <- as.data.frame(cbind(results.df, Cindex_LUSC))
results.df <- as.data.frame(cbind(results.df, Cindex_LUAD))

results.df$Cindex_Test <-round(results.df$Cindex_Test, 3)
results.df_multimodal_noclin <- results.df 
results.df_multimodal_noclin
```


# VISUALIZATIONS

## Stratified by Survival Status - 5 Year Survival

```{r, echo=FALSE}
multimodal_prediction <- rbind(multimodal_LUSC_indices, multimodal_LUSC_indices_test_multi)

X <- multimodal_prediction %>% dplyr::select(c(starts_with("L"),starts_with("M"),starts_with("G"),starts_with("L"),starts_with("m"), vital_status, censor_time, barcode))
X <- X %>% dplyr::select(-c(gender))

X <- merge(X, barcode_disease_mapping, by="barcode")

X$vital_status <- as.factor(X$vital_status)
levels(X$vital_status) <- c("Alive", "Dead")

X_deceased <- X %>% filter(vital_status=="Dead" & censor_time < 1825)
X_alive_5years <- X %>% filter(censor_time > 1825)
X_alive_5years$vital_status <- "Alive"

X <- rbind(X_deceased, X_alive_5years)


disease_vital <- NULL
for (i in 1:nrow(X)) {
  if (X$vital_status[i]=="Dead" & X$name[i]=="Lung Squamous Cell Carcinoma") {
  disease_vital[i] <- "LUSC - Dead"
}
 if (X$vital_status[i]=="Alive" & X$name[i]=="Lung Squamous Cell Carcinoma") {
  disease_vital[i] <- "LUSC - Alive"
}
 if (X$vital_status[i]=="Dead" & X$name[i]=="Lung Adenocarcinoma") {
  disease_vital[i] <- "LUAD - Dead"
}
 if (X$vital_status[i]=="Alive" & X$name[i]=="Lung Adenocarcinoma") {
  disease_vital[i] <- "LUAD - Alive"
}
}


X <- as.data.frame(cbind(X, disease_vital))

vital_status_vector_MULTINOCLIN <- X$vital_status
disease_vitalstatus <- X$disease_vital

X <- X %>%
  dplyr::select(-c(vital_status, censor_time, barcode, disease_vital, name))


LUSC_PCA_MULTINOCLIN <- prcomp(X)

ggplot(as.data.frame(LUSC_PCA_MULTINOCLIN$x), aes(x = PC1, y = PC2, col = factor(disease_vitalstatus))) + geom_point() + scale_color_discrete(name = "Survival Status") + labs(title="5 Year Survival - Multimodal")
```

## T-SNE 5 Years

```{r, echo=FALSE}
set.seed(1234) # reproducibility
tsne <- Rtsne(X, dims = 2, perplexity = 10, check_duplicates=FALSE)
t.df_MULTI <- as.data.frame(tsne$Y)
colnames(t.df_MULTI) <- c("V1", "V2")
ggplot(t.df_MULTI, aes(x = V1, y = V2, color = factor(disease_vitalstatus))) + geom_point()+ labs(title="Multimodal - tSNE") + scale_color_discrete(name = "Survival Status")
```


### Trying Another Round of Feature Selection on Multimodal Data

```{r, echo=FALSE}
if (another_round_of_featureselection == "yes") {
y <- multimodal_train$vital_status
censor_time <- multimodal_train$censor_time
association_mat <- multimodal_train %>% dplyr::select(-c(vital_status, censor_time))
association_mat <- as.matrix(association_mat)

set.seed(10)
association <- uni.selection(t.vec = censor_time1, d.vec = y, X.mat=association_mat,
                          P.value = p_value_featselection, randomize=FALSE, K=5)

col_filtered <- rownames(as.data.frame(association$P))
multimodal_train <- multimodal_train %>% dplyr::select(c(all_of(col_filtered), vital_status, censor_time))
multimodal_test <- multimodal_test %>% dplyr::select(c(all_of(col_filtered), vital_status, censor_time))

multimodal_train <- multimodal_train[, !duplicated(colnames(multimodal_train))]
multimodal_test <- multimodal_test[, !duplicated(colnames(multimodal_train))]

multimodal_train_ <- multimodal_train %>% dplyr::select(c(all_of(col_filtered), vital_status, censor_time))
multimodal_test <- multimodal_test %>% dplyr::select(c(all_of(col_filtered), vital_status, censor_time))


multimodal_test_LUAD <- multimodal_test_LUAD %>% dplyr::select(c(all_of(col_filtered), vital_status, censor_time))
multimodal_test_LUSC <- multimodal_test_LUSC %>% dplyr::select(c(all_of(col_filtered), vital_status, censor_time))

#Overall Model
cox <- coxph(Surv(censor_time, vital_status) ~ ., data=multimodal_train)
cox_fit <- survfit(cox)
cox_overfit_preds <- predict(cox, multimodal_test)
cox_testingdata <- Cindex(cox_overfit_preds, Surv(multimodal_test$censor_time,multimodal_test$vital_status))


#Cancer Specific - LUSC
cox_overfit_preds_LUSC <- predict(cox, multimodal_test_LUSC)
cox_testingdata_LUSC <- Cindex(cox_overfit_preds_LUSC, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
cox_overfit_preds_LUAD <- predict(cox, multimodal_test_LUAD)
cox_testingdata_LUAD <- Cindex(cox_overfit_preds_LUAD, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))

rfsrc.pbc_plot <- rfsrc(Surv(censor_time, vital_status) ~ ., # survival object
                  data = multimodal_train, # data
                  nsplit = 1, # number of random splits to consider
                  tree.err = FALSE
                  ,importance = TRUE
                  )
rfsrc.pbc.testplot <- predict(rfsrc.pbc_plot,multimodal_test)
rf2_cindex_testingdata <- Cindex(rfsrc.pbc.testplot$predicted, Surv(multimodal_test$censor_time,multimodal_test$vital_status))

#Cancer Specific - LUSC
rf_overfit_preds_LUSC <- predict(rfsrc.pbc_plot, multimodal_test_LUSC)
rf_testingdata_LUSC <- Cindex(rf_overfit_preds_LUSC$predicted, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
rf_overfit_preds_LUAD <- predict(rfsrc.pbc_plot, multimodal_test_LUAD)
rf_testingdata_LUAD <- Cindex(rf_overfit_preds_LUAD$predicted, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))

multimodal_cv <-as.matrix(multimodal_train[1:nrow(multimodal_train), 1:(ncol(multimodal_train)-2)])
multimodal_cv_test <-as.matrix(multimodal_test[1:nrow(multimodal_test), 1:(ncol(multimodal_test)-2)])

cv.fit = cv.glmnet(multimodal_cv,  Surv(multimodal_train$censor_time,multimodal_train$vital_status),
                   alpha =0.5, # lasso: alpha = 1; ridge: alpha=0
                   family = "cox", maxit = 10000, type.measure = "C")
#plot(cv.fit)
est.coef = coef(cv.fit, s = cv.fit$lambda.min) # returns the p length coefficient vector
active.k = which(est.coef != 0)

glm_pred <- predict(cv.fit, multimodal_cv_test)
glm_c_testingdata <- Cindex(glm_pred,Surv(multimodal_test$censor_time,multimodal_test$vital_status))

#Individual Cancer Types
multimodal_test_LUSC_cv <-as.matrix(multimodal_test_LUSC[1:nrow(multimodal_test_LUSC ), 1:(ncol(multimodal_test_LUSC)-2)])
multimodal_test_LUAD_cv <-as.matrix(multimodal_test_LUAD[1:nrow(multimodal_test_LUAD ), 1:(ncol(multimodal_test_LUAD)-2)])


#Cancer Specific - LUSC
elastic_overfit_preds_LUSC <- predict(cv.fit, multimodal_test_LUSC_cv)
elastic_testingdata_LUSC <- Cindex(elastic_overfit_preds_LUSC, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
elastic_overfit_preds_LUAD <- predict(cv.fit, multimodal_test_LUAD_cv)
elastic_testingdata_LUAD <- Cindex(elastic_overfit_preds_LUAD, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))


multimodal_xg <- multimodal_train
multimodal_xg$vital_status <- as.integer(multimodal_xg$vital_status)
label <- ifelse(multimodal_xg$vital_status == 1, multimodal_xg$censor_time, -multimodal_xg$censor_time)

matrix_xg <- as.matrix(multimodal_xg[1:nrow(multimodal_xg), 1:(ncol(multimodal_xg)-2)])

prediction_df <- multimodal_xg[1:nrow(multimodal_train),
                               (ncol(multimodal_train)-1):ncol(multimodal_train)]

#Overall for Plot:
val_ind <- sample.int(nrow(multimodal_xg), 0.1 * nrow(multimodal_xg))
x_train <- as.matrix(multimodal_xg [-val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")])

x_label <- label[-val_ind]
x_val <- xgb.DMatrix(as.matrix(multimodal_xg[val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")]), label = label[val_ind])

#Testing Data
multimodal_xg_test <- multimodal_test
multimodal_xg_test$vital_status <- as.integer(multimodal_xg_test$vital_status)
test_label <- ifelse(multimodal_xg_test$vital_status == 1, multimodal_xg_test$censor_time, -multimodal_xg_test$censor_time)
multimodal_test_xgboost <- multimodal_xg_test %>% dplyr::select(-c(vital_status,censor_time))

surv_xgboost_model_all <- xgb.train.surv(
  params = list(
    objective = "survival:cox",
    eval_metric = "cox-nloglik",
    eta = 0.001, subsample=.5,
    #,learning_rate=0.001
               max_depth=2
  ), data = matrix_xg, label = label,
  watchlist = list(val3 = x_val),
  nrounds = 50
  , early_stopping_rounds = 15

)

# predict survival curves
xg_df = predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost))
C_xg_overfit <- Cindex(xg_df, Surv(multimodal_xg_test$censor_time, multimodal_xg_test$vital_status))

#Examining risk scores (should be somewhat spread out)
risk_scores <- predict(object = surv_xgboost_model_all, newdata = as.matrix(multimodal_test_xgboost), type = "risk")
hist(risk_scores)


#Testing Data - Individual Cancer Types
multimodal_xg_test_LUSC <- multimodal_test_LUSC
multimodal_xg_test_LUSC$vital_status <- as.integer(multimodal_xg_test_LUSC$vital_status)
test_label <- ifelse(multimodal_xg_test_LUSC$vital_status == 1, multimodal_xg_test_LUSC$censor_time, -multimodal_xg_test_LUSC$censor_time)
multimodal_test_xgboost_LUSC <- multimodal_xg_test_LUSC %>% dplyr::select(-c(vital_status,censor_time))

#Testing Data
multimodal_xg_test_LUAD <- multimodal_test_LUAD
multimodal_xg_test_LUAD$vital_status <- as.integer(multimodal_xg_test_LUAD$vital_status)
test_label <- ifelse(multimodal_xg_test_LUAD$vital_status == 1, multimodal_xg_test_LUAD$censor_time, -multimodal_xg_test_LUAD$censor_time)
multimodal_test_xgboost_LUAD <- multimodal_xg_test_LUAD %>% dplyr::select(-c(vital_status,censor_time))


#Cancer Specific - LUSC
xg_overfit_preds_LUSC <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUSC))
xg_testingdata_LUSC <- Cindex(xg_overfit_preds_LUSC, Surv(multimodal_xg_test_LUSC$censor_time,multimodal_xg_test_LUSC$vital_status))

#LUAD
xg_overfit_preds_LUAD <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUAD))
xg_testingdata_LUAD <- Cindex(xg_overfit_preds_LUAD, Surv(multimodal_xg_test_LUAD$censor_time,multimodal_xg_test_LUAD$vital_status))



elastic_res <- glm_c_testingdata
cox_res <- cox_testingdata
xg_res <- C_xg_overfit
rsf_res <- rf2_cindex_testingdata

Cindex_Test <- c(xg_res, cox_res, elastic_res, rsf_res)
Models <- c("XGBoost", "Cox", "ElasticNet", "RSF")
Data_Types <- rep("Multi_Selection", 4)
results.df <- as.data.frame(cbind(Models, Cindex_Test))
colnames(results.df)[2] <- "Cindex_Test"
results.df <- cbind(results.df, Data_Types)
results.df$Cindex_Test <- as.numeric(results.df$Cindex_Test)
Dimension_Reduction <- rep(autoencoder_type, 4)
results.df <- cbind(results.df, Dimension_Reduction)

Cindex_LUSC <- c(xg_testingdata_LUSC, cox_testingdata_LUSC, elastic_testingdata_LUSC, rf_testingdata_LUSC) 
Cindex_LUAD <- c(xg_testingdata_LUAD, cox_testingdata_LUAD, elastic_testingdata_LUAD, rf_testingdata_LUAD)
results.df <- as.data.frame(cbind(results.df, Cindex_LUSC))
results.df <- as.data.frame(cbind(results.df, Cindex_LUAD))

results.df$Cindex_Test <-round(results.df$Cindex_Test, 3)
results.df_multimodal_featselection <- results.df 
}
```

### Multimodal PCA Comparison

```{r, echo=FALSE}
if (PCA == "yes") {
#all generated earlier
mirna_train_pca <- mirna_train_pca %>% dplyr::select(-c(vital_status, censor_time))
gene_train_pca  <- gene_train_pca %>% dplyr::select(-c(vital_status, censor_time))
lnc_train_pca <- lnc_train_pca %>% dplyr::select(-c(vital_status, censor_time))
meth_train_pca <- meth_train_pca %>% dplyr::select(-c(vital_status, censor_time))

mirna_test_pca <- mirna_test_pca %>% dplyr::select(-c(vital_status, censor_time))
gene_test_pca <- gene_test_pca %>% dplyr::select(-c(vital_status, censor_time))
lnc_test_pca <- lnc_test_pca %>% dplyr::select(-c(vital_status, censor_time))
meth_test_pca <- meth_test_pca %>% dplyr::select(-c(vital_status, censor_time))

#Training Data Combination
index <- which(colnames(clin_train)=="censor_time1")
colnames(clin_train)[index] <- "censor_time"
  
df2 <- merge(lnc_train_pca, mirna_train_pca, by="barcode")
df2 <- merge(df2, gene_train_pca, by="barcode")
df2 <- merge(df2, meth_train_pca, by="barcode")
multimodal_train <- merge(df2,clin_train, by="barcode")

#Testing Data Combination
index <- which(colnames(clin_test)=="censor_time1")
colnames(clin_test)[index] <- "censor_time"

df3 <- merge(lnc_test_pca, mirna_test_pca, by ="barcode")
df3 <- merge(df3, gene_test_pca, by="barcode")
df3 <- merge(df3, meth_test_pca, by="barcode")
multimodal_test <- merge(df3, clin_test,by="barcode")

multimodal_test <- multimodal_test %>% dplyr::select(-c(barcode))
multimodal_train <- multimodal_train %>% dplyr::select(-c(barcode))

multimodal_test <- as.data.frame(sapply(multimodal_test, as.numeric))
multimodal_train <- as.data.frame(sapply(multimodal_train, as.numeric))

#Overall Model
cox <- coxph(Surv(censor_time, vital_status) ~ ., data=multimodal_train)
cox_fit <- survfit(cox)
cox_overfit_preds <- predict(cox, multimodal_test)
cox_testingdata <- Cindex(cox_overfit_preds, Surv(multimodal_test$censor_time,multimodal_test$vital_status))

multimodal_train$vital_status <- multimodal_train$vital_status - 1
multimodal_test$vital_status <- multimodal_test$vital_status-1

rfsrc.pbc_plot <- rfsrc(Surv(censor_time, vital_status) ~ ., # survival object
                  data = multimodal_train, # data
                  nsplit = 1, # number of random splits to consider
                  tree.err = FALSE
                  ,importance = TRUE
                  )
rfsrc.pbc.testplot <- predict(rfsrc.pbc_plot,multimodal_test)
rf2_cindex_testingdata <- Cindex(rfsrc.pbc.testplot$predicted, Surv(multimodal_test$censor_time,multimodal_test$vital_status))

multimodal_cv <-as.matrix(multimodal_train[1:nrow(multimodal_train), 1:(ncol(multimodal_train)-2)])
multimodal_cv_test <-as.matrix(multimodal_test[1:nrow(multimodal_test), 1:(ncol(multimodal_test)-2)])

cv.fit = cv.glmnet(multimodal_cv,  Surv(multimodal_train$censor_time,multimodal_train$vital_status),
                   alpha =0.5, # lasso: alpha = 1; ridge: alpha=0
                   family = "cox", maxit = 10000, type.measure = "C")
est.coef = coef(cv.fit, s = cv.fit$lambda.min) # returns the p length coefficient vector
active.k = which(est.coef != 0)

glm_pred <- predict(cv.fit, multimodal_cv_test)
glm_c_testingdata <- Cindex(glm_pred,Surv(multimodal_test$censor_time,multimodal_test$vital_status))

library(survXgboost)
library(xgboost)

multimodal_xg <- multimodal_train
multimodal_xg$vital_status <- as.integer(multimodal_xg$vital_status)
label <- ifelse(multimodal_xg$vital_status == 1, multimodal_xg$censor_time, -multimodal_xg$censor_time)

matrix_xg <- as.matrix(multimodal_xg[1:nrow(multimodal_xg), 1:(ncol(multimodal_xg)-2)])

prediction_df <- multimodal_xg[1:nrow(multimodal_train),
                               (ncol(multimodal_train)-1):ncol(multimodal_train)]

#Overall for Plot:
val_ind <- sample.int(nrow(multimodal_xg), 0.1 * nrow(multimodal_xg))
x_train <- as.matrix(multimodal_xg [-val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")])

x_label <- label[-val_ind]
x_val <- xgb.DMatrix(as.matrix(multimodal_xg[val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")]), label = label[val_ind])

#Testing Data
multimodal_xg_test <- multimodal_test
multimodal_xg_test$vital_status <- as.integer(multimodal_xg_test$vital_status)
test_label <- ifelse(multimodal_xg_test$vital_status == 1, multimodal_xg_test$censor_time, -multimodal_xg_test$censor_time)
multimodal_test_xgboost <- multimodal_xg_test %>% dplyr::select(-c(vital_status,censor_time))

surv_xgboost_model_all <- xgb.train.surv(
  params = list(
    objective = "survival:cox",
    eval_metric = "cox-nloglik",
    eta = 0.001, subsample=.5,
               max_depth=2
  ), data = matrix_xg, label = label,
  watchlist = list(val3 = x_val),
  nrounds = 50
  , early_stopping_rounds = 15

)
# predict survival curves
xg_df = predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost))
C_xg_overfit <- Cindex(xg_df, Surv(multimodal_xg_test$censor_time, multimodal_xg_test$vital_status))

#Examining risk scores (should be somewhat spread out)
risk_scores <- predict(object = surv_xgboost_model_all, newdata = as.matrix(multimodal_test_xgboost), type = "risk")
hist(risk_scores)

#Not overfit
elastic_res <- glm_c_testingdata
cox_res <- cox_testingdata
xg_res <- C_xg_overfit
rsf_res <- rf2_cindex_testingdata

Cindex_Test <- c(xg_res, cox_res, elastic_res, rsf_res)
Models <- c("XGBoost", "Cox", "ElasticNet", "RSF")
Data_Types <- rep("Multimodal", 5)
results.df <- as.data.frame(cbind(Models, Cindex_Test))
colnames(results.df)[2] <- "Cindex_Test"
results.df <- cbind(results.df, Data_Types)
results.df$Cindex_Test <- as.numeric(results.df$Cindex_Test)
Dimension_Reduction <- rep("PCA", 5)
results.df <- cbind(results.df, Dimension_Reduction)

results.df$Cindex_Test <-round(results.df$Cindex_Test, 3)
results.df_multimodal_pca <- results.df 
results.df_multimodal_pca
}
```

### Are separate autoencoders more effective than one large one?

```{r,echo=FALSE}
if (full_AE=="yes") {

#Training Data - all generated earlier on the in the pipeline in case you wanted to do this analysis
df3 <- merge(lncrna_fullAE, mrna_fullAE, by="barcode")
df3 <- merge(df3, mirna_fullAE, by="barcode")
fullAE <- merge(df3, meth_fullAE, by="barcode")

#Testing Data
df4 <- merge(lncrna_fullAE_test, mrna_fullAE_test, by="barcode")
df4 <- merge(df4, mirna_fullAE_test, by="barcode")
fullAE_test <- merge(df4, meth_fullAE_test, by="barcode")

#Save vital status and outcome into a dataframe
fullAE_outcome <- fullAE %>% dplyr::select(c(barcode))
fullAE_outcome_test <- fullAE_test %>% dplyr::select(c(barcode))

full_LUSC_indices <- merge(fullAE, barcode_disease_mapping, by="barcode")
full_LUAD_indices_train <- which(full_LUSC_indices$name=="Lung Adenocarcinoma") 
full_LUSC_indices_train <- which(full_LUSC_indices$name=="Lung Squamous Cell Carcinoma")

full_LUSC_indices_test <- merge(fullAE_test, barcode_disease_mapping, by="barcode")
full_LUAD_indices_test <- which(full_LUSC_indices_test$name=="Lung Adenocarcinoma") 
full_LUSC_indices_test <- which(full_LUSC_indices_test$name=="Lung Squamous Cell Carcinoma")

fullAE <- fullAE %>% dplyr::select(-c(barcode))
fullAE_test <- fullAE_test %>% dplyr::select(-c(barcode))

fullAE <- as.data.frame(sapply(fullAE, as.numeric))
fullAE_test <- as.data.frame(sapply(fullAE_test, as.numeric))

fullAE_test <- as.matrix(fullAE_test)

fullAE_test <- scale(fullAE_test)
fullAE <- scale(fullAE)

if (autoencoder_type == "stacked_denoising_zeros") {
set.seed(123)

# one-of corruption
corrupt_with_ones <- function(x) {
  n_to_sample <- floor(length(x) * .3)
  elements_to_corrupt <- sample(seq_along(x), n_to_sample, replace = FALSE)
  x[elements_to_corrupt] <- 0
  return(x)
}

inputs_currupted_ones <- fullAE %>%
  as.data.frame() %>%
  purrr::map_df(corrupt_with_ones)

features <- as.matrix(fullAE)
inputs_currupted_ones <- as.matrix(inputs_currupted_ones)

model1 <- keras_model_sequential()
model1 %>%
  layer_dense(units = fullAE_features, activation = "tanh", name = "BottleNeck",input_shape=ncol(inputs_currupted_ones)) %>%
  layer_dense(units = ncol(inputs_currupted_ones))

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

history5 <- model1 %>% keras::fit(
  x = inputs_currupted_ones,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)
plot(history5)

mse.ae1 <- evaluate(model1, features,inputs_currupted_ones)

intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_multimodal_auto <- predict(intermediate_layer_model1, inputs_currupted_ones)

first_AE_RMSE <- evaluate(model1,fullAE_test, fullAE_test)
LUSC_multimodal_auto_output <- predict(intermediate_layer_model1,fullAE_test)

}
}
```

### Testing Different Autoencoder Types - Full Autoencoder

```{r, echo=FALSE}
if (full_AE == "yes") {

if (autoencoder_type == "sparse") {
  
features <- as.matrix(fullAE)

model1 <- keras_model_sequential()
model1 %>%
  layer_dense(units = fullAE_features, activation = "relu", name = "BottleNeck", input_shape = ncol(features), activity_regularizer=regularizer_l1(10e-5)) %>%
  layer_dense(units = ncol(features), activation='sigmoid')


model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,features)


intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_multimodal_auto <- predict(intermediate_layer_model1, features)

first_AE_RMSE <- evaluate(model1,fullAE_test, fullAE_test)
LUSC_multimodal_auto_output <- predict(intermediate_layer_model1,fullAE_test)
}


if (autoencoder_type == "variational") {

features <- as.matrix(fullAE)
  
enc_input <- ncol(features)
latent_size <- fullAE_features
epsilon_std <- 1.0

enc_input <- layer_input(shape = c(ncol(features)))
layer_one <- layer_dense(enc_input, units=256, activation = "relu")
z_mean <- layer_dense(layer_one, latent_size, name = "BottleNeck")
z_log_var <- layer_dense(layer_one, latent_size)
 
encoder <- keras_model(enc_input, z_mean)
summary(encoder)


sampling <- function(arg){
  z_mean <- arg[, 1:(latent_size)]
  z_log_var <- arg[, (latent_size + 1):(2 * latent_size)]
  epsilon <- k_random_normal(shape = c(k_shape(z_mean)[[1]]), mean=0)
  z_mean + k_exp(z_log_var/2)*epsilon
}

z <- layer_concatenate(list(z_mean, z_log_var)) %>% 
  layer_lambda(sampling)


decoder_layer <- layer_dense(units = 256, activation = "relu")
decoder_mean <- layer_dense(units = ncol(features), activation = "sigmoid")
h_decoded <- decoder_layer(z)
x_decoded_mean <- decoder_mean(h_decoded)
 
model1 <- keras_model(enc_input, x_decoded_mean)

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% fit(
  features, features, 
  epochs = 100, 
  validation_split=0.1
)


mse.ae1 <- evaluate(model1, features,features)
intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_multimodal_auto <- predict(intermediate_layer_model1, features)
first_AE_RMSE <- evaluate(model1,fullAE_test, fullAE_test)
LUSC_multimodal_auto_output <- predict(intermediate_layer_model1,fullAE_test)
}

if (autoencoder_type=="denoising_gaussian") {

features <- as.matrix(fullAE)

model1 <- keras_model_sequential()
model1 %>%
  layer_gaussian_noise(stddev = 1, input_shape = ncol(features)) %>%
  layer_dense(units = fullAE_features, activation = "tanh", name = "BottleNeck") %>%
  layer_dense(units = ncol(features))

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,features)
intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_multimodal_auto <- predict(intermediate_layer_model1, features)
first_AE_RMSE <- evaluate(model1,fullAE_test, fullAE_test)
LUSC_multimodal_auto_output <- predict(intermediate_layer_model1,fullAE_test)
}

if (autoencoder_type=="basic") {

features <- as.matrix(fullAE)

model1 <- keras_model_sequential()
model1 %>%
  layer_dense(units = fullAE_features, activation = "tanh", name = "BottleNeck", input_shape = ncol(features)) %>%
  layer_dense(units = ncol(features))

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)

model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,features)
intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_multimodal_auto <- predict(intermediate_layer_model1, features)
first_AE_RMSE <- evaluate(model1,fullAE_test, fullAE_test)
LUSC_multimodal_auto_output <- predict(intermediate_layer_model1,fullAE_test)
}

if (autoencoder_type=="denoising_gaussian_stacked") {

features <- as.matrix(fullAE)

model1 <- keras_model_sequential()
model1 %>%
  layer_gaussian_noise(stddev = 1, input_shape = ncol(features)) %>%
  layer_dense(units = 200, activation = "tanh") %>%
  layer_dense(units = fullAE_features, activation = "tanh", name = "BottleNeck") %>%
  layer_dense(units = 200, activation = "tanh") %>%
  layer_dense(units = ncol(features))

model1 %>% keras::compile(
  loss = "mean_squared_error",
  optimizer = optimizer_adam(lr=0.001)
)
model1 %>% keras::fit(
  x = features,
  y = features,
  epochs = 100,
  validation_split=0.1,
    callbacks = list(
    early_stop)
)

mse.ae1 <- evaluate(model1, features,features)
intermediate_layer_model1 <- keras_model(inputs = model1$input, outputs = get_layer(model1, "BottleNeck")$output)
LUSC_multimodal_auto <- predict(intermediate_layer_model1, features)
first_AE_RMSE <- evaluate(model1,fullAE_test, fullAE_test)
LUSC_multimodal_auto_output <- predict(intermediate_layer_model1,fullAE_test)
}
}
```

### Setting up Full Autoencoded Data

```{r, echo=FALSE} 
if (full_AE == "yes") {
AE_RMSE[5] <- first_AE_RMSE
AE_RMSE_train[5] <- mse.ae1

#incorporating outcome data
LUSC_AE <- as.data.frame(cbind(LUSC_multimodal_auto, fullAE_outcome))
LUSC_AE_test <- as.data.frame(cbind(LUSC_multimodal_auto_output,fullAE_outcome_test))

colnames(LUSC_AE)[ncol(LUSC_AE)] <- "barcode"
colnames(LUSC_AE_test)[ncol(LUSC_AE_test)] <- "barcode"

#incorporating clinical data
LUSC_fullAE <- merge(LUSC_AE, clin_train, by="barcode")
LUSC_fullAE_test <- merge(LUSC_AE_test, clin_test, by="barcode")

LUSC_fullAE_barcode <- LUSC_fullAE$barcode
LUSC_fullAE_barcode_test <- LUSC_fullAE_test$barcode

LUSC_fullAE <- LUSC_fullAE %>% dplyr::select(-c(barcode))
LUSC_fullAE_test <- LUSC_fullAE_test %>% dplyr::select(-c(barcode))

LUSC_fullAE <- as.data.frame(sapply(LUSC_fullAE, as.numeric))
LUSC_fullAE_test <- as.data.frame(sapply(LUSC_fullAE_test, as.numeric))

#Overall Model
cox <- coxph(Surv(censor_time, vital_status) ~ ., data=LUSC_fullAE)
cox_fit <- survfit(cox)
cox_overfit_preds <- predict(cox, LUSC_fullAE_test)
cox_testingdata <- Cindex(cox_overfit_preds, Surv(LUSC_fullAE_test$censor_time,LUSC_fullAE_test$vital_status))

#Cancer Specific - LUSC
multimodal_test_LUSC <- LUSC_fullAE_test[c(full_LUSC_indices_test), ]
cox_overfit_preds_LUSC <- predict(cox, multimodal_test_LUSC)
cox_testingdata_LUSC <- Cindex(cox_overfit_preds_LUSC, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
multimodal_test_LUAD <- LUSC_fullAE_test[c(full_LUAD_indices_test), ]
cox_overfit_preds_LUAD <- predict(cox, multimodal_test_LUAD)
cox_testingdata_LUAD <- Cindex(cox_overfit_preds_LUAD, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))

LUSC_fullAE$vital_status <- LUSC_fullAE$vital_status - 1
LUSC_fullAE_test$vital_status <- LUSC_fullAE_test$vital_status - 1

rfsrc.pbc_plot <- rfsrc(Surv(censor_time, vital_status) ~ ., # survival object
                  data = LUSC_fullAE, # data
                  nsplit = 1, # number of random splits to consider
                  tree.err = FALSE
                  ,importance = TRUE
                  )
rfsrc.pbc.testplot <- predict(rfsrc.pbc_plot,LUSC_fullAE_test)
rf2_cindex_testingdata <- Cindex(rfsrc.pbc.testplot$predicted, Surv(LUSC_fullAE_test$censor_time,LUSC_fullAE_test$vital_status))

multimodal_test_LUAD$vital_status <- multimodal_test_LUAD$vital_status - 1
multimodal_test_LUSC$vital_status <- multimodal_test_LUSC$vital_status - 1

#Cancer Specific - LUSC
rf_overfit_preds_LUSC <- predict(rfsrc.pbc_plot, multimodal_test_LUSC)
rf_testingdata_LUSC <- Cindex(rf_overfit_preds_LUSC$predicted, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
rf_overfit_preds_LUAD <- predict(rfsrc.pbc_plot, multimodal_test_LUAD)
rf_testingdata_LUAD <- Cindex(rf_overfit_preds_LUAD$predicted, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))

multimodal_cv <-as.matrix(LUSC_fullAE[1:nrow(LUSC_fullAE), 1:(ncol(LUSC_fullAE)-2)])
multimodal_cv_test <-as.matrix(LUSC_fullAE_test[1:nrow(LUSC_fullAE_test), 1:(ncol(LUSC_fullAE_test)-2)])

cv.fit = cv.glmnet(multimodal_cv,  Surv(LUSC_fullAE$censor_time,LUSC_fullAE$vital_status),
                   alpha =0.5, # lasso: alpha = 1; ridge: alpha=0
                   family = "cox", maxit = 10000, type.measure = "C")
est.coef = coef(cv.fit, s = cv.fit$lambda.min) 
active.k = which(est.coef != 0)

glm_pred <- predict(cv.fit, multimodal_cv_test)
glm_c_testingdata <- Cindex(glm_pred,Surv(LUSC_fullAE_test$censor_time,LUSC_fullAE_test$vital_status))

#Individual Cancer Types
multimodal_test_LUSC_cv <-as.matrix(multimodal_test_LUSC[1:nrow(multimodal_test_LUSC ), 1:(ncol(multimodal_test_LUSC)-2)])
multimodal_test_LUAD_cv <-as.matrix(multimodal_test_LUAD[1:nrow(multimodal_test_LUAD ), 1:(ncol(multimodal_test_LUAD)-2)])

#Cancer Specific - LUSC
elastic_overfit_preds_LUSC <- predict(cv.fit, multimodal_test_LUSC_cv)
elastic_testingdata_LUSC <- Cindex(elastic_overfit_preds_LUSC, Surv(multimodal_test_LUSC$censor_time,multimodal_test_LUSC$vital_status))

#LUAD
elastic_overfit_preds_LUAD <- predict(cv.fit, multimodal_test_LUAD_cv)
elastic_testingdata_LUAD <- Cindex(elastic_overfit_preds_LUAD, Surv(multimodal_test_LUAD$censor_time,multimodal_test_LUAD$vital_status))

multimodal_xg <- LUSC_fullAE
multimodal_xg$vital_status <- as.integer(multimodal_xg$vital_status)
label <- ifelse(multimodal_xg$vital_status == 1, multimodal_xg$censor_time, -multimodal_xg$censor_time)

matrix_xg <- as.matrix(multimodal_xg[1:nrow(multimodal_xg), 1:(ncol(multimodal_xg)-2)])

prediction_df <- multimodal_xg[1:nrow(LUSC_fullAE),
                               (ncol(LUSC_fullAE)-1):ncol(LUSC_fullAE)]
#Overall for Plot:
val_ind <- sample.int(nrow(multimodal_xg), 0.1 * nrow(multimodal_xg))
x_train <- as.matrix(multimodal_xg [-val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")])

x_label <- label[-val_ind]
x_val <- xgb.DMatrix(as.matrix(multimodal_xg[val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")]), label = label[val_ind])

#Testing Data
multimodal_xg_test <- LUSC_fullAE_test
multimodal_xg_test$vital_status <- as.integer(multimodal_xg_test$vital_status)
test_label <- ifelse(multimodal_xg_test$vital_status == 1, multimodal_xg_test$censor_time, -multimodal_xg_test$censor_time)
multimodal_test_xgboost <- multimodal_xg_test %>% dplyr::select(-c(vital_status,censor_time))

surv_xgboost_model_all <- xgb.train.surv(
  params = list(
    objective = "survival:cox",
    eval_metric = "cox-nloglik",
    eta = 0.001, subsample=.5,
    #,learning_rate=0.001
               max_depth=2
  ), data = matrix_xg, label = label,
  watchlist = list(val3 = x_val),
  nrounds = 50
  , early_stopping_rounds = 15
)

# predict survival curves
xg_df = predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost))
C_xg_overfit <- Cindex(xg_df, Surv(multimodal_xg_test$censor_time, multimodal_xg_test$vital_status))

#Examining risk scores (should be somewhat spread out)
risk_scores <- predict(object = surv_xgboost_model_all, newdata = as.matrix(multimodal_test_xgboost), type = "risk")

#Testing Data - Individual Cancer Types
multimodal_xg_test_LUSC <- multimodal_test_LUSC
multimodal_xg_test_LUSC$vital_status <- as.integer(multimodal_xg_test_LUSC$vital_status)
test_label <- ifelse(multimodal_xg_test_LUSC$vital_status == 1, multimodal_xg_test_LUSC$censor_time, -multimodal_xg_test_LUSC$censor_time)
multimodal_test_xgboost_LUSC <- multimodal_xg_test_LUSC %>% dplyr::select(-c(vital_status,censor_time))

#Testing Data
multimodal_xg_test_LUAD <- multimodal_test_LUAD
multimodal_xg_test_LUAD$vital_status <- as.integer(multimodal_xg_test_LUAD$vital_status)
test_label <- ifelse(multimodal_xg_test_LUAD$vital_status == 1, multimodal_xg_test_LUAD$censor_time, -multimodal_xg_test_LUAD$censor_time)
multimodal_test_xgboost_LUAD <- multimodal_xg_test_LUAD %>% dplyr::select(-c(vital_status,censor_time))

#Cancer Specific - LUSC
xg_overfit_preds_LUSC <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUSC))
xg_testingdata_LUSC <- Cindex(xg_overfit_preds_LUSC, Surv(multimodal_xg_test_LUSC$censor_time,multimodal_xg_test_LUSC$vital_status))

#LUAD
xg_overfit_preds_LUAD <- predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost_LUAD))
xg_testingdata_LUAD <- Cindex(xg_overfit_preds_LUAD, Surv(multimodal_xg_test_LUAD$censor_time,multimodal_xg_test_LUAD$vital_status))

#Not overfit
elastic_res <- glm_c_testingdata
cox_res <- cox_testingdata
xg_res <- C_xg_overfit
rsf_res <- rf2_cindex_testingdata

Cindex_Test <- c(xg_res, cox_res, elastic_res, rsf_res)
Models <- c("XGBoost", "Cox", "ElasticNet", "RSF")
Data_Types <- rep("Full AE", 4)
results.df <- as.data.frame(cbind(Models, Cindex_Test))
colnames(results.df)[2] <- "Cindex_Test"
results.df <- cbind(results.df, Data_Types)
results.df$Cindex_Test <- as.numeric(results.df$Cindex_Test)
Dimension_Reduction <- rep(autoencoder_type, 4)
results.df <- cbind(results.df, Dimension_Reduction)

Cindex_LUSC <- c(xg_testingdata_LUSC, cox_testingdata_LUSC, elastic_testingdata_LUSC, rf_testingdata_LUSC) 
Cindex_LUAD <- c(xg_testingdata_LUAD, cox_testingdata_LUAD, elastic_testingdata_LUAD, rf_testingdata_LUAD)
results.df <- as.data.frame(cbind(results.df, Cindex_LUSC))
results.df <- as.data.frame(cbind(results.df, Cindex_LUAD))

results.df$Cindex_Test <-round(results.df$Cindex_Test, 3)
results.df_fullAE <- results.df
}
```

# VISUALIZATIONS

## Stratified by Survival Status - 5 Year Survival

```{r, echo=FALSE}
if (full_AE == "yes") {
LUSC_fullAE <- as.data.frame(cbind(LUSC_fullAE, LUSC_fullAE_barcode))
LUSC_fullAE_test <- as.data.frame(cbind(LUSC_fullAE_test, LUSC_fullAE_barcode_test))

colnames(LUSC_fullAE)[ncol(LUSC_fullAE)] <- "barcode"
colnames(LUSC_fullAE_test)[ncol(LUSC_fullAE_test)] <- "barcode"

multimodal_prediction <- rbind(LUSC_fullAE, LUSC_fullAE_test)
X <- multimodal_prediction
X <- merge(X, barcode_disease_mapping, by="barcode")


X$vital_status <- as.factor(X$vital_status)
levels(X$vital_status) <- c("Alive", "Dead")

X_deceased <- X %>% filter(vital_status=="Dead" & censor_time < 1825)
X_alive_5years <- X %>% filter(censor_time > 1825)
X_alive_5years$vital_status <- "Alive"

X <- rbind(X_deceased, X_alive_5years)

disease_vital <- NULL
for (i in 1:nrow(X)) {
  if (X$vital_status[i]=="Dead" & X$name.y[i]=="Lung Squamous Cell Carcinoma") {
  disease_vital[i] <- "LUSC - Dead"
}
 if (X$vital_status[i]=="Alive" & X$name.y[i]=="Lung Squamous Cell Carcinoma") {
  disease_vital[i] <- "LUSC - Alive"
}
 if (X$vital_status[i]=="Dead" & X$name.y[i]=="Lung Adenocarcinoma") {
  disease_vital[i] <- "LUAD - Dead"
}
 if (X$vital_status[i]=="Alive" & X$name.y[i]=="Lung Adenocarcinoma") {
  disease_vital[i] <- "LUAD - Alive"
}
}

X <- as.data.frame(cbind(X, disease_vital))
vital_status_vector_FULLAE <- X$vital_status
disease_vitalstatus <- X$disease_vital

X <- X %>%
  dplyr::select(-c(vital_status, censor_time, barcode, disease_vital, name.y, volume))

X <- as.data.frame(X[,2:121])

LUSC_PCA_FULLAE <- prcomp(X)

ggplot(as.data.frame(LUSC_PCA_FULLAE$x), aes(x = PC1, y = PC2, col = factor(disease_vitalstatus))) + geom_point() + scale_color_discrete(name = "Survival Status") + labs(title="5 Year Survival - Full AE")

set.seed(1234) # reproducibility
tsne <- Rtsne(X, dims = 2, perplexity = 10)
t.df_FULLAE <- as.data.frame(tsne$Y)
colnames(t.df_FULLAE) <- c("V1", "V2")
ggplot(t.df_FULLAE, aes(x = V1, y = V2, color = factor(disease_vitalstatus))) + geom_point()+ labs(title="Full AE - tSNE") + scale_color_discrete(name = "Survival Status")
}
```

### Full Multimodal Data with PCA instead of Autoencoder

```{r,echo=FALSE}
if (PCA == "yes") {
if (full_AE=="yes") {

#Training Data
df3 <- merge(lncrna_fullAE, mrna_fullAE, by="barcode")
df3 <- merge(df3, mirna_fullAE, by="barcode")
fullAE <- merge(df3, meth_fullAE, by="barcode")

#Testing Data
df4 <- merge(lncrna_fullAE_test, mrna_fullAE_test, by="barcode")
df4 <- merge(df4, mirna_fullAE_test, by="barcode")
fullAE_test <- merge(df4, meth_fullAE_test, by="barcode")

#Save vital status and outcome into a dataframe
fullAE_outcome <- fullAE %>% dplyr::select(c(barcode))
fullAE_outcome_test <- fullAE_test %>% dplyr::select(c(barcode))

fullAE <- fullAE %>% dplyr::select(-c(barcode))
fullAE_test <- fullAE_test %>% dplyr::select(-c(barcode))

fullAE <- as.data.frame(sapply(fullAE, as.numeric))
fullAE_test <- as.data.frame(sapply(fullAE_test, as.numeric))

fullAE_test <- as.matrix(fullAE_test)

fullAE_test <- scale(fullAE_test)
fullAE <- scale(fullAE)


#PCA 
#Training Data
pc <- prcomp(fullAE) 
multimodal_pc <- as.data.frame(pc$x) %>% dplyr::select(PC1:PC120)

#Testing Set
pc_test <- predict(pc, newdata = fullAE_test)
multimodal_pc_test <- as.data.frame(pc_test) %>% dplyr::select(PC1:PC120)

PCA_RMSE_COMBINED <- pca.recon(pc, fullAE, 120)$rmse
PCA_RMSE_COMBINED_TEST <- pca.recon_test(pc_test, fullAE_test)$rmse

}
}
```

### Setting up Full Autoencoded Data - PCA

```{r, echo=FALSE} 
if (PCA == "yes") {
if (full_AE == "yes") {


#incorporating outcome data
LUSC_AE <- as.data.frame(cbind(multimodal_pc, fullAE_outcome))
LUSC_AE_test <- as.data.frame(cbind(multimodal_pc_test,fullAE_outcome_test))

colnames(LUSC_AE)[ncol(LUSC_AE)] <- "barcode"
colnames(LUSC_AE_test)[ncol(LUSC_AE_test)] <- "barcode"

#incorporating clinical data
LUSC_fullAE <- merge(LUSC_AE, clin_train, by="barcode")
LUSC_fullAE_test <- merge(LUSC_AE_test, clin_test, by="barcode")

LUSC_fullAE <- LUSC_fullAE %>% dplyr::select(-c(barcode))
LUSC_fullAE_test <- LUSC_fullAE_test %>% dplyr::select(-c(barcode))

LUSC_fullAE <- as.data.frame(sapply(LUSC_fullAE, as.numeric))
LUSC_fullAE_test <- as.data.frame(sapply(LUSC_fullAE_test, as.numeric))

#Overall Model
cox <- coxph(Surv(censor_time, vital_status) ~ ., data=LUSC_fullAE)
cox_fit <- survfit(cox)
cox_overfit_preds <- predict(cox, LUSC_fullAE_test)
cox_testingdata <- Cindex(cox_overfit_preds, Surv(LUSC_fullAE_test$censor_time,LUSC_fullAE_test$vital_status))

LUSC_fullAE$vital_status <- LUSC_fullAE$vital_status - 1
LUSC_fullAE_test$vital_status <- LUSC_fullAE_test$vital_status - 1

rfsrc.pbc_plot <- rfsrc(Surv(censor_time, vital_status) ~ ., # survival object
                  data = LUSC_fullAE, # data
                  nsplit = 1, # number of random splits to consider
                  tree.err = FALSE
                  ,importance = TRUE
                  )
rfsrc.pbc.testplot <- predict(rfsrc.pbc_plot,LUSC_fullAE_test)
rf2_cindex_testingdata <- Cindex(rfsrc.pbc.testplot$predicted, Surv(LUSC_fullAE_test$censor_time,LUSC_fullAE_test$vital_status))


multimodal_cv <-as.matrix(LUSC_fullAE[1:nrow(LUSC_fullAE), 1:(ncol(LUSC_fullAE)-2)])
multimodal_cv_test <-as.matrix(LUSC_fullAE_test[1:nrow(LUSC_fullAE_test), 1:(ncol(LUSC_fullAE_test)-2)])

cv.fit = cv.glmnet(multimodal_cv,  Surv(LUSC_fullAE$censor_time,LUSC_fullAE$vital_status),
                   alpha =0.5, # lasso: alpha = 1; ridge: alpha=0
                   family = "cox", maxit = 10000, type.measure = "C")
#plot(cv.fit)
est.coef = coef(cv.fit, s = cv.fit$lambda.min) # returns the p length coefficient vector
active.k = which(est.coef != 0)

glm_pred <- predict(cv.fit, multimodal_cv_test)
glm_c_testingdata <- Cindex(glm_pred,Surv(LUSC_fullAE_test$censor_time,LUSC_fullAE_test$vital_status))

multimodal_xg <- LUSC_fullAE
multimodal_xg$vital_status <- as.integer(multimodal_xg$vital_status)
label <- ifelse(multimodal_xg$vital_status == 1, multimodal_xg$censor_time, -multimodal_xg$censor_time)

matrix_xg <- as.matrix(multimodal_xg[1:nrow(multimodal_xg), 1:(ncol(multimodal_xg)-2)])

prediction_df <- multimodal_xg[1:nrow(LUSC_fullAE),
                               (ncol(LUSC_fullAE)-1):ncol(LUSC_fullAE)]

#Overall for Plot:
val_ind <- sample.int(nrow(multimodal_xg), 0.1 * nrow(multimodal_xg))
x_train <- as.matrix(multimodal_xg [-val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")])

x_label <- label[-val_ind]
x_val <- xgb.DMatrix(as.matrix(multimodal_xg[val_ind, !names(multimodal_xg) %in% c("censor_time", "vital_status")]), label = label[val_ind])

#Testing Data
multimodal_xg_test <- LUSC_fullAE_test
multimodal_xg_test$vital_status <- as.integer(multimodal_xg_test$vital_status)
test_label <- ifelse(multimodal_xg_test$vital_status == 1, multimodal_xg_test$censor_time, -multimodal_xg_test$censor_time)
multimodal_test_xgboost <- multimodal_xg_test %>% dplyr::select(-c(vital_status,censor_time))

surv_xgboost_model_all <- xgb.train.surv(
  params = list(
    objective = "survival:cox",
    eval_metric = "cox-nloglik",
    eta = 0.001, subsample=.5,
               max_depth=2
  ), data = matrix_xg, label = label,
  watchlist = list(val3 = x_val),
  nrounds = 50
  , early_stopping_rounds = 15

)

# predict survival curves
xg_df = predict(surv_xgboost_model_all, as.matrix(multimodal_test_xgboost))
C_xg_overfit <- Cindex(xg_df, Surv(multimodal_xg_test$censor_time, multimodal_xg_test$vital_status))

#Examining risk scores (should be somewhat spread out)
risk_scores <- predict(object = surv_xgboost_model_all, newdata = as.matrix(multimodal_test_xgboost), type = "risk")

#Testing Data - Individual Cancer Types
multimodal_xg_test_LUSC <- multimodal_test_LUSC
multimodal_xg_test_LUSC$vital_status <- as.integer(multimodal_xg_test_LUSC$vital_status)
test_label <- ifelse(multimodal_xg_test_LUSC$vital_status == 1, multimodal_xg_test_LUSC$censor_time, -multimodal_xg_test_LUSC$censor_time)
multimodal_test_xgboost_LUSC <- multimodal_xg_test_LUSC %>% dplyr::select(-c(vital_status,censor_time))

#Not overfit
elastic_res <- glm_c_testingdata
cox_res <- cox_testingdata
xg_res <- C_xg_overfit
rsf_res <- rf2_cindex_testingdata

Cindex_Test <- c(xg_res, cox_res, elastic_res, rsf_res)
Models <- c("XGBoost", "Cox", "ElasticNet", "RSF")
Data_Types <- rep("Full AE", 5)
results.df <- as.data.frame(cbind(Models, Cindex_Test))
colnames(results.df)[2] <- "Cindex_Test"
results.df <- cbind(results.df, Data_Types)
results.df$Cindex_Test <- as.numeric(results.df$Cindex_Test)
Dimension_Reduction <- rep("PCA", 5)
results.df <- cbind(results.df, Dimension_Reduction)

results.df$Cindex_Test <-round(results.df$Cindex_Test, 3)
results.df_fullAE_PCA <- results.df
}
}
```

# Overall Plot of Different Modalities - Post Processing Data

```{r,echo=FALSE}
std <- function(x) sd(x)/sqrt(length(x))

df1 <- rbind(results.df_gene, results.df_meth)
df1 <- rbind(df1, results.df_lncs)
df1 <- rbind(df1, results.df_mirna)
df1 <- rbind(df1, results.df_clin)

if (another_round_of_featureselection == "yes") {
 df1 <- rbind(df1, results.df_multimodal_featselection)
}
if (full_AE=="yes") {
df1 <- rbind(df1,results.df_fullAE)
}

if(PCA == "yes") {
  
df4 <- rbind(results.df_lncs_pca,results.df_mirna_pca)
df4 <- rbind(df4,results.df_meth_pca)
df4 <- rbind(df4,results.df_multimodal_pca)
df4 <- rbind(df4,results.df_gene_pca)
df4 <- df4 %>% mutate(Cindex_LUSC=0)
df4 <- df4 %>% mutate(Cindex_LUAD=0)
  
df1 <- rbind(df1, df4)

if (full_AE=="yes") {
results.df_fullAE_PCA <- results.df_fullAE_PCA %>% mutate(Cindex_LUAD=0)
results.df_fullAE_PCA <- results.df_fullAE_PCA %>% mutate(Cindex_LUSC=0)
df1 <- rbind(df1,results.df_fullAE_PCA)

}
}
  
df1 <- rbind(df1, results.df_multimodal_noclin)
overall_results <- rbind(df1, results.df_multimodal)
overall_results <- overall_results %>% filter(Models!="DeepSurv")

#Grouped by Model Type
overall_plot <- overall_results %>%
  filter(Dimension_Reduction!="PCA") %>%
  group_by(Data_Types) %>%
  summarise(C_Index = mean(Cindex_Test), sd=std(Cindex_Test))

surv_predictions <- overall_plot %>%
  ggplot(aes(x=reorder(Data_Types, C_Index) , y=C_Index)) + geom_point() + geom_errorbar(aes(ymin=C_Index-sd,ymax=C_Index+sd)) + labs(x="Modalities", y="C Index", title="Survival Predictions of Biological Modalities")

overall_plot %>%
  ggplot(aes(x=reorder(Data_Types, C_Index) , y=C_Index)) + geom_point() + geom_errorbar(aes(ymin=C_Index-sd,ymax=C_Index+sd)) + labs(x="Modalities", y="C Index", title="Survival Predictions of Biological Modalities")
```

## Saving Results from Each Run

```{r, echo=FALSE}
#Saving Results in New Directory
dir.create(paste0("~/desktop/Multimodal/MultimodalPipeline_Results/", runinfo, "/"))

# Save Overall Table of all Results
write.csv(overall_results, file=paste0("~/desktop/Multimodal/MultimodalPipeline_Results/", runinfo, "/overall_results_run", run_number,".csv"))

# Save .png of modality comparison
ggsave(filename= paste0("~/desktop/Multimodal/MultimodalPipeline_Results/", runinfo, "/ModalityComparison", run_number ,".png"), plot=surv_predictions)

#Run Information Saving for Reference
vec1 <- c(n_lncs, ngenes, n_mirna, nprobes, n_clinical, linear_feature_selection, p_value, autoencoder_type, meth_features, gene_features, mirna_features, lnc_features, PCA)
vec2 <- c("LNC RNAs", "Genes", "miRNAs", "Meth Probes", "Clinical", "linear feature selection", "P value", "Autoencoder Choice", "Methylation Features","Gene Features", "miRNA Features", "LNC Features", "PCA")
info <- as.data.frame(cbind(vec2, vec1))

write.csv(info, file=paste0("~/desktop/Multimodal/MultimodalPipeline_Results/", runinfo, "/RunInformation.csv"))

#Saving Autoencoder Info:
#PCA Performance
Modalities <- c("LNC_RNA", "mRNA", "miRNA", "Methylation", "Combined")

if (linear_feature_selection != "yes") {
if (PCA=="yes") {
pca_train_info <- c(PCA_RMSE_LNC, PCA_RMSE_GENE, PCA_RMSE_MIRNA, PCA_RMSE_METH, PCA_RMSE_COMBINED)
pca_test_info  <- c(PCA_RMSE_LNC_TEST, PCA_RMSE_GENE_TEST, PCA_RMSE_MIRNA_TEST, PCA_RMSE_METH_TEST, PCA_RMSE_COMBINED_TEST)
PCA_Info <- as.data.frame(cbind(pca_test_info, pca_train_info))
colnames(PCA_Info) <- c("Test_RMSE", "Train_RMSE")
PCA_Info <- as.data.frame(cbind(PCA_Info, Modalities))
write.csv(PCA_Info, file=paste0("~/desktop/Multimodal/MultimodalPipeline_Results/", runinfo, "/PCA_Information.csv"))
}
AE_Info <- as.data.frame(cbind(AE_RMSE, AE_RMSE_train))
colnames(AE_Info) <- c("Test_RMSE", "Train_RMSE")
AE_Info <- as.data.frame(cbind(AE_Info, Modalities))
write.csv(AE_Info, file=paste0("~/desktop/Multimodal/MultimodalPipeline_Results/", runinfo, "/AE_Information.csv"))
}

#Creating new directories for visualizations
dir.create(paste0("~/desktop/Multimodal/MultimodalPipeline_Results/", runinfo, "/pics/"))

ggsave(filename=paste0("~/desktop/Multimodal/MultimodalPipeline_Results/", runinfo, "/pics/Multimodal_FeatureImportance.png"), plot = variable_importance)
```
